{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# plottingNewAndOldReads.ipynb\n",
    "## Marcus Viscardi,    January 16, 2023\n",
    "\n",
    "This script is going to take the methods from finalizingReadAndCoveragePlotting_matplotlib.py and use those to plot the new reads!"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.insert(0, '/data16/marcus/scripts/nanoporePipelineScripts')\n",
    "import nanoporePipelineCommon as npCommon\n",
    "from nanoporeReadPlotting.finalizingReadAndCoveragePlotting_matplotlib import plot_reads, coverage_plotting_5tera\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "print(f\"Imports done at {npCommon.get_dt(for_print=True)}\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "try:\n",
    "    reads_df_genes = pd.read_parquet(npCommon.find_newest_matching_file(f\"./output_files/*_quad5TERA.reads_df.parquet\"))\n",
    "    compressed_df_genes = pd.read_parquet(npCommon.find_newest_matching_file(f\"./output_files/*_quad5TERA.compressed_df.parquet\"))\n",
    "    conversion_dict = {\"xrn-1-5tera\": \"oldN2\",\n",
    "                       \"xrn-1-5tera-smg-6\": \"oldS6\",\n",
    "                       \"5tera_xrn-1-KD_wt\": \"newN2\",\n",
    "                       \"5tera_xrn-1-KD_smg-5\": \"newS5\",\n",
    "                       \"5tera_xrn-1-KD_smg-6\": \"newS6\",\n",
    "                       \"5tera_xrn-1-KD_smg-7\": \"newS7\",\n",
    "                       \"sPM57\": \"sPM57\",\n",
    "                       \"sPM58\": \"sPM58\",\n",
    "                       }\n",
    "    reads_df_genes.replace(conversion_dict, inplace=True)\n",
    "    compressed_df_genes.replace(conversion_dict, inplace=True)\n",
    "    print(f\"Finished library loading at {npCommon.get_dt(for_print=True)}\")\n",
    "except ValueError:\n",
    "    reads_df_genes, compressed_df_genes = None, None\n",
    "    print(\"Could not find pre-compressed dataframes saved. Try running the first few cells of initialTestingAndScratchPaper.ipynb\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Plotting of reads and coverage for specific genes\n",
    "Some genes of interest:\n",
    "- ubl-1\n",
    "  - ubl-1 has been a classic gene for me. It has one NMD sensitive isoform that is obviously stabilized with the inclusion of NMD-breaking alleles\n",
    "- rpl-30\n",
    "  - rpl-30 is pretty interesting as it seems to have one NMD sensitive isoform that is related to a longer 5' UTR region. Maybe this is due to a uORF? This gene might also be one of the annotated NMD targets in elegans.\n",
    "  - This isoform is identified as Y106G6H.3c, its is annotated as protein coding, but the cleavages all seem to be in the 5'UTR! very likely uORF, I need to check the literature on the NMD-sensitivity of this allele!\n",
    "- rpl-7A\n",
    "  - Alt. 5'SS in the third intron produces a NMD sensitive isoform that is stabilized by NMD inhibition. This is a \"known\" NMD target in worms.\n",
    "- F19B2.5\n",
    "  - Seems like there is only one isoform that is NMD sensitive for F19B2.5, interestingly this gene shows a pretty dramatic increase in RPM with NMD-breaking alleles.\n",
    "- rpl-3\n",
    "  - This gene has an alt. 3'SS in the first intron that seems to produce an NMD sensitive isoform. It is stabilized with NMD inhibition. This isoform is F13B10.2b, which is annotated as having no CDS and being non-productive.\n",
    "- Y73B3A.18\n",
    "- rpl-26\n",
    "  - Has a really obvious NMD targeted isoform that we can zoom in on"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "parissas_parsed_gtf_path = \"/data16/parissa/genomes/221213_210414_srf1004/210414_srf1004Genome_221213mod.gtf.parquet\"\n",
    "\n",
    "target_genes_plus = [  # This now also includes the divide factors and the adapted adjustment factors\n",
    "                     # ('odc-1', 50, 5),\n",
    "                     # ('F19B2.5', 50, 5),\n",
    "                     # ('rpl-3', 250, 5),\n",
    "                     # ('rpl-7A', 200, 5),\n",
    "                     # ('rpl-30', 250, 5),\n",
    "                     # ('ubl-1', 400, 15),\n",
    "                     # ('Y73B3A.18', 100, 5),\n",
    "                     # From Q.M. Mitrovich & P. Anderson (Genes & Dev., 2000):\n",
    "                     # ID'ed 4 ribosomal proteins that showed change in \"expression\" upon NMD knockout:\n",
    "                     # ('rpl-7A', 100, 5),\n",
    "                     # ('rpl-1', 100, 5),  # aka rpl-10a!!\n",
    "                     # ('rpl-12', 100, 5),\n",
    "                     # ('rpl-3', 250, 5),\n",
    "                     # From V.S. Muir, A.P.Gash, & P. Anderson (G3, 2018):\n",
    "                     # Class I (showing up in all analyses)\n",
    "                     # ('B0495.8', 100, 5),\n",
    "                     # Some negative controls, random genes:\n",
    "                     # ('unc-54', 150, 5),\n",
    "                     # More genes I've found:\n",
    "                     # ('rpl-26', 150, 25),\n",
    "                     # ('rps-22', 150, 25),\n",
    "                     # Parissa's Stuff:\n",
    "                     # ('unc-54', 150, 25),\n",
    "                     # ('eef-2', 150, 25),\n",
    "                     # ('vig-1', 150, 25),\n",
    "                     # ('car-1', 150, 25),\n",
    "                     # Some more genes from comparing the fisher's p-value rankings:\n",
    "                     ('F19B2.5', 0, 0),\n",
    "                     ('Y73B3A.18', 0, 0),\n",
    "                     ('tos-1', 0, 0),\n",
    "                     ('C30E1.9', 0, 0),\n",
    "                     ('odc-1', 0, 0),\n",
    "                     ('rpl-5', 0, 0),\n",
    "                     ('smd-1', 0, 0),\n",
    "                     ('Y37E3.8', 0, 0),\n",
    "                     ('lep-5', 0, 0),\n",
    "                     ]\n",
    "divide_factor_bases = []  # Moved above # Larger number means less reads plotted! Generally 10-500 works well\n",
    "adapted_adjustment_factor_bases = []  # Moved above # Larger numbers here mean we'll get proportionally more adapted reads plotted, this is b/c they're so rare normally!\n",
    "\n",
    "for target_gene, divide_factor_base, adapted_adjustment_factor_base in target_genes_plus:\n",
    "    \n",
    "    target_gene_folder_path = f\"/home/marcus/Insync/marcus.viscardi@gmail.com/Google Drive/insync_folder/5TERA_ReadsAndTails_Plots/raw_figures_from_python/{npCommon.get_dt()}_plots_from_{target_gene}\"\n",
    "    \n",
    "    try:\n",
    "        os.mkdir(target_gene_folder_path)\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    \n",
    "    bar_plot_dicts = []\n",
    "    \n",
    "    for lib in [\n",
    "                'newN2',\n",
    "                'newS5',\n",
    "                'newS6',\n",
    "                # 'newS7',\n",
    "                'oldN2',\n",
    "                'oldS6',\n",
    "                # 'sPM57',\n",
    "                # 'sPM58',\n",
    "                ]:\n",
    "        # if lib.startswith('old'):\n",
    "        #     divide_factor = divide_factor_base * 1\n",
    "        #     adapted_adjustment_factor =  adapted_adjustment_factor_base / 2\n",
    "        # else:\n",
    "        #     divide_factor = divide_factor_base\n",
    "        #     adapted_adjustment_factor = adapted_adjustment_factor_base\n",
    "        # Here I use a round about way to pull the RPM information to normalize the number of reads plotted!\n",
    "        try:\n",
    "            adapted_rpm = compressed_df_genes.query(f\"lib == '{lib}'\").query(f\"gene_name == '{target_gene}'\").query(f\"t5 == '+'\")[\n",
    "                'gene_rpm'].values[0]\n",
    "        except IndexError:\n",
    "            # An IndexError would indicate that this place in the dataframe doesn't exist, because of the way these frames were made,\n",
    "            # if no reads match the parameters, the cell never gets made!! so this indicates there are 0 reads of the searched parameters!\n",
    "            adapted_rpm = 0\n",
    "        try:\n",
    "            unadapted_rpm = compressed_df_genes.query(f\"lib == '{lib}'\").query(f\"gene_name == '{target_gene}'\").query(f\"t5 == '-'\")[\n",
    "                'gene_rpm'].values[0]\n",
    "        except IndexError:\n",
    "            unadapted_rpm = 0\n",
    "        \n",
    "        sea.set_style(\"whitegrid\")\n",
    "        \n",
    "        total_reads_plotted = 8\n",
    "        \n",
    "        if lib[3] == 'S':\n",
    "            num_adapt_to_plot = 1\n",
    "        if lib[3] == 'N':\n",
    "            num_adapt_to_plot = 4\n",
    "        num_unadpt_to_plot = total_reads_plotted - num_adapt_to_plot\n",
    "        \n",
    "        # num_adapt_to_plot = int(adapted_rpm / (divide_factor / adapted_adjustment_factor))\n",
    "        # num_unadpt_to_plot = int(unadapted_rpm / divide_factor)\n",
    "        \n",
    "        # print(f\"For {lib} we are plotting {num_adapt_to_plot} adapted, and {num_unadpt_to_plot} unadapted reads.\\nThis is based on the RPM values of t5+:{adapted_rpm} & t5-:{unadapted_rpm}\")\n",
    "        \n",
    "        # sea.set_style(\"white\")\n",
    "        fig, axes = plt.subplots(nrows=3, ncols=1,\n",
    "                                 height_ratios=(1,2,4),\n",
    "                                 constrained_layout=True,\n",
    "                                 figsize=(10, 3),\n",
    "                                 )\n",
    "        plt.suptitle(f\"Coverage and Reads for {target_gene} from {lib}\")\n",
    "        \n",
    "        \n",
    "        coverage_plotting_5tera(reads_df_genes.query(f\"lib == '{lib}'\"),\n",
    "                                gene_name=target_gene,\n",
    "                                # save_dir=target_gene_folder_path, save_suffix=f\"_{lib}\",\n",
    "                                rpm_normalize=True,\n",
    "                                provide_axes=axes[:2],\n",
    "                                specify_gtf_path=parissas_parsed_gtf_path,\n",
    "                                )\n",
    "        \n",
    "        plot_reads(reads_df_genes.query(f\"lib == '{lib}'\"), gene_name_to_plot=target_gene,\n",
    "                   pad_x_axis_bounds_by=50,\n",
    "                   # subsample_fraction=0.125,\n",
    "                   t5_pos_count=num_adapt_to_plot, t5_neg_count=num_unadpt_to_plot,\n",
    "                   only_keep_reads_matched_to_gene=True, plot_width_and_height=(4, 3),\n",
    "                   # save_dir=target_gene_folder_path, save_suffix=f\"_{lib}_subSampledByRPM\",\n",
    "                   provided_axis=axes[-1],\n",
    "                   )\n",
    "        fig.supylabel('Reads Per Million')\n",
    "        plt.tight_layout()\n",
    "        read_plot_save_path = f\"{target_gene_folder_path}/readPlots_{target_gene}_{lib}\"\n",
    "        for file_type in ['.png', '.svg']:\n",
    "            plt.savefig(read_plot_save_path + file_type,\n",
    "                        dpi=300)\n",
    "        \n",
    "        # Save read information for bar plot!\n",
    "        bar_plot_dicts.append({\"gene_name\": target_gene,\n",
    "                               \"lib\": lib,\n",
    "                               \"unadapted_rpm\": unadapted_rpm,\n",
    "                               \"adapted_rpm\": adapted_rpm})\n",
    "    \n",
    "    sea.set_style(\"whitegrid\")\n",
    "    pd.DataFrame(bar_plot_dicts).set_index('lib').plot(kind='bar', stacked=True, color=['k', 'red'])\n",
    "    bar_plot_save_path = f\"{target_gene_folder_path}/barPlot_geneRPMS_{target_gene}\"\n",
    "    plt.title(f\"RPMs on {target_gene}\")\n",
    "    for file_type in ['.png', '.svg']:\n",
    "        plt.tight_layout()\n",
    "        plt.xticks(rotation=45)\n",
    "        plt.savefig(bar_plot_save_path + file_type,\n",
    "                    dpi=300)"
   ],
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
