{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# decappingQuantification2_metaPlots.ipynb\n",
    "## Marcus Viscardi,    July 11, 2024\n",
    "\n",
    "Check out `README.md` for a more overarching view of what is happening here.\n",
    "\n",
    "The `decappingQuantification.ipynb` script was getting REALLY bloated.\n",
    "This script trys to take the parts of that script that were related to producing meta plots and has them a little more streamlined!"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "72a8643cb3e55bad"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "from typing import Tuple\n",
    "\n",
    "import nanoporePipelineCommon as npCommon\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "from icecream import ic\n",
    "from datetime import datetime\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from scipy.stats import mannwhitneyu, ks_2samp\n",
    "\n",
    "import pickle as pkl\n",
    "\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "def __time_formatter__():\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    return f\"ic: {now} | > \"\n",
    "ic.configureOutput(prefix=__time_formatter__)\n",
    "\n",
    "_ = ic(\"Imports done!\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "gtf_parquet_path = Path(\"/data16/marcus/genomes/plus_cerENO2_elegansRelease100/230327_allChrs_plus-cerENO2.gtf.parquet\")\n",
    "\n",
    "gtf_df = pd.read_parquet(gtf_parquet_path)\n",
    "\n",
    "current_dir = Path.cwd()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "72801e8222ca3f5e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "libs_to_run = [\n",
    "    \"oldN2\",\n",
    "    # \"oldS6\",\n",
    "    \"newerN2\",\n",
    "    \"newerS6\",\n",
    "    \"newerS5\",\n",
    "    \"thirdN2\",\n",
    "    \"thirdS5\",\n",
    "    \"thirdS6\",\n",
    "    \"temp25cN2\",\n",
    "    \"temp25cS5\",\n",
    "    \"temp25cS6\",\n",
    "    \"temp25cS7\",\n",
    "]\n",
    "genes = [\n",
    "    'rps-22',\n",
    "    'ubl-1',\n",
    "    'rpl-30',\n",
    "    # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "    'rla-1',  # rev\n",
    "    'rps-25',  # rev\n",
    "    'Y37E3.8',  # rev\n",
    "    'rpl-10',\n",
    "    'rla-0',\n",
    "    'rpl-18',\n",
    "    'rps-3',\n",
    "    'rpl-38',\n",
    "    'rpl-24.1',\n",
    "    'hel-1',\n",
    "    'rpl-7A',\n",
    "    'ZK228.4',\n",
    "    'R06C1.4',\n",
    "    'rpl-12',\n",
    "    'pqn-70',\n",
    "    'rsp-5',\n",
    "    'rpl-1',\n",
    "    'C53H9.2',\n",
    "    'rpl-3',\n",
    "    'rsp-6',\n",
    "    'rpl-26',\n",
    "    'K08D12.3',\n",
    "    'aly-3',\n",
    "]\n",
    "genes = [ # Anything still commented below is because it is too long of a gene for use to consistently hit it's TSS\n",
    "    'rps-22',\n",
    "    'ubl-1',\n",
    "    'rpl-30',\n",
    "    # # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "    'rla-1',  # rev\n",
    "    'rps-25',  # rev\n",
    "    'Y37E3.8',  # rev\n",
    "    'rpl-10',\n",
    "    # 'rla-0',\n",
    "    'rpl-18',\n",
    "    'rps-3',\n",
    "    # 'rpl-38',\n",
    "    'rpl-24.1',\n",
    "    # 'hel-1',\n",
    "    # 'rpl-7A',\n",
    "    # 'ZK228.4',\n",
    "    'R06C1.4',\n",
    "    # 'rpl-12',\n",
    "    # 'pqn-70',\n",
    "    # 'rsp-5',\n",
    "    # 'rpl-1',\n",
    "    # 'C53H9.2',\n",
    "    # 'rpl-3',\n",
    "    # 'rsp-6',\n",
    "    'rpl-26',\n",
    "    'K08D12.3',\n",
    "    # 'aly-3',\n",
    "    'C45B2.8'  # 240815: New one!\n",
    "]\n",
    "plot_genes = genes[:]\n",
    "\n",
    "run_info = {\n",
    "    'libs': sorted(libs_to_run),\n",
    "    'genes': sorted(plot_genes),\n",
    "}\n",
    "run_info"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "b73600bfa4bcdd4e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "load_preprocessed_data = True\n",
    "preprocessed_data_path = current_dir / 'preprocessed_data' / 'decapping_data.parquet'\n",
    "preprocessed_info_path = current_dir / 'preprocessed_data' / 'decapping_data_info.pkl'\n",
    "if not preprocessed_data_path.exists() or not preprocessed_info_path.exists():\n",
    "    load_preprocessed_data = False\n",
    "    print(\"Preprocessed data not found, will reload data.\")\n",
    "if load_preprocessed_data:\n",
    "    print(\"Checking preprocessed data...\")\n",
    "    with open(preprocessed_info_path, 'rb') as f:\n",
    "        preprocessed_info = pkl.load(f)\n",
    "    if preprocessed_info != run_info:\n",
    "        print(\"Preprocessed data does not match current run info, will reload data.\")\n",
    "        load_preprocessed_data = False\n",
    "    else:\n",
    "        print(\"Preprocessed data matches current run info, will load data.\")\n",
    "if not load_preprocessed_data:\n",
    "    print(\"Will reload data and save preprocessed data.\")\n",
    "    with open(preprocessed_info_path, 'wb') as f:\n",
    "        pkl.dump(run_info, f)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "63a51ab06de5e1ce",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "if not load_preprocessed_data:\n",
    "    obj_dict = {}\n",
    "    for lib in libs_to_run:\n",
    "        print(f\"\\nLoading {lib}...\", end=\"\")\n",
    "        obj = npCommon.NanoporeRun(run_nickname=lib)\n",
    "        obj_dict[lib] = obj\n",
    "        # obj.load_mergedOnReads()\n",
    "        obj.load_nmd_targets(return_sambamobj=False)\n",
    "        print(\" Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "8b32a5eb01bc10ef",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "if not load_preprocessed_data:\n",
    "    gene_ids_to_names_dict = gtf_df.query('feature == \"gene\"').set_index('gene_id')['gene_name'].to_dict()\n",
    "    gene_names_to_ids_dict = gtf_df.query('feature == \"gene\"').set_index('gene_name')['gene_id'].to_dict()\n",
    "    for gene_name in plot_genes:\n",
    "        gene_id = gene_names_to_ids_dict[gene_name]\n",
    "        print(f\"{gene_name:>10} -> {gene_id}\")\n",
    "    gene_ids = [gene_names_to_ids_dict[gene_name] for gene_name in plot_genes]"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4dfc0ffc6bd47df6",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "def load_simple_nmd_targets(nmd_targets_path, target_gene_ids_to_names_dict) -> pd.DataFrame:\n",
    "    target_gene_names_to_ids_dict = {v: k for k, v in target_gene_ids_to_names_dict.items()}\n",
    "    reads_dict = {}\n",
    "    with pysam.AlignmentFile(nmd_targets_path, 'rb') as test_bam:\n",
    "        for read in test_bam.fetch():\n",
    "            read_assigned_gene_id = read.get_tag('gA')\n",
    "            # print(read.query_name, read_assigned_gene_id, target_gene_ids_to_names_dict[read_assigned_gene_id], read.get_tag('t5'), read.get_tag('nC'))\n",
    "            reads_dict[read.query_name] = {\n",
    "                'gene_id': read_assigned_gene_id,\n",
    "                'gene_name': target_gene_ids_to_names_dict[read_assigned_gene_id],\n",
    "                't5': read.get_tag('t5'),\n",
    "                'nmd_assignment': read.get_tag('nC'),\n",
    "                'cigar': read.cigarstring,\n",
    "                'chr_pos': read.reference_start,\n",
    "                'chr': read.reference_name,\n",
    "                'seq': read.query_sequence,\n",
    "            }\n",
    "    df = pd.DataFrame(reads_dict).T\n",
    "    return df\n",
    "\n",
    "if not load_preprocessed_data:\n",
    "    import pysam\n",
    "    \n",
    "    nmd_genes = []\n",
    "    \n",
    "    for lib, lib_obj in obj_dict.items():\n",
    "        print(f\"Loading {lib} NMD Targets from {lib_obj.nmd_targets_bam_path}...\", end=\"\")\n",
    "        lib_obj.load_nmd_targets(return_sambamobj=False)\n",
    "        print(\" Done!\")\n",
    "        lib_obj.nmd_targets_df = load_simple_nmd_targets(lib_obj.nmd_targets_bam_path, gene_ids_to_names_dict)\n",
    "        # print(lib_obj.nmd_targets_df.gene_name.unique())\n",
    "        nmd_genes.extend(lib_obj.nmd_targets_df.gene_name.unique())\n",
    "    nmd_genes = list(set(nmd_genes))\n",
    "    print(nmd_genes)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "91c7a552c1daa42a",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "UNANNOTATED_TSS_DICT = {  # these are manual annotations of transcription start sites that are not in the GTF file\n",
    "    'rpl-7A': [4_390_579, 4_390_464],\n",
    "}\n",
    "\n",
    "UNANNOTATED_STOPS_DICT = {  # these are manual annotations of stops that are not in the GTF file\n",
    "    'rps-22': [1_950_853],  # cassette exon contains PTC\n",
    "    'rpl-30': [10_436_332],  # extension of first exon contains PTC\n",
    "    'rpl-7A': [4_389_880, 4_389_745],  # unannotated extension of 3rd exon contains PTC (second loc due to potential FS)\n",
    "    'rsp-6': [7_790_576],  # longer unannotated form of cassette exon causes frame shift\n",
    "    'C53H9.2': [1_833_467],  # shorter first exon leaves frame shift\n",
    "    'pqn-70': [11_226_879],  # skipped exon leads to a frame change\n",
    "    'rpl-1': [2_875_908],  # new orf started in the 3' extension of the first exon, causing a frame change\n",
    "    'R06C1.4': [11_931_081],  # unannotated cassette exon contains a frame shift\n",
    "    'rpl-3': [3_868_335],  # incorrectly annotated first intron 3' truncation (2nd exon extended) contains stop\n",
    "    'hel-2': [8_327_970],  # later 5' SS adds to 3rd exon and contains PTC\n",
    "    'rpl-12': [13_240_078],  # later 5' SS of second intron adds extension to 2nd exon with a PTC\n",
    "    'aly-3': [12_123_859], # intron retention of the third intron has a PTC\n",
    "}\n",
    "\n",
    "BAD_TSS_ANNOTATIONS_DICT = {  # these are annotations that exist in the GTF that are incorrect (based on never being seen in FL nanopore libs)\n",
    "    'rpl-30': [10_436_466, 10_436_408],\n",
    "    'rpl-1': [2_876_039, 2_876_034],\n",
    "    'rpl-12': [13_240_204],\n",
    "    \n",
    "}\n",
    "\n",
    "def get_other_read_end(row):\n",
    "    cigar = row['cigar']\n",
    "    read_pos = row['chr_pos']\n",
    "    if cigar is None:\n",
    "        raise ValueError(\"Cigar is None!\")\n",
    "    if read_pos is None:\n",
    "        raise ValueError(\"read_pos is None!\")\n",
    "    pattern = r\"(\\d*)([MDISN])\"\n",
    "    matches = re.findall(pattern, cigar)\n",
    "    for match in matches:\n",
    "        count, key = match\n",
    "        if key in ['M', 'D', 'N']:\n",
    "            read_pos += int(count)\n",
    "    return read_pos\n",
    "\n",
    "def near_a_tss_or_stop(row, _strand, tss_positions, stop_positions, col_for_pos='chr_pos', full_window_size=50, window_bias_downstream=0.50, make_window_correction=True) -> Tuple[bool, bool]:\n",
    "    chr_pos = row[col_for_pos]\n",
    "    if _strand == '-' and make_window_correction:\n",
    "        window_bias_downstream = 1 - window_bias_downstream\n",
    "    elif _strand != '+' and make_window_correction:\n",
    "        raise ValueError(f\"Strand must be '+' or '-' not {_strand}\")\n",
    "    \n",
    "    near_tss = False\n",
    "    near_stop = False\n",
    "    chr_pos_upstream = chr_pos - full_window_size * (1 - window_bias_downstream)\n",
    "    chr_pos_downstream = chr_pos + full_window_size * window_bias_downstream\n",
    "    for tss in set(tss_positions):\n",
    "        if chr_pos_upstream <= tss <= chr_pos_downstream:\n",
    "            near_tss = True\n",
    "            break\n",
    "    for stop in set(stop_positions):\n",
    "        if chr_pos_upstream <= stop <= chr_pos_downstream:\n",
    "            near_stop = True\n",
    "            break\n",
    "    return near_tss, near_stop\n",
    "\n",
    "def convert_near_tss_and_stop_to_hue(row, tss_color='g', stop_color='r', neither_color='k', both_color='y'):\n",
    "    near_tss, near_stop = row['near_tss'], row['near_stop']\n",
    "    if near_tss and near_stop:\n",
    "        return both_color\n",
    "    elif near_tss:\n",
    "        return tss_color\n",
    "    elif near_stop:\n",
    "        return stop_color\n",
    "    else:\n",
    "        return neither_color\n",
    "\n",
    "def just_extract_gene_data(gtf_df, test_gene, plot_lib_df_dict, window_size=100,\n",
    "                           up_v_down_fraction=0.25, number_of_bins=40,\n",
    "                            flanking_gene_fraction=0.05, print_stats=False):    \n",
    "    extra_stops_dict = UNANNOTATED_STOPS_DICT\n",
    "    extra_tss_dict = UNANNOTATED_TSS_DICT\n",
    "    bad_tss_dict = BAD_TSS_ANNOTATIONS_DICT\n",
    "    \n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    gene_start, gene_end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    gene_length = gene_end - gene_start\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    if print_stats:\n",
    "        print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    if strand == '-':\n",
    "        if print_stats:\n",
    "            print(\"Because this gene is on the reverse strand, we will have to calculate out the other end of the reads with the CIGAR strings. This will take slightly longer.\")\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'end'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'end'].values\n",
    "        up_v_down_fraction = 1 - up_v_down_fraction\n",
    "        flipped = True\n",
    "    else:\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "        flipped = False\n",
    "        \n",
    "    stop_locations = list(stop_locations)\n",
    "    start_sites = list(start_sites)\n",
    "    \n",
    "    if test_gene in extra_stops_dict:\n",
    "        for stop_location in extra_stops_dict[test_gene]:\n",
    "            stop_locations.append(stop_location)\n",
    "    if test_gene in extra_tss_dict:\n",
    "        for tss_location in extra_tss_dict[test_gene]:\n",
    "            start_sites.append(tss_location)\n",
    "    if test_gene in bad_tss_dict:\n",
    "        for bad_tss_location in bad_tss_dict[test_gene]:\n",
    "            if bad_tss_location in start_sites:\n",
    "                start_sites.remove(bad_tss_location)\n",
    "            else:\n",
    "                raise ValueError(f\"Bad TSS location {bad_tss_location} not in start_sites for gene {test_gene}\\n\"\n",
    "                                 f\"Start Sites: {start_sites}\")\n",
    "    summary_dict_for_gene = {}\n",
    "    for i, (lib, lib_df) in enumerate(plot_lib_df_dict.items()):\n",
    "\n",
    "        if print_stats:\n",
    "            print(f\"\\t{lib}:\")\n",
    "        \n",
    "        gene_df = lib_df.query('gene_name == @test_gene').copy()\n",
    "        gene_df.replace({'t5': {0: '-', 1: '+', '+': '+', '-': '-'}}, inplace=True)\n",
    "        if flipped:\n",
    "            gene_df.loc[:, 'chr_pos'] = gene_df.apply(get_other_read_end, axis=1)\n",
    "        gene_df.loc[:, 'near_tss'], gene_df.loc[:, 'near_stop'] = zip(*gene_df.apply(near_a_tss_or_stop, args=(strand, start_sites, stop_locations),\n",
    "            full_window_size=window_size,window_bias_downstream=up_v_down_fraction, axis=1))\n",
    "        gene_df.loc[:, 'hue'] = gene_df.apply(convert_near_tss_and_stop_to_hue, axis=1)\n",
    "        \n",
    "        unadapted_df = gene_df.query('t5 == \"-\"').copy()\n",
    "        adapted_df = gene_df.query('t5 == \"+\"').copy()\n",
    "        \n",
    "        # Create annotations to put in ax5\n",
    "        ada_near_stop_count = adapted_df.query('near_stop').shape[0]\n",
    "        ada_near_tss_count = adapted_df.query('near_tss').shape[0]\n",
    "        unada_near_stop_count = unadapted_df.query('near_stop').shape[0]\n",
    "        unada_near_tss_count = unadapted_df.query('near_tss').shape[0]\n",
    "        ada_total_count = adapted_df.shape[0]\n",
    "        unada_total_count = unadapted_df.shape[0]\n",
    "        summary_dict_for_gene[lib] = {\n",
    "            'Adapted Near Stop': ada_near_stop_count,\n",
    "            'Adapted Near TSS': ada_near_tss_count,\n",
    "            'Total Adapted': ada_total_count,\n",
    "            'Unadapted Near Stop': unada_near_stop_count,\n",
    "            'Unadapted Near TSS': unada_near_tss_count,\n",
    "            'Total Unadapted': unada_total_count,\n",
    "        }\n",
    "    return summary_dict_for_gene\n",
    "\n",
    "if not load_preprocessed_data:\n",
    "    plot_libs = libs_to_run\n",
    "    \n",
    "    window_size_for_search = 100  # 100nt seems to make sense (+50 and -50 from the TSS or STOP)\n",
    "    \n",
    "    target_nmd_types = ('nmd_target', 'ambiguous', 'non_nmd_target')\n",
    "    subset_nmd_types_list = [target_nmd_types[0], target_nmd_types[1], target_nmd_types[2], target_nmd_types[:2], target_nmd_types[:]]\n",
    "    subset_nmd_types_names = ['NMD_Targets', 'Ambiguous_Reads', 'NMD_NonTargets', 'NMD_TargetsAndAmbiguous', 'All_Types']\n",
    "    \n",
    "    plot_lib_objs_dict = {lib: obj_dict[lib] for lib in plot_libs}\n",
    "    overall_summaries_nested_dict = {}\n",
    "    for gene in nmd_genes[::-1]:\n",
    "        # if gene != \"rpl-7A\":\n",
    "        #     continue\n",
    "        print(f\"{gene:#^50}\")\n",
    "        gene_summary_dicts = {}\n",
    "        \n",
    "        for subset_nmd_types, subset_nmd_types_name in zip(subset_nmd_types_list, subset_nmd_types_names):\n",
    "            try:\n",
    "                lib_df_dict_for_gene = {lib: lib_obj.nmd_targets_df.query('gene_name == @gene & nmd_assignment in @subset_nmd_types')\n",
    "                                        for lib, lib_obj in plot_lib_objs_dict.items()}\n",
    "                # summary_dict = plot_gene_data3(gtf_df, gene, lib_df_dict_for_gene, print_stats=False,\n",
    "                #                                title_override=f\"{gene}\\nNMD Target Types: {subset_nmd_types}\",\n",
    "                #                                save_name_override=f\"{gene}_{subset_nmd_types_name}\",\n",
    "                #                                current_dir=current_dir,)\n",
    "                summary_dict = just_extract_gene_data(gtf_df, gene, lib_df_dict_for_gene, print_stats=False, window_size=window_size_for_search)\n",
    "                gene_summary_dicts[subset_nmd_types_name] = summary_dict\n",
    "            except Exception as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                print(f\"Gene: {gene}\")\n",
    "                print(f\"NMD Types: {subset_nmd_types}\")\n",
    "                print(f\"Continuing...\")\n",
    "        overall_summaries_nested_dict[gene] = gene_summary_dicts\n",
    "    \n",
    "    plotting_list_for_df = []\n",
    "    \n",
    "    dict_index = 0\n",
    "    for gene, gene_dict in overall_summaries_nested_dict.items():\n",
    "        for nmd_target_type, target_type_dict in gene_dict.items():\n",
    "            print(f\"{gene} ({nmd_target_type}):\")\n",
    "            for lib, lib_dict in target_type_dict.items():\n",
    "                print(f\"\\t{lib}:\")\n",
    "                print(f\"\\t\\tAdapted Near Stop:   {lib_dict['Adapted Near Stop']}\")\n",
    "                print(f\"\\t\\tAdapted Near TSS:    {lib_dict['Adapted Near TSS']}\")\n",
    "                print(f\"\\t\\tTotal Adapted:       {lib_dict['Total Adapted']}\")\n",
    "                print(f\"\\t\\tUnadapted Near Stop: {lib_dict['Unadapted Near Stop']}\")\n",
    "                print(f\"\\t\\tUnadapted Near TSS:  {lib_dict['Unadapted Near TSS']}\")\n",
    "                print(f\"\\t\\tTotal Unadapted:     {lib_dict['Total Unadapted']}\")\n",
    "                try:\n",
    "                    print(f\"\\t\\tAda Stop/Unada FL: {lib_dict['Adapted Near Stop']/lib_dict['Unadapted Near TSS']:.2%}\")\n",
    "                except ZeroDivisionError:\n",
    "                    print(f\"\\t\\tAda Stop/Unada FL: -- NO UNADAPTED FL --\")\n",
    "                plotting_list_for_df.append(\n",
    "                    {\n",
    "                        'gene_name': gene,\n",
    "                        'nmd_target_type': nmd_target_type,\n",
    "                        'lib': lib,\n",
    "                        'adapted_near_stop': lib_dict['Adapted Near Stop'],\n",
    "                        'adapted_near_tss': lib_dict['Adapted Near TSS'],\n",
    "                        'total_adapted': lib_dict['Total Adapted'],\n",
    "                        'unadapted_near_stop': lib_dict['Unadapted Near Stop'],\n",
    "                        'unadapted_near_tss': lib_dict['Unadapted Near TSS'],\n",
    "                        'total_unadapted': lib_dict['Total Unadapted'],\n",
    "                    }\n",
    "                )\n",
    "    \n",
    "    plotting_summary_df = pd.DataFrame.from_records(plotting_list_for_df)\n",
    "    plotting_summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "aaec035386a3538e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "if not load_preprocessed_data:\n",
    "    # Before anything let's copy down code so we can regenerate the input df\n",
    "    summary_df = plotting_summary_df.copy()\n",
    "    \n",
    "    lib_strain_dict = {'N2': 'wildtype',\n",
    "                       'S5': 'smg-5',\n",
    "                       'S6': 'smg-6',\n",
    "                       'S7': 'smg-7'}\n",
    "    \n",
    "    summary_df['strain'] = summary_df.lib.str[-2:].map(lib_strain_dict)\n",
    "    \n",
    "    summary_df['total_near_tss'] = summary_df['adapted_near_tss'] + summary_df['unadapted_near_tss']\n",
    "    summary_df['total_near_stop'] = summary_df['adapted_near_stop'] + summary_df['unadapted_near_stop']\n",
    "    summary_df['total_reads'] = summary_df['total_adapted'] + summary_df['total_unadapted']\n",
    "    \n",
    "    # There are a few different ways to do this!\n",
    "    summary_df['frac_near_tss'] = summary_df['total_near_tss'] / summary_df['total_reads']\n",
    "    \n",
    "    summary_df['frac_near_tss'] = summary_df['adapted_near_tss'] / summary_df['total_reads']\n",
    "    \n",
    "    print(f\"Saving preprocessed data to {preprocessed_data_path}\")\n",
    "    summary_df.to_parquet(preprocessed_data_path)\n",
    "else:\n",
    "    print(\"Loading preprocessed data...\")\n",
    "    summary_df = pd.read_parquet(preprocessed_data_path)\n",
    "summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "13413112661b48ef",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import scipy.stats as stats\n",
    "from colorama import Fore, Style\n",
    "\n",
    "def run_stats_intralib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       strains=('wildtype', 'smg-5', 'smg-6'),\n",
    "                       output_file: Path = None):\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for strain in strains:\n",
    "        nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_Targets'\")\n",
    "        non_nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_NonTargets'\")\n",
    "        merge_df = nmd_df.merge(non_nmd_df,\n",
    "                                on=('gene_name', 'strain', 'rep', 'lib', 'name'),\n",
    "                                suffixes=('_nmd', '_non_nmd'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_nmd'], merge_df[f'{compare_col}_non_nmd'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[strain] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    output_string += f\"{'Strain':^8} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-' * 50 + '\\n'\n",
    "    for strain, test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{strain:>8} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string\n",
    "\n",
    "def run_stats_interlib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       comparisons=(('wildtype', 'smg-5'), ('wildtype', 'smg-6'), ('smg-5', 'smg-6')),\n",
    "                       nmd_type='NMD_NonTargets',\n",
    "                       output_file=None):\n",
    "    # Now let's do a similar comparison but between non-NMD-targets between libs\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for comp in comparisons:\n",
    "        nmd_df = df_in.query(\"strain == @comp[0] & nmd_target_type == @nmd_type\")\n",
    "        non_nmd_df = df_in.query(\"strain == @comp[1] & nmd_target_type == @nmd_type\")\n",
    "        drop_cols = ['nmd_target_type', 'strain', 'lib', 'name']\n",
    "        merge_df = nmd_df.drop(columns=drop_cols).merge(non_nmd_df.drop(columns=drop_cols), on=('gene_name', 'rep'), suffixes=('_lib1', '_lib2'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_lib1'], merge_df[f'{compare_col}_lib2'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[comp] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    \n",
    "    output_string += f\"{'Comp':^14} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-'*56 + '\\n'\n",
    "    \n",
    "    for (lib1, lib2), test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if lib1 == 'wildtype':\n",
    "            lib1 = 'wt'\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{lib1:>5} vs {lib2:<5} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5f03344e9f7fbe75",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Maybe the NMD targets have higher adapted_TSS fractions in the WT not because they have high numbers of adapted_TSS reads, but b/c they have lower numbers of the denominator (all reads)\n",
    "# To remedy this, let's try comparing the adapted_TSS to the unadapted_TSS\n",
    "\n",
    "# Additionally, it would be nice to do an adapted RPM comparison against Adapt TSS reads. But this needs more reads!!\n",
    "\n",
    "# I need to circle WAY back to retain both adapt and unadapted TSSes.. UGH LETS DO IT\n",
    "in_df = summary_df.copy()\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type']).sort_index()\n",
    "better_df = in_df.loc[:, :, :, :, ['NMD_Targets', 'NMD_NonTargets'], :].drop(columns=['frac_near_tss'])\n",
    "better_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4d684aee73a38627",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "min_total = 1\n",
    "min_adapted = 5\n",
    "\n",
    "def run_box_and_ttests(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       add_tss_ada_pseudo: float = -1.0,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> None:\n",
    "    input_df = input_df.copy()\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(# range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    output_file = output_dir / f'{npCommon.get_dt(for_file=True)}_boxPlotStats_{save_friendly_target_col}.txt'\n",
    "    output_file.touch(exist_ok=True)\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(f\"Stats for {target_col}\\n\")\n",
    "        f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    print(\"Targets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    print(\"NonTargets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    \n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nInter Library Stats:\\n\")\n",
    "    run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "def run_cdf_and_ttests(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.ecdf(box_df,\n",
    "                 facet_col='strain',\n",
    "                 x=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                 # height=500,\n",
    "                 # width=1000,\n",
    "                 )\n",
    "    fig.update_xaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    # fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    save_friendly_target_col += f\"_minTotal_{min_total_reads}_minAdapted_{min_adapted_reads}\"\n",
    "    fig.write_html(output_dir / f\"cdfPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"cdfPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    output_file = output_dir / f'cdfPlotStats_{save_friendly_target_col}.txt'\n",
    "    output_file.touch(exist_ok=True)\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(f\"Stats for {target_col}\\n\")\n",
    "        f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    print(\"Targets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    print(\"NonTargets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    \n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nInter Library Stats:\\n\")\n",
    "    run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "def run_xy_and_ttests(input_df: pd.DataFrame,\n",
    "                      target_num: str, target_denom: str,\n",
    "                      min_total_reads: int = 1,\n",
    "                      min_adapted_reads: int = 1,\n",
    "                      min_tss_ada_reads: int = 0,\n",
    "                      add_tss_ada_pseudo: int = -1,\n",
    "                      log_y: bool = True,\n",
    "                      extra_savename: str = '',\n",
    "                      output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                      template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.scatter(box_df,\n",
    "                 x=target_num,\n",
    "                 y=target_denom,\n",
    "                 facet_col='strain',\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 hover_name='name',\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 log_y=log_y,\n",
    "                 title=f\"{target_num} x {target_denom} by Strain<br>\"\n",
    "                       f\"Min Total: {min_total_reads}, Min Adapt: {min_adapted_reads}, Min TSS Adapt: {min_tss_ada_reads}\",\n",
    "                 opacity=0.7,\n",
    "                 )\n",
    "    # fig2 = px.density_contour(box_df,\n",
    "    #              x=target_num,\n",
    "    #              y=target_denom,\n",
    "    #              color='nmd_target_type',\n",
    "    #              color_discrete_map=colors_dict,\n",
    "    #              hover_name='name',\n",
    "    #              hover_data=[target_num, target_denom],\n",
    "    #              template=template,\n",
    "    #              log_y=log_y,\n",
    "    #              title=f\"{target_num} x {target_denom} by Strain<br>\"\n",
    "    #                    f\"Min Total: {min_total_reads}, Min Adapt: {min_adapted_reads}, Min TSS Adapt: {min_tss_ada_reads}\",\n",
    "    #              height=500,\n",
    "    #              width=700,\n",
    "    #              )\n",
    "    # fig = go.Figure(data=fig1.data + fig2.data)\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    \n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "    \n",
    "    save_friendly_target_col += extra_savename\n",
    "    \n",
    "    fig.write_html(output_dir / f\"xyPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"xyPlot_{save_friendly_target_col}.png\", scale=5,\n",
    "                 height=500,\n",
    "                 width=700,)\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    # output_file = output_dir / f'boxPlotStats_{save_friendly_target_col}.txt'\n",
    "    # output_file.touch(exist_ok=True)\n",
    "    # with open(output_file, 'w') as f:\n",
    "    #     f.write(f\"Stats for {target_col}\\n\")\n",
    "    #     f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    # print(\"Targets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    # print(\"NonTargets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    # \n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nInter Library Stats:\\n\")\n",
    "    # run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "min_adapteds = [\n",
    "    1,\n",
    "    # 2, 3, 4,\n",
    "    5,\n",
    "    # 10,\n",
    "]\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    'total_adapted',\n",
    "    # 'total_unadapted',\n",
    "]\n",
    "for min_adapted in min_adapteds:\n",
    "    for target_denominator in target_denominators:\n",
    "        print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "        # run_cdf_and_ttests(better_df, 'adapted_near_tss', target_denominator, log_y=True, min_adapted_reads=min_adapted,\n",
    "        #                    min_tss_ada_reads=1)\n",
    "        # run_xy_and_ttests(\n",
    "        #     better_df, # better_df.query(\"strain == 'wildtype'\"),\n",
    "        #     'adapted_near_tss', target_denominator,\n",
    "        #     extra_savename=f\"_wFacets\",\n",
    "        #     min_adapted_reads=min_adapted, min_tss_ada_reads=0, log_y=True)\n",
    "        run_box_and_ttests(\n",
    "            better_df, # better_df.query(\"strain == 'wildtype'\"),\n",
    "            'adapted_near_tss', target_denominator,\n",
    "            min_adapted_reads=min_adapted, min_tss_ada_reads=0, log_y=True,\n",
    "            add_tss_ada_pseudo=0.1,\n",
    "        )\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "48e22040fc9f4129",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Below was an individual gene plotter that doesn't really fit in this analysis now\n",
    "# ToDo: Move it to `decappingQuantification2_perGenePlots.ipynb` and update it to work in there\n",
    "\n",
    "# # Let's see if we can make quick bar plots for individual gene species\n",
    "# indv_df = better_df.copy()\n",
    "# \n",
    "# per_gene_plots_dir = current_dir / 'plots' / 'decappingPerGene'\n",
    "# per_gene_plots_dir.mkdir(exist_ok=True)\n",
    "# \n",
    "# target_num = 'adapted_near_tss'\n",
    "# target_denom = 'total_adapted'\n",
    "# \n",
    "# target_col = f\"{target_num} / {target_denom}\"\n",
    "# add_pseudocount = False\n",
    "# log_y = False\n",
    "# \n",
    "# # Let's try adding a pseudo-count to the adapted_near_tss and total_adapted columns\n",
    "# if add_pseudocount:\n",
    "#     indv_df['adapted_near_tss'] += 0.1\n",
    "#     indv_df['total_adapted'] += 0.1\n",
    "#     indv_df['adapted_near_tss / total_adapted'] = indv_df['adapted_near_tss'] / indv_df['total_adapted']\n",
    "# \n",
    "# # Drop replicate 4 (the temp 25C data)\n",
    "# indv_df = indv_df.query(\"rep != 4\")\n",
    "# \n",
    "# \n",
    "# genes_to_ind_plot = [\n",
    "#     'rpl-30',\n",
    "#     'rps-22',\n",
    "#     'rpl-7A',\n",
    "#     'ubl-1',\n",
    "# ]\n",
    "# \n",
    "# genes_to_ind_plot = indv_df.index.get_level_values('gene_name').unique()\n",
    "# \n",
    "# # These colors are updated from the final plots!\n",
    "# plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "#                ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "#                ('#676767', '#414141'),  # ambiguous greys\n",
    "#                )\n",
    "# colors_dict = {'NMD_Targets': plot_colors[0][1],\n",
    "#                'NMD_NonTargets': plot_colors[1][1],}\n",
    "# \n",
    "# for gene_name in genes_to_ind_plot:\n",
    "#     print(f\"Working on {gene_name}...\")\n",
    "#     gene_df = indv_df.query(\"gene_name == @gene_name\")\n",
    "#     # Drop things with NA values:\n",
    "#     gene_df = gene_df.dropna(axis=0)\n",
    "#     # Drop things without any adapted reads near the TSS:\n",
    "#     # gene_df = gene_df.query(\"adapted_near_tss > 0\")\n",
    "#     # Simplify the indexes:\n",
    "#     gene_df = gene_df.reset_index()\n",
    "#     gene_df['strain'] = pd.Categorical(gene_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "#     gene_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "#     gene_df['name'] = gene_df['gene_name'] + ' ' + gene_df['strain'].astype(str) + ' rep' + gene_df['rep'].astype(str)\n",
    "#     fig = px.box(gene_df,\n",
    "#                  x='strain',\n",
    "#                  y=target_col,\n",
    "#                  color='nmd_target_type',\n",
    "#                  color_discrete_map={'NMD_Targets': plot_colors[0][0],\n",
    "#                                      'NMD_NonTargets': plot_colors[1][0],},\n",
    "#                  points='all',\n",
    "#                  hover_name='name',\n",
    "#                  log_y=log_y,\n",
    "#                  title=f\"{gene_name} Adapted Near TSS / Total Adapted\",\n",
    "#                  hover_data=['adapted_near_tss', 'total_adapted', 'total_reads'],\n",
    "#                  template='plotly_white',\n",
    "#                  height=500,\n",
    "#                  width=1000,\n",
    "#                  )\n",
    "#     fig.update_yaxes(range=[-0.1, 1.1], title_text=f'{target_num} / <br>{target_denom}')\n",
    "#     fig.update_layout(\n",
    "#         font=dict(\n",
    "#             # family=\"Courier New, monospace\",\n",
    "#             size=20,\n",
    "#             # color=\"RebeccaPurple\",\n",
    "#         ),\n",
    "#     )\n",
    "#     # fig = px.bar(gene_df,\n",
    "#     #          x='strain',\n",
    "#     #          y=target_col,\n",
    "#     #          color='nmd_target_type',\n",
    "#     #          color_discrete_map={'NMD_Targets': plot_colors[0][1],\n",
    "#     #                              'NMD_NonTargets': plot_colors[1][1],},\n",
    "#     #          hover_name='name',\n",
    "#     #          barmode='group',\n",
    "#     #          facet_col='rep',\n",
    "#     #          log_y=log_y,\n",
    "#     #          title=f\"{gene_name} Adapted Near TSS / Total Adapted\",\n",
    "#     #          hover_data=['adapted_near_tss', 'total_adapted', 'total_reads'],\n",
    "#     #          )\n",
    "#     fig.write_image(per_gene_plots_dir / f\"{gene_name}_adaptedNearTSS_vs_totalAdapted.png\")\n",
    "#     fig.write_image(per_gene_plots_dir / f\"{gene_name}_adaptedNearTSS_vs_totalAdapted.svg\")\n",
    "#     fig.write_html(per_gene_plots_dir / f\"{gene_name}_adaptedNearTSS_vs_totalAdapted.html\")\n",
    "#     fig.show()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2764b731f650454e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# 8/16/24 After meeting with Josh\n",
    "One of the major factors that confuses the box plots above is the massive difference in depth between the NMD-targets and the Non-targets in the WT libraries\n",
    "\n",
    "Another question is about if each gene has similar effects across the different replicates.\n",
    "\n",
    "The solution for the first issue is to bootstrap the non-NMD targets to the same depth as the NMD targets and then compare the two.\n",
    "\n",
    "The solution for the second is to try the means of the replicates AND look at genes on their own."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f61b9e455e093ffc"
  },
  {
   "cell_type": "markdown",
   "source": [
    "# First let's work on bootstrapping the libraries"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "95e7509a7fe634a8"
  },
  {
   "cell_type": "code",
   "source": [
    "better_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "eee61d459a6737ff",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "def bootstrap_row_optimized(index: tuple, row: pd.Series, num_samples: int, sample_size: int) -> pd.Series:\n",
    "\n",
    "    num_ada_tss = row['adapted_near_tss']\n",
    "    num_unada_tss = row['unadapted_near_tss']\n",
    "    num_ada_not_tss = row['total_adapted'] - row['adapted_near_tss']\n",
    "    num_unada_not_tss = row['total_unadapted'] - row['unadapted_near_tss']\n",
    "\n",
    "    set_up_array = np.array([\n",
    "        [1, 1],  # adapted, near TSS\n",
    "        [1, 0],  # adapted, not TSS\n",
    "        [0, 1],  # unadapted, near TSS\n",
    "        [0, 0],  # unadapted, not TSS\n",
    "    ])\n",
    "    bootstrap_array = np.repeat(set_up_array, [num_ada_tss, num_ada_not_tss, num_unada_tss, num_unada_not_tss], axis=0)\n",
    "\n",
    "    bootstrap_results = []\n",
    "    for cycle in range(num_samples):\n",
    "        sample_indices = np.random.choice(bootstrap_array.shape[0], size=sample_size, replace=True)\n",
    "        sample = bootstrap_array[sample_indices]\n",
    "        unique, counts = np.unique(sample, axis=0, return_counts=True)\n",
    "        bootstrap_results.append(dict(zip(map(tuple, unique), counts)))\n",
    "\n",
    "    col_conversion_dict = {\n",
    "        (1, 1): 'ada_near_tss',\n",
    "        (1, 0): 'ada_not_tss',\n",
    "        (0, 1): 'unada_near_tss',\n",
    "        (0, 0): 'unada_not_tss',\n",
    "    }\n",
    "\n",
    "    # Convert list of dicts to DataFrame\n",
    "    bootstrap_df = pd.DataFrame(bootstrap_results).fillna(0).astype(int)\n",
    "    bootstrap_df.columns = [col_conversion_dict[col] for col in bootstrap_df.columns]\n",
    "    \n",
    "    means = bootstrap_df.mean().rename(index)\n",
    "    return means\n",
    "\n",
    "def bootstrap_row_for_apply(row, bootstrap_cycles: int = 1000, sample_size: int = 10):\n",
    "    index = row.name\n",
    "    return bootstrap_row_optimized(index, row, num_samples=bootstrap_cycles, sample_size=sample_size)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5199c4b2466d0e25",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "tqdm.pandas(desc=\"Bootstrapping\")\n",
    "bootstrap_df = better_df.query(\"total_adapted >= 5\").progress_apply(bootstrap_row_for_apply, axis=1).fillna(0).astype(float)\n",
    "bootstrap_df['total_reads'] = bootstrap_df['ada_near_tss'] + bootstrap_df['ada_not_tss'] + bootstrap_df['unada_near_tss'] + bootstrap_df['unada_not_tss']\n",
    "bootstrap_df['total_adapted'] = bootstrap_df['ada_near_tss'] + bootstrap_df['ada_not_tss']"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "990eac7c9de66f10",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "\n",
    "bootstrap_df['adapted_near_tss / total_reads'] = bootstrap_df['ada_near_tss'] / bootstrap_df['total_reads']\n",
    "bootstrap_df['adapted_near_tss / total_adapted'] = bootstrap_df['ada_not_tss'] / bootstrap_df['total_adapted']"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f15be65e36a62e9d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "bootstrap_df.head(10)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "7937d29f9ea86d3d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "better_df.head(10)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "75f71e82532cca7b",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "def run_box_bootstrapped(input_df: pd.DataFrame,\n",
    "                         target_num: str, target_denom: str,\n",
    "                         min_total_reads: int = 1,\n",
    "                         min_adapted_reads: int = 1,\n",
    "                         min_tss_ada_reads: int = 0,\n",
    "                         add_tss_ada_pseudo: float = -1.0,\n",
    "                         output_dir: Path = current_dir / 'plots' / 'decapping_bootstrapped',\n",
    "                         bootstrap_cycles: int = 1000,\n",
    "                         bootstrap_sample_size: int = 10,\n",
    "                         log_y: bool = True,\n",
    "                         log2_y: bool = False,\n",
    "                         template: str = 'none') -> None:\n",
    "    input_df = input_df.copy()\n",
    "    input_df = input_df.query(\"adapted_near_tss >= @min_tss_ada_reads\")\n",
    "    input_df = input_df.query(\"total_reads >= @min_total_reads\")\n",
    "    input_df = input_df.query(\"total_adapted >= @min_adapted_reads\")\n",
    "    \n",
    "    \n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    tqdm.pandas(desc=\"Bootstrapping\")\n",
    "    boot_df = input_df.query(\"total_adapted >= @min_adapted_reads\").progress_apply(bootstrap_row_for_apply, bootstrap_cycles=bootstrap_cycles, sample_size=bootstrap_sample_size, axis=1).fillna(0).astype(float)\n",
    "    boot_df.rename(columns={'ada_near_tss': 'adapted_near_tss',\n",
    "                            'ada_not_tss': 'ada_not_tss',\n",
    "                            'unada_near_tss': 'unadapted_near_tss',\n",
    "                            'unada_not_tss': 'unada_not_tss'},\n",
    "                        inplace=True)\n",
    "    print(boot_df.columns)\n",
    "    boot_df['total_reads'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss'] + boot_df['unadapted_near_tss'] + boot_df['unada_not_tss']\n",
    "    boot_df['total_adapted'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss']\n",
    "    boot_df['total_near_tss'] = boot_df['adapted_near_tss'] + boot_df['unadapted_near_tss']\n",
    "    input_df = boot_df\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>\"\n",
    "                       f\"Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}<br>\"\n",
    "                       f\"Bootstrapped {bootstrap_cycles} cycles of {bootstrap_sample_size} samples\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    \n",
    "    # # Can we add a line for the means?\n",
    "    # means = box_df.groupby(['strain', 'nmd_target_type']).mean().reset_index()\n",
    "    # fig.add_trace(go.Scatter(x=means['strain'], y=means[target_col], mode='lines+markers', marker=dict(size=10, color='black'), name='Mean'))\n",
    "    \n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(# range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=16,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    save_friendly_target_col += f\"_bootstrapped{bootstrap_cycles}cycleX{bootstrap_sample_size}samples\"\n",
    "    print(f\"Saving file to {output_dir / f'{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png/html'}\")\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "\n",
    "\n",
    "min_adapteds = [\n",
    "    1,\n",
    "    # 2,\n",
    "    3,\n",
    "    # 4,\n",
    "    5,\n",
    "    # 10,\n",
    "]\n",
    "\n",
    "target_numerators = [\n",
    "    'adapted_near_tss',\n",
    "    'total_near_tss',\n",
    "]\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    'total_adapted',\n",
    "    # 'total_unadapted',\n",
    "]\n",
    "for min_adapted in min_adapteds:\n",
    "    for target_numerator in target_numerators:\n",
    "        for target_denominator in target_denominators:\n",
    "            print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "            run_box_bootstrapped(\n",
    "                better_df,\n",
    "                target_numerator,\n",
    "                target_denominator,\n",
    "                min_adapted_reads=min_adapted,\n",
    "                min_tss_ada_reads=0,\n",
    "                bootstrap_cycles=100,\n",
    "                bootstrap_sample_size=10,\n",
    "                log_y=True,\n",
    "                add_tss_ada_pseudo=0.001,\n",
    "            )\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "31e4c89aa833afd9",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "def run_box_with_means(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       add_tss_ada_pseudo: float = -1.0,\n",
    "                       min_number_reps_observed: int = -1,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> pd.DataFrame:\n",
    "    input_df = input_df.copy()\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    groupby_obj = box_df.reset_index().set_index(['strain', 'gene_name', 'nmd_target_type']).groupby(['strain', 'gene_name', 'nmd_target_type'])\n",
    "    keep_cols = [target_col, target_num, target_denom]\n",
    "    groupby_df = groupby_obj[keep_cols].mean()\n",
    "    std_cols = [f'{col}_std' for col in keep_cols]\n",
    "    groupby_df[std_cols] = groupby_obj[keep_cols].std()\n",
    "    groupby_df['successful_reps'] = groupby_obj['rep'].apply(lambda x: ','.join(map(str, x)))\n",
    "    groupby_df['num_successful_reps'] = groupby_obj['rep'].count()\n",
    "    groupby_df.reset_index(inplace=True)\n",
    "    if min_number_reps_observed > 0:\n",
    "        groupby_df = groupby_df.query(\"num_successful_reps >= @min_number_reps_observed\")\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.box(groupby_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='gene_name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom, 'successful_reps', 'num_successful_reps'] + std_cols,\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>\"\n",
    "                       f\"Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}<br>\"\n",
    "                       f\"Means of Replicates\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(# range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=16,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    save_friendly_target_col += f\"_meanOnGene\"\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    return box_df\n",
    "\n",
    "def run_box_with_means2(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       add_tss_ada_pseudo: float = -1.0,\n",
    "                       min_number_reps_observed: int = -1,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> pd.DataFrame:\n",
    "    input_df = input_df.copy()\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    groupby_obj = box_df.reset_index().set_index(['strain', 'gene_name', 'nmd_target_type']).groupby(['strain', 'gene_name', 'nmd_target_type'])\n",
    "    keep_cols = [target_col, target_num, target_denom]\n",
    "    groupby_df = groupby_obj[keep_cols].mean()\n",
    "    std_cols = [f'{col}_std' for col in keep_cols]\n",
    "    groupby_df[std_cols] = groupby_obj[keep_cols].std()\n",
    "    groupby_df['successful_reps'] = groupby_obj['rep'].apply(lambda x: ','.join(map(str, x)))\n",
    "    groupby_df['num_successful_reps'] = groupby_obj['rep'].count()\n",
    "    groupby_df.reset_index(inplace=True)\n",
    "    if min_number_reps_observed > 0:\n",
    "        groupby_df = groupby_df.query(\"num_successful_reps >= @min_number_reps_observed\")\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.box(groupby_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 # points='all',\n",
    "                 hover_name='gene_name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom, 'successful_reps', 'num_successful_reps'] + std_cols,\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>\"\n",
    "                       f\"Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}<br>\"\n",
    "                       f\"Means of Replicates\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    \n",
    "    # Add strip plot\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=box_df['strain'],\n",
    "        y=box_df[target_col],\n",
    "        mode='markers',\n",
    "        marker=dict(color=box_df['nmd_target_type'].map(colors_dict)),\n",
    "        text=box_df['gene_name'],\n",
    "        name='Strip Plot',\n",
    "    ))\n",
    "    \n",
    "    # Add lines between points of the same gene\n",
    "    for gene in box_df['gene_name'].unique():\n",
    "        gene_df = box_df[box_df['gene_name'] == gene]\n",
    "        if len(gene_df) > 1:\n",
    "            fig.add_trace(go.Scatter(\n",
    "                x=gene_df['strain'],\n",
    "                y=gene_df[target_col],\n",
    "                mode='lines',\n",
    "                line=dict(color='gray', dash='dash'),\n",
    "                showlegend=False\n",
    "            ))\n",
    "    \n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(# range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=16,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    save_friendly_target_col += f\"_meanOnGene\"\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    return box_df\n",
    "\n",
    "min_adapteds = [\n",
    "    # 0,\n",
    "    1,\n",
    "    # 2,\n",
    "    3,\n",
    "    # 4,\n",
    "    # 5,\n",
    "    # 10,\n",
    "]\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    'total_adapted',\n",
    "    # 'total_unadapted',\n",
    "]\n",
    "for min_adapted in min_adapteds:\n",
    "    for target_denominator in target_denominators:\n",
    "        print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "        test_df = run_box_with_means2(\n",
    "            better_df,\n",
    "            'adapted_near_tss',\n",
    "            target_denominator,\n",
    "            min_adapted_reads=min_adapted,\n",
    "            min_tss_ada_reads=-1,\n",
    "            log_y=True,\n",
    "            add_tss_ada_pseudo=0.001,\n",
    "            min_number_reps_observed=-1,\n",
    "        )\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "aeba614daf50890b",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "test_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4daea33c36db5690",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "122fd7e532c535bc",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "d4352d62f3f2afdc",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
