{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# decappingQuantification3_allGenes.ipynb\n",
    "## Marcus Viscardi,    July 18, 2024\n",
    "\n",
    "Check out `README.md` for a more overarching view of what is happening here.\n",
    "\n",
    "This script is taking items from both `decappingQuantification.ipynb` and `decappingQuantification2_metaPlots.ipynb` and trying to combine them into a single script that can be run on all genes. This should give us an opportunity to see the general trends in decapping efficiency across all genes and see if there are any interesting patterns that emerge.\n",
    "\n",
    "The initial pass will just be with wildtype libraries. I can add NMD-mutants afterwards if it seems like it would be useful."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f40b24bf90506e36"
  },
  {
   "cell_type": "code",
   "source": [
    "from typing import Tuple\n",
    "\n",
    "import nanoporePipelineCommon as npCommon\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "from icecream import ic\n",
    "from datetime import datetime\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from scipy.stats import mannwhitneyu, ks_2samp\n",
    "\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "def __time_formatter__():\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    return f\"ic: {now} | > \"\n",
    "ic.configureOutput(prefix=__time_formatter__)\n",
    "\n",
    "_ = ic(\"Imports done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "1e4a2d9aa8ecba7d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "gtf_parquet_path = Path(\"/data16/marcus/genomes/plus_cerENO2_elegansRelease100/230327_allChrs_plus-cerENO2.gtf.parquet\")\n",
    "\n",
    "gtf_df = pd.read_parquet(gtf_parquet_path)\n",
    "\n",
    "current_dir = Path.cwd()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "obj_dict = {}\n",
    "pretty_name_map = {\n",
    "    ### The classics:\n",
    "    # \"oldN2\": \"Wildtype (rep1)\",\n",
    "    # \"oldS6\": \"<i>smg-6</i> (rep1)\",\n",
    "    ### The terrible second replicates:\n",
    "    # \"newN2\": \"Wildtype (bad rep)\",\n",
    "    # \"newS5\": \"<i>smg-5</i> (bad rep)\",\n",
    "    # \"newS6\": \"<i>smg-6</i> (bad rep)\",\n",
    "    ### The \"better\" second replicates:\n",
    "    \"newerN2\": \"Wildtype (rep2)\",\n",
    "    \"newerS6\": \"<i>smg-6</i> (rep2)\",\n",
    "    \"newerS5\": \"<i>smg-5</i> (rep2)\",\n",
    "    ### The triplicates!\n",
    "    \"thirdN2\": \"Wildtype (rep3)\",\n",
    "    \"thirdS5\": \"<i>smg-5</i> (rep3)\",\n",
    "    \"thirdS6\": \"<i>smg-6</i> (rep3)\",\n",
    "    ### At 25C for smg-7 (and fourth replicates in a way):\n",
    "    # \"temp25cN2\": \"Wildtype (25째C)\",\n",
    "    # \"temp25cS5\": \"<i>smg-5</i> (25째C)\",\n",
    "    # \"temp25cS6\": \"<i>smg-6</i> (25째C)\",\n",
    "    # \"temp25cS7\": \"<i>smg-7</i> (25째C)\",\n",
    "}\n",
    "libs_to_run = list(pretty_name_map.keys())\n",
    "for lib in libs_to_run:\n",
    "    print(f\"\\nLoading {lib}...\", end=\"\")\n",
    "    obj = npCommon.NanoporeRun(run_nickname=lib)\n",
    "    obj_dict[lib] = obj\n",
    "    obj.load_mergedOnReads()\n",
    "    # obj.load_nmd_targets(return_sambamobj=False)\n",
    "    print(\" Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3048be639258a2e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# genes = [\n",
    "#     'rps-22',\n",
    "#     'ubl-1',\n",
    "#     'rpl-30',\n",
    "#     # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "#     'rla-1',  # rev\n",
    "#     'rps-25',  # rev\n",
    "#     'Y37E3.8',  # rev\n",
    "#     'rpl-10',\n",
    "#     'rla-0',\n",
    "#     'rpl-18',\n",
    "#     'rps-3',\n",
    "#     'rpl-38',\n",
    "#     'rpl-24.1',\n",
    "#     'hel-1',\n",
    "#     'rpl-7A',\n",
    "#     'ZK228.4',\n",
    "#     'R06C1.4',\n",
    "#     'rpl-12',\n",
    "#     'pqn-70',\n",
    "#     'rsp-5',\n",
    "#     'rpl-1',\n",
    "#     'C53H9.2',\n",
    "#     'rpl-3',\n",
    "#     'rsp-6',\n",
    "#     'rpl-26',\n",
    "#     'K08D12.3',\n",
    "#     'aly-3',\n",
    "# ]\n",
    "# plot_genes = genes[:]\n",
    "plot_genes = gtf_df.query('feature == \"gene\"').gene_name.unique()\n",
    "gene_ids_to_names_dict = gtf_df.query('feature == \"gene\"').set_index('gene_id')['gene_name'].to_dict()\n",
    "gene_names_to_ids_dict = gtf_df.query('feature == \"gene\"').set_index('gene_name')['gene_id'].to_dict()\n",
    "for gene_name in plot_genes:\n",
    "    gene_id = gene_names_to_ids_dict[gene_name]\n",
    "    print(f\"{gene_name:>10} -> {gene_id}\")\n",
    "gene_ids = [gene_names_to_ids_dict[gene_name] for gene_name in plot_genes]"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "9ef4e10ad7e1adb2",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import pysam"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "38f290d6398dae90",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "UNANNOTATED_TSS_DICT = {  # these are manual annotations of transcription start sites that are not in the GTF file\n",
    "    'rpl-7A': [4_390_579, 4_390_464],\n",
    "}\n",
    "\n",
    "UNANNOTATED_STOPS_DICT = {  # these are manual annotations of stops that are not in the GTF file\n",
    "    'rps-22': [1_950_853],  # cassette exon contains PTC\n",
    "    'rpl-30': [10_436_332],  # extension of first exon contains PTC\n",
    "    'rpl-7A': [4_389_880, 4_389_745],  # unannotated extension of 3rd exon contains PTC (second loc due to potential FS)\n",
    "    'rsp-6': [7_790_576],  # longer unannotated form of cassette exon causes frame shift\n",
    "    'C53H9.2': [1_833_467],  # shorter first exon leaves frame shift\n",
    "    'pqn-70': [11_226_879],  # skipped exon leads to a frame change\n",
    "    'rpl-1': [2_875_908],  # new orf started in the 3' extension of the first exon, causing a frame change\n",
    "    'R06C1.4': [11_931_081],  # unannotated cassette exon contains a frame shift\n",
    "    'rpl-3': [3_868_335],  # incorrectly annotated first intron 3' truncation (2nd exon extended) contains stop\n",
    "    'hel-2': [8_327_970],  # later 5' SS adds to 3rd exon and contains PTC\n",
    "    'rpl-12': [13_240_078],  # later 5' SS of second intron adds extension to 2nd exon with a PTC\n",
    "    'aly-3': [12_123_859], # huge extension of second exon contains as PTC\n",
    "}\n",
    "\n",
    "def get_other_read_end(row):\n",
    "    cigar = row['cigar']\n",
    "    read_pos = row['chr_pos']\n",
    "    if cigar is None:\n",
    "        raise ValueError(\"Cigar is None!\")\n",
    "    if read_pos is None:\n",
    "        raise ValueError(\"read_pos is None!\")\n",
    "    pattern = r\"(\\d*)([MDISN])\"\n",
    "    matches = re.findall(pattern, cigar)\n",
    "    for match in matches:\n",
    "        count, key = match\n",
    "        if key in ['M', 'D', 'N']:\n",
    "            read_pos += int(count)\n",
    "    return read_pos\n",
    "\n",
    "def near_a_tss(row, _strand, tss_positions, col_for_pos='chr_pos', full_window_size=50, window_bias_downstream=0.50, make_window_correction=True) -> Tuple[bool, bool]:\n",
    "    chr_pos = row[col_for_pos]\n",
    "    if _strand == '-' and make_window_correction:\n",
    "        window_bias_downstream = 1 - window_bias_downstream\n",
    "    elif _strand != '+' and make_window_correction:\n",
    "        raise ValueError(f\"Strand must be '+' or '-' not {_strand}\")\n",
    "    \n",
    "    near_tss = False\n",
    "    chr_pos_upstream = chr_pos - full_window_size * (1 - window_bias_downstream)\n",
    "    chr_pos_downstream = chr_pos + full_window_size * window_bias_downstream\n",
    "    for tss in set(tss_positions):\n",
    "        if chr_pos_upstream <= tss <= chr_pos_downstream:\n",
    "            near_tss = True\n",
    "            break\n",
    "    return near_tss\n",
    "\n",
    "# TODO: Rework the following to be more efficient!\n",
    "def just_extract_gene_data(gtf_df, test_gene, plot_lib_df_dict, window_size=100,\n",
    "                           up_v_down_fraction=0.25, print_stats=False):    \n",
    "    # extra_stops_dict = UNANNOTATED_STOPS_DICT\n",
    "    extra_tss_dict = UNANNOTATED_TSS_DICT\n",
    "    \n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    gene_start, gene_end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    gene_length = gene_end - gene_start\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    if print_stats:\n",
    "        print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    if strand == '-':\n",
    "        if print_stats:\n",
    "            print(\"Because this gene is on the reverse strand, we will have to calculate out the other end of the reads with the CIGAR strings. This will take slightly longer.\")\n",
    "        # stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'end'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'end'].values\n",
    "        up_v_down_fraction = 1 - up_v_down_fraction\n",
    "        flipped = True\n",
    "    else:\n",
    "        # stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "        flipped = False\n",
    "        \n",
    "    # stop_locations = list(stop_locations)\n",
    "    start_sites = list(start_sites)\n",
    "    \n",
    "    # if test_gene in extra_stops_dict:\n",
    "    #     for stop_location in extra_stops_dict[test_gene]:\n",
    "    #         stop_locations.append(stop_location)\n",
    "    if test_gene in extra_tss_dict:\n",
    "        for tss_location in extra_tss_dict[test_gene]:\n",
    "            start_sites.append(tss_location)\n",
    "    summary_dict_for_gene = {}\n",
    "    for i, (lib, lib_df) in enumerate(plot_lib_df_dict.items()):\n",
    "\n",
    "        if print_stats:\n",
    "            print(f\"\\t{lib}:\")\n",
    "        \n",
    "        gene_df = lib_df.query('gene_name == @test_gene').copy()\n",
    "        gene_df.replace({'t5': {0: '-', 1: '+', '+': '+', '-': '-'}}, inplace=True)\n",
    "        if flipped:\n",
    "            gene_df.loc[:, 'chr_pos'] = gene_df.apply(get_other_read_end, axis=1)\n",
    "        gene_df.loc[:, 'near_tss'] = gene_df.apply(near_a_tss, args=(strand, start_sites),\n",
    "            full_window_size=window_size,window_bias_downstream=up_v_down_fraction, axis=1)\n",
    "        \n",
    "        unadapted_df = gene_df.query('t5 == \"-\"').copy()\n",
    "        adapted_df = gene_df.query('t5 == \"+\"').copy()\n",
    "        \n",
    "        ada_near_tss_count = adapted_df.query('near_tss').shape[0]\n",
    "        unada_near_tss_count = unadapted_df.query('near_tss').shape[0]\n",
    "        ada_total_count = adapted_df.shape[0]\n",
    "        unada_total_count = unadapted_df.shape[0]\n",
    "        summary_dict_for_gene[lib] = {\n",
    "            # 'Adapted Near Stop': ada_near_stop_count,\n",
    "            'Adapted Near TSS': ada_near_tss_count,\n",
    "            'Total Adapted': ada_total_count,\n",
    "            # 'Unadapted Near Stop': unada_near_stop_count,\n",
    "            'Unadapted Near TSS': unada_near_tss_count,\n",
    "            'Total Unadapted': unada_total_count,\n",
    "            'Gene Length': gene_length,\n",
    "        }\n",
    "    return summary_dict_for_gene\n",
    "\n",
    "\n",
    "plot_libs = [\n",
    "    ### The classics:\n",
    "    #\"oldN2\",\n",
    "    # \"oldS6\",\n",
    "    ### The terrible second replicates:\n",
    "    # \"newN2\",\n",
    "    # \"newS5\",\n",
    "    # \"newS6\",\n",
    "    ### The \"better\" second replicates:\n",
    "    \"newerN2\",\n",
    "    \"newerS6\",\n",
    "    \"newerS5\",\n",
    "    ### The triplicates!\n",
    "    \"thirdN2\",\n",
    "    \"thirdS5\",\n",
    "    \"thirdS6\",\n",
    "    ### At 25C for smg-7 (and fourth replicates in a way):\n",
    "    # \"temp25cN2\",\n",
    "    # \"temp25cS5\",\n",
    "    # \"temp25cS6\",\n",
    "    # \"temp25cS7\",\n",
    "    \n",
    "]\n",
    "window_size_for_search = 100  # 100nt seems to make sense (+50 and -50 from the TSS or STOP)\n",
    "\n",
    "plot_lib_objs_dict = {lib: obj_dict[lib] for lib in plot_libs}\n",
    "overall_summaries_nested_dict = {}\n",
    "subset_nmd_types_name, subset_nmd_types = 'indifferent', 'indifferent'\n",
    "for gene in tqdm(plot_genes[:]):\n",
    "    # print(f\"{gene:-^50}\")\n",
    "    gene_summary_dicts = {}\n",
    "    try:\n",
    "        lib_df_dict_for_gene = {lib: lib_obj.mergedOnReads_df.query('gene_name == @gene')\n",
    "                                for lib, lib_obj in plot_lib_objs_dict.items()}\n",
    "        summary_dict = just_extract_gene_data(gtf_df, gene, lib_df_dict_for_gene, print_stats=False, window_size=window_size_for_search)\n",
    "        gene_summary_dicts[subset_nmd_types_name] = summary_dict\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        print(f\"Gene: {gene}\")\n",
    "        # print(f\"Continuing...\")\n",
    "        raise e\n",
    "    overall_summaries_nested_dict[gene] = gene_summary_dicts\n",
    "plotting_list_for_df = []\n",
    "\n",
    "dict_index = 0\n",
    "for gene, gene_dict in tqdm(overall_summaries_nested_dict.items()):\n",
    "    for nmd_target_type, target_type_dict in gene_dict.items():\n",
    "        print(f\"{gene} ({nmd_target_type}):\")\n",
    "        for lib, lib_dict in target_type_dict.items():\n",
    "            # print(f\"\\t{lib}:\")\n",
    "            # # print(f\"\\t\\tAdapted Near Stop:   {lib_dict['Adapted Near Stop']}\")\n",
    "            # print(f\"\\t\\tAdapted Near TSS:    {lib_dict['Adapted Near TSS']}\")\n",
    "            # print(f\"\\t\\tTotal Adapted:       {lib_dict['Total Adapted']}\")\n",
    "            # # print(f\"\\t\\tUnadapted Near Stop: {lib_dict['Unadapted Near Stop']}\")\n",
    "            # print(f\"\\t\\tUnadapted Near TSS:  {lib_dict['Unadapted Near TSS']}\")\n",
    "            # print(f\"\\t\\tTotal Unadapted:     {lib_dict['Total Unadapted']}\")\n",
    "            # try:\n",
    "            #     print(f\"\\t\\tAda Stop/Unada FL: {lib_dict['Adapted Near Stop']/lib_dict['Unadapted Near TSS']:.2%}\")\n",
    "            # except ZeroDivisionError:\n",
    "            #     print(f\"\\t\\tAda Stop/Unada FL: -- NO UNADAPTED FL --\")\n",
    "            plotting_list_for_df.append(\n",
    "                {\n",
    "                    'gene_name': gene,\n",
    "                    'gene_length': lib_dict['Gene Length'],\n",
    "                    'nmd_target_type': nmd_target_type,\n",
    "                    'lib': lib,\n",
    "                    # 'adapted_near_stop': lib_dict['Adapted Near Stop'],\n",
    "                    'adapted_near_tss': lib_dict['Adapted Near TSS'],\n",
    "                    'total_adapted': lib_dict['Total Adapted'],\n",
    "                    # 'unadapted_near_stop': lib_dict['Unadapted Near Stop'],\n",
    "                    'unadapted_near_tss': lib_dict['Unadapted Near TSS'],\n",
    "                    'total_unadapted': lib_dict['Total Unadapted'],\n",
    "                }\n",
    "            )\n",
    "\n",
    "plotting_summary_df = pd.DataFrame.from_records(plotting_list_for_df)\n",
    "plotting_summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "45587dd712452fde",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Before anything let's copy down code so we can regenerate the input df\n",
    "summary_df = plotting_summary_df.copy()\n",
    "\n",
    "lib_strain_dict = {'N2': 'wildtype',\n",
    "                   'S5': 'smg-5',\n",
    "                   'S6': 'smg-6',\n",
    "                   'S7': 'smg-7'}\n",
    "\n",
    "summary_df['strain'] = summary_df.lib.str[-2:].map(lib_strain_dict)\n",
    "\n",
    "summary_df['total_near_tss'] = summary_df['adapted_near_tss'] + summary_df['unadapted_near_tss']\n",
    "# summary_df['total_near_stop'] = summary_df['adapted_near_stop'] + summary_df['unadapted_near_stop']\n",
    "summary_df['total_reads'] = summary_df['total_adapted'] + summary_df['total_unadapted']\n",
    "\n",
    "# # There are a few different ways to do this!\n",
    "# summary_df['frac_near_tss'] = summary_df['total_near_tss'] / summary_df['total_reads']\n",
    "# \n",
    "# summary_df['frac_near_tss'] = summary_df['adapted_near_tss'] / summary_df['total_reads']\n",
    "\n",
    "summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ad2cf2facb4618b8",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import scipy.stats as stats\n",
    "from colorama import Fore, Style\n",
    "\n",
    "def run_stats_intralib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       strains=('wildtype', 'smg-5', 'smg-6'),\n",
    "                       output_file: Path = None):\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for strain in strains:\n",
    "        nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_Targets'\")\n",
    "        non_nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_NonTargets'\")\n",
    "        merge_df = nmd_df.merge(non_nmd_df,\n",
    "                                on=('gene_name', 'strain', 'rep', 'lib', 'name'),\n",
    "                                suffixes=('_nmd', '_non_nmd'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_nmd'], merge_df[f'{compare_col}_non_nmd'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[strain] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    output_string += f\"{'Strain':^8} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-' * 50 + '\\n'\n",
    "    for strain, test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{strain:>8} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string\n",
    "\n",
    "def run_stats_interlib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       comparisons=(('wildtype', 'smg-5'), ('wildtype', 'smg-6'), ('smg-5', 'smg-6')),\n",
    "                       nmd_type='NMD_NonTargets',\n",
    "                       output_file=None):\n",
    "    # Now let's do a similar comparison but between non-NMD-targets between libs\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for comp in comparisons:\n",
    "        nmd_df = df_in.query(\"strain == @comp[0] & nmd_target_type == @nmd_type\")\n",
    "        non_nmd_df = df_in.query(\"strain == @comp[1] & nmd_target_type == @nmd_type\")\n",
    "        drop_cols = ['nmd_target_type', 'strain', 'lib', 'name']\n",
    "        merge_df = nmd_df.drop(columns=drop_cols).merge(non_nmd_df.drop(columns=drop_cols), on=('gene_name', 'rep'), suffixes=('_lib1', '_lib2'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_lib1'], merge_df[f'{compare_col}_lib2'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[comp] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    \n",
    "    output_string += f\"{'Comp':^14} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-'*56 + '\\n'\n",
    "    \n",
    "    for (lib1, lib2), test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if lib1 == 'wildtype':\n",
    "            lib1 = 'wt'\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{lib1:>5} vs {lib2:<5} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "b915d643e42bdd98",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Maybe the NMD targets have higher adapted_TSS fractions in the WT not because they have high numbers of adapted_TSS reads, but b/c they have lower numbers of the denominator (all reads)\n",
    "# To remedy this, let's try comparing the adapted_TSS to the unadapted_TSS\n",
    "\n",
    "# Additionally, it would be nice to do an adapted RPM comparison against Adapt TSS reads. But this needs more reads!!\n",
    "\n",
    "# I need to circle WAY back to retain both adapt and unadapted TSSes.. UGH LETS DO IT\n",
    "in_df = summary_df.copy()\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type']).sort_index()\n",
    "better_df = in_df.loc[:, :, :, :, ['indifferent'], :]\n",
    "better_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c2b6fb02bbc134ce",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "min_total = 1\n",
    "min_adapted = 5\n",
    "\n",
    "def run_box_and_ttests(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    hover_columns = [\n",
    "        'name',\n",
    "        target_num,\n",
    "        target_denom,\n",
    "        'adapted_near_tss',\n",
    "        'unadapted_near_tss',\n",
    "        'total_reads',\n",
    "        'total_adapted',\n",
    "    ]\n",
    "    \n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=set(hover_columns),\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    fig.write_html(output_dir / f\"boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    # \n",
    "    # output_file = output_dir / f'boxPlotStats_{save_friendly_target_col}.txt'\n",
    "    # output_file.touch(exist_ok=True)\n",
    "    # with open(output_file, 'w') as f:\n",
    "    #     f.write(f\"Stats for {target_col}\\n\")\n",
    "    #     f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    # print(\"Targets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    # print(\"NonTargets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    # \n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nInter Library Stats:\\n\")\n",
    "    # run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "def plot_xy(input_df: pd.DataFrame,\n",
    "            target_num: str, target_denom: str,\n",
    "            x_col: str,\n",
    "            min_total_reads: int = 1,\n",
    "            min_adapted_reads: int = 1,\n",
    "            min_tss_ada_reads: int = 0,\n",
    "            output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "            log_y: bool = True,\n",
    "            log_x: bool = True,\n",
    "            color_by: str = 'strain',\n",
    "            template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df['near_tss'] = input_df['adapted_near_tss'] + input_df['unadapted_near_tss']\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    hover_columns = [\n",
    "        'name',\n",
    "        'gene_length',\n",
    "        target_num,\n",
    "        target_denom,\n",
    "        'adapted_near_tss',\n",
    "        'unadapted_near_tss',\n",
    "        'total_reads',\n",
    "        'total_adapted',\n",
    "    ]\n",
    "    \n",
    "    fig = px.scatter(box_df,\n",
    "                     x=x_col,\n",
    "                     y=target_col,\n",
    "                     color=color_by,\n",
    "                     color_discrete_map=colors_dict,\n",
    "                     hover_name='name',\n",
    "                     log_y=log_y,\n",
    "                     log_x=log_x,\n",
    "                     hover_data=set(hover_columns),\n",
    "                     template=template,\n",
    "                     # title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                     opacity=0.4,\n",
    "                     height=500,\n",
    "                     width=700,\n",
    "                     )\n",
    "    # if target_denom == 'total_reads':\n",
    "    #     fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    # else:\n",
    "    #     fig.update_yaxes(range=[-0.1, 1.1],\n",
    "    #                      title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(font=dict(size=18))\n",
    "    # fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    fig.write_html(output_dir / f\"boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "\n",
    "def run_cdf_and_ttests(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_total_reads: int = 1,\n",
    "                       min_adapted_reads: int = 1,\n",
    "                       min_tss_ada_reads: int = 0,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                       log_y: bool = True,\n",
    "                       log2_y: bool = False,\n",
    "                       template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.ecdf(box_df,\n",
    "                 facet_col='strain',\n",
    "                 x=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                 # height=500,\n",
    "                 # width=1000,\n",
    "                 )\n",
    "    fig.update_xaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    # fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    save_friendly_target_col += f\"_minTotal_{min_total_reads}_minAdapted_{min_adapted_reads}\"\n",
    "    fig.write_html(output_dir / f\"cdfPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"cdfPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    output_file = output_dir / f'cdfPlotStats_{save_friendly_target_col}.txt'\n",
    "    output_file.touch(exist_ok=True)\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(f\"Stats for {target_col}\\n\")\n",
    "        f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    print(\"Targets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    print(\"NonTargets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    \n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nInter Library Stats:\\n\")\n",
    "    run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "def run_xy_and_ttests(input_df: pd.DataFrame,\n",
    "                      target_num: str, target_denom: str,\n",
    "                      min_total_reads: int = 1,\n",
    "                      min_adapted_reads: int = 1,\n",
    "                      min_tss_ada_reads: int = 0,\n",
    "                      log_y: bool = True,\n",
    "                      extra_savename: str = '',\n",
    "                      output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "                      template: str = 'none') -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.scatter(box_df,\n",
    "                 x=target_num,\n",
    "                 y=target_denom,\n",
    "                 facet_col='strain',\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 hover_name='name',\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 log_y=log_y,\n",
    "                 title=f\"{target_num} x {target_denom} by Strain<br>\"\n",
    "                       f\"Min Total: {min_total_reads}, Min Adapt: {min_adapted_reads}, Min TSS Adapt: {min_tss_ada_reads}\",\n",
    "                 opacity=0.7,\n",
    "                 )\n",
    "    # fig2 = px.density_contour(box_df,\n",
    "    #              x=target_num,\n",
    "    #              y=target_denom,\n",
    "    #              color='nmd_target_type',\n",
    "    #              color_discrete_map=colors_dict,\n",
    "    #              hover_name='name',\n",
    "    #              hover_data=[target_num, target_denom],\n",
    "    #              template=template,\n",
    "    #              log_y=log_y,\n",
    "    #              title=f\"{target_num} x {target_denom} by Strain<br>\"\n",
    "    #                    f\"Min Total: {min_total_reads}, Min Adapt: {min_adapted_reads}, Min TSS Adapt: {min_tss_ada_reads}\",\n",
    "    #              height=500,\n",
    "    #              width=700,\n",
    "    #              )\n",
    "    # fig = go.Figure(data=fig1.data + fig2.data)\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=18,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    \n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "    \n",
    "    save_friendly_target_col += extra_savename\n",
    "    \n",
    "    fig.write_html(output_dir / f\"xyPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"xyPlot_{save_friendly_target_col}.png\", scale=5,\n",
    "                 height=500,\n",
    "                 width=700,)\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    # output_file = output_dir / f'boxPlotStats_{save_friendly_target_col}.txt'\n",
    "    # output_file.touch(exist_ok=True)\n",
    "    # with open(output_file, 'w') as f:\n",
    "    #     f.write(f\"Stats for {target_col}\\n\")\n",
    "    #     f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    # print(\"Targets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    # print(\"NonTargets\")\n",
    "    # run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    # \n",
    "    # with open(output_file, 'a') as f:\n",
    "    #     f.write(\"\\nInter Library Stats:\\n\")\n",
    "    # run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "min_adapteds = [\n",
    "    # 1,\n",
    "    # 2, 3, 4,\n",
    "    # 5,\n",
    "    10,\n",
    "]\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    # 'total_adapted',\n",
    "]\n",
    "for min_adapted in min_adapteds:\n",
    "    for target_denominator in target_denominators:\n",
    "        print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "        # run_box_and_ttests(\n",
    "        #     better_df, # better_df.query(\"strain == 'wildtype'\"),\n",
    "        #     'adapted_near_tss', target_denominator,\n",
    "        #     min_adapted_reads=min_adapted, min_tss_ada_reads=0, log_y=False)\n",
    "        # plot_xy(better_df,\n",
    "        #         'near_tss', target_denominator,\n",
    "        #         'gene_length',\n",
    "        #         min_total_reads=1, min_adapted_reads=min_adapted, min_tss_ada_reads=0,\n",
    "        #         log_y=False, log_x=True)\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3baee8a32c5a1f08",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# We need to produce a better gene_length information\n",
    "gtf_df['genomic_length'] = gtf_df['end'] - gtf_df['start']"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "10fb89c24d559f04",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# The magic of pandas:\n",
    "transcript_lengths = gtf_df.query(\"feature == 'exon'\").groupby(['gene_name', 'gene_id', 'transcript_id'])['genomic_length'].sum()\n",
    "# transcript_lengths.plot.hist(bins=100)\n",
    "transcript_lengths.reset_index().query(\"gene_name == 'unc-54'\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "aa570eed4f9db625",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# The values we were using before (we actually just used the largest of these)\n",
    "overal_transcript_length = gtf_df.query(\"feature == 'transcript'\").set_index(['gene_name', 'gene_id', 'transcript_id']).sort_index()['genomic_length']\n",
    "overal_transcript_length.plot.hist(bins=100)\n",
    "overal_transcript_length.reset_index().query(\"gene_name == 'unc-54'\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bf42531ddfb19728",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# let's try taking the average of the transcript lengths for each gene\n",
    "transcript_lengths = transcript_lengths.reset_index()\n",
    "gene_lengths = transcript_lengths.groupby('gene_name')['genomic_length'].agg(['mean', 'std'])\n",
    "gene_lengths.rename(columns={'mean': 'avg_transcript_length', 'std': 'std_transcript_length'}, inplace=True)\n",
    "gene_lengths.fillna(0, inplace=True)\n",
    "# I think a bunch of these tiny genes aren't even real.\n",
    "# The requirement for reads in the library means they'll get dropped so no worries.\n",
    "# gene_lengths.plot.hist(bins=100)\n",
    "gene_lengths.reset_index().query(\"gene_name == 'unc-54'\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "af881aedaff4391a",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "in_df = summary_df.copy()\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type']).sort_index()\n",
    "better_df = in_df.loc[:, :, :, :, ['indifferent'], :]\n",
    "better_df = better_df.reset_index().merge(gene_lengths.reset_index(), on='gene_name').set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type'])\n",
    "better_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5c6e036aaef84ff6",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "genes = [\n",
    "    'rps-22',\n",
    "    'ubl-1',\n",
    "    'rpl-30',\n",
    "    # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "    'rla-1',  # rev\n",
    "    'rps-25',  # rev\n",
    "    'Y37E3.8',  # rev\n",
    "    'rpl-10',\n",
    "    'rla-0',\n",
    "    'rpl-18',\n",
    "    'rps-3',\n",
    "    'rpl-38',\n",
    "    'rpl-24.1',\n",
    "    'hel-1',\n",
    "    'rpl-7A',\n",
    "    'ZK228.4',\n",
    "    'R06C1.4',\n",
    "    'rpl-12',\n",
    "    'pqn-70',\n",
    "    'rsp-5',\n",
    "    'rpl-1',\n",
    "    'C53H9.2',\n",
    "    'rpl-3',\n",
    "    'rsp-6',\n",
    "    'rpl-26',\n",
    "    'K08D12.3',\n",
    "    'aly-3',\n",
    "    'C45B2.8'  # 240815: New one!\n",
    "]\n",
    "def plot_xy(input_df: pd.DataFrame,\n",
    "            target_num: str, target_denom: str,\n",
    "            x_col: str,\n",
    "            min_total_reads: int = 1,\n",
    "            min_adapted_reads: int = 1,\n",
    "            min_tss_ada_reads: int = 0,\n",
    "            output_dir: Path = current_dir / 'plots' / 'decapping',\n",
    "            log_y: bool = True,\n",
    "            log_x: bool = True,\n",
    "            color_by: str = 'strain',\n",
    "            template: str = 'none',\n",
    "            trend_line=True,\n",
    "            x_cutoff_line_val=-1) -> None:\n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df['near_tss'] = input_df['adapted_near_tss'] + input_df['unadapted_near_tss']\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str) + ' rep' + box_df['rep'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    hover_columns = [\n",
    "        'name',\n",
    "        'gene_length',\n",
    "        target_num,\n",
    "        target_denom,\n",
    "        'adapted_near_tss',\n",
    "        'unadapted_near_tss',\n",
    "        'total_reads',\n",
    "        'total_adapted',\n",
    "    ]\n",
    "    if trend_line:\n",
    "        from scipy.optimize import curve_fit\n",
    "    \n",
    "        # Define the logistic function\n",
    "        def inverse_logistic(x, k, x0):\n",
    "            return 1 - (1 / (1 + np.exp(-k * (x - x0))))\n",
    "        \n",
    "        # Fit the logistic model to the data\n",
    "        x_data = box_df[x_col]\n",
    "        y_data = box_df[target_col]\n",
    "        \n",
    "        # Initial guess for the parameters\n",
    "        initial_guess = [1, np.median(x_data)]\n",
    "        \n",
    "        # Fit the model\n",
    "        params, _ = curve_fit(inverse_logistic, x_data, y_data, p0=initial_guess, maxfev=10000)\n",
    "        print(params)\n",
    "        \n",
    "        # Generate trend line data\n",
    "        x_fit = np.linspace(x_data.min(), x_data.max(), 100)\n",
    "        y_fit = inverse_logistic(x_fit, *params)\n",
    "        trendline_df = pd.DataFrame({x_col: x_fit, target_col: y_fit})\n",
    "    \n",
    "    fig = px.scatter(box_df,\n",
    "                     x=x_col,\n",
    "                     y=target_col,\n",
    "                     color=color_by,\n",
    "                     color_discrete_map=colors_dict,\n",
    "                     hover_name='name',\n",
    "                     log_y=log_y,\n",
    "                     log_x=log_x,\n",
    "                     hover_data=set(hover_columns),\n",
    "                     template=template,\n",
    "                     # title=f\"{target_col} by Strain<br>Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}\",\n",
    "                     opacity=0.4,\n",
    "                     height=500, # *1.5,\n",
    "                     width=700, # *1.5,\n",
    "                     )\n",
    "    if trend_line:\n",
    "        fig.add_trace(go.Scatter(x=trendline_df[x_col], y=trendline_df[target_col], mode='lines', name='Trend Line'))\n",
    "    if x_cutoff_line_val > 0:\n",
    "        fig.add_shape(\n",
    "            # Line Horizontal\n",
    "            dict(\n",
    "                type=\"line\",\n",
    "                x0=x_cutoff_line_val,\n",
    "                y0=0,\n",
    "                x1=x_cutoff_line_val,\n",
    "                y1=1,\n",
    "                line=dict(\n",
    "                    color=\"Red\",\n",
    "                    width=3,\n",
    "                    dash=\"dashdot\",\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    # if target_denom == 'total_reads':\n",
    "    #     fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    # else:\n",
    "    #     fig.update_yaxes(range=[-0.1, 1.1],\n",
    "    #                      title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(font=dict(size=18))\n",
    "    # fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_xyPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_xyPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    # 'total_adapted',\n",
    "]\n",
    "target_numinators = [\n",
    "    'near_tss',\n",
    "    # 'adapted_near_tss',\n",
    "]\n",
    "\n",
    "# Let's run it back:\n",
    "for min_adapted in min_adapteds:\n",
    "    for target_denominator in target_denominators:\n",
    "        for target_numinator in target_numinators:\n",
    "            print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "            better_df['is_nmd_target'] = better_df.index.get_level_values('gene_name').isin(genes)\n",
    "            # run_box_and_ttests(\n",
    "            #     better_df, # better_df.query(\"strain == 'wildtype'\"),\n",
    "            #     'adapted_near_tss', target_denominator,\n",
    "            #     min_adapted_reads=min_adapted, min_tss_ada_reads=0, log_y=False)\n",
    "            plot_xy(better_df,\n",
    "                    target_numinator, target_denominator,\n",
    "                    'avg_transcript_length',\n",
    "                    color_by='is_nmd_target',\n",
    "                    min_total_reads=1, min_adapted_reads=min_adapted, min_tss_ada_reads=0,\n",
    "                    log_y=False, log_x=True,\n",
    "                    trend_line=False,\n",
    "                    x_cutoff_line_val=800)\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "32452199d68f80d2",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "better_df.to_parquet(current_dir / 'preprocessed_data' / 'all_genes_avgTranscriptLength_and_nearTSS.parquet')\n",
    "print(\"Saved!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2d19aa8aed0de888",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# better_df['log(std_transcript_length)'] = np.log(better_df['std_transcript_length'])\n",
    "# fig = px.scatter(better_df.reset_index().query(\"total_reads >= 10\"),\n",
    "#                  x='gene_length',\n",
    "#                  y='avg_transcript_length',\n",
    "#                  hover_name='gene_name',\n",
    "#                  color='log(std_transcript_length)',\n",
    "#                  log_y=True,\n",
    "#                  log_x=True,\n",
    "#                  opacity=0.5,\n",
    "#                  template='none')\n",
    "# fig.update_yaxes(\n",
    "#     scaleanchor = \"x\",\n",
    "#     scaleratio = 1,\n",
    "#   )\n",
    "# fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f0f8b8a74d02ef5e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Picking genes that we know we can catch FL with:\n",
    "\n",
    "Using the near_tss/total_reads vs avg_transcript_length plot I am setting a ROUGH cutoff of around 50% captured and trying to pick out the NMD genes that that are above that threshold.\n",
    "\n",
    "More specifically, I am generally looking at the trend and calling a rough length that looks to intersect with 50%, then pickign any genes with a length below that!\n",
    "\n",
    "# Decided on a cutoff of ~800nt for avg_transcript_length\n",
    "\n",
    "I am going to take these genes over to `decappingQuantification2_metaPlots.ipynb` to see how they look in the context of the other data."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6f76fae7f5d8d6ee"
  },
  {
   "cell_type": "code",
   "source": [
    "genes = [ # Anything still commented below is because it is too long of a gene for use to consistently hit it's TSS\n",
    "    'rps-22',\n",
    "    'ubl-1',\n",
    "    'rpl-30',\n",
    "    # # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "    'rla-1',  # rev\n",
    "    'rps-25',  # rev\n",
    "    'Y37E3.8',  # rev\n",
    "    'rpl-10',\n",
    "    # 'rla-0',\n",
    "    'rpl-18',\n",
    "    'rps-3',\n",
    "    # 'rpl-38',\n",
    "    'rpl-24.1',\n",
    "    # 'hel-1',\n",
    "    # 'rpl-7A',\n",
    "    # 'ZK228.4',\n",
    "    'R06C1.4',\n",
    "    # 'rpl-12',\n",
    "    # 'pqn-70',\n",
    "    # 'rsp-5',\n",
    "    # 'rpl-1',\n",
    "    # 'C53H9.2',\n",
    "    # 'rpl-3',\n",
    "    # 'rsp-6',\n",
    "    'rpl-26',\n",
    "    'K08D12.3',\n",
    "    # 'aly-3',\n",
    "    'C45B2.8'  # 240815: New one!\n",
    "]"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c78dd219060bd9f1",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
