{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# decappingQuantification.ipynb\n",
    "## Marcus Viscardi,    April 15, 2024\n",
    "\n",
    "Check out `README.md` for a more overarching view of what is happening here.\n",
    "\n",
    "Specifically, the goal is to produce comparisons of true full-length reads (spanning out to the TSS) and the reads that have cleavages near their stop codons (and the other reads).\n",
    "\n",
    "The largest choice to make here, is if we want to manually annotate a subset of genes to look at for this, or if we should rather look into a reference-based approach. I think the reference-based approach is likely to be more robust, but it will require a big up-front investment in time to get the reference data in a usable format.\n",
    "\n",
    "Initially, I think I will start with the manual annotation approach, as it will be quicker to get started with and will allow me to get a better feel for the data and the types of things that I will be looking for."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5cc859e08f35b675"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "from typing import Tuple\n",
    "\n",
    "import nanoporePipelineCommon as npCommon\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "from icecream import ic\n",
    "from datetime import datetime\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from scipy.stats import mannwhitneyu, ks_2samp\n",
    "\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "def __time_formatter__():\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    return f\"ic: {now} | > \"\n",
    "ic.configureOutput(prefix=__time_formatter__)\n",
    "\n",
    "_ = ic(\"Imports done!\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "gtf_parquet_path = Path(\"/data16/marcus/genomes/plus_cerENO2_elegansRelease100/230327_allChrs_plus-cerENO2.gtf.parquet\")\n",
    "\n",
    "gtf_df = pd.read_parquet(gtf_parquet_path)\n",
    "\n",
    "current_dir = Path.cwd()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bbd97a94e7f3036b",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "obj_dict = {}\n",
    "libs_to_run = [\n",
    "    \"oldN2\",\n",
    "    # \"oldS6\",\n",
    "    \"newerN2\",\n",
    "    \"newerS6\",\n",
    "    \"newerS5\",\n",
    "    \"thirdN2\",\n",
    "    \"thirdS5\",\n",
    "    \"thirdS6\",\n",
    "    \"temp25cN2\",\n",
    "    \"temp25cS5\",\n",
    "    \"temp25cS6\",\n",
    "    \"temp25cS7\",\n",
    "]\n",
    "for lib in libs_to_run:\n",
    "    print(f\"\\nLoading {lib}...\", end=\"\")\n",
    "    obj = npCommon.NanoporeRun(run_nickname=lib)\n",
    "    obj_dict[lib] = obj\n",
    "    print(\" Done!\")\n",
    "    obj.load_mergedOnReads()\n",
    "    obj.load_nmd_targets(return_sambamobj=False)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "7914dbed126f9722",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "def histo_plot_ends(test_gene, gtf_df, obj_dict, flanking_dist=200):\n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    start, end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    \n",
    "    stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "    start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "    \n",
    "    fig, axes = plt.subplots(2, 1, figsize=(10, 5), sharex=True)\n",
    "    ax1, ax2 = axes\n",
    "    \n",
    "    sea.histplot(data=obj_dict['thirdN2'].mergedOnReads_df.query('gene_name == @test_gene and t5 == 0'),\n",
    "                 x='chr_pos',\n",
    "                 binwidth=10,\n",
    "                 # stat='percent',\n",
    "                 common_norm=False,\n",
    "                 ax=ax1,\n",
    "                 color='k',\n",
    "                 alpha=0.5)\n",
    "    ax1.set_ylabel('Unadapted Reads')\n",
    "    sea.histplot(data=obj_dict['thirdN2'].mergedOnReads_df.query('gene_name == @test_gene and t5 == 1'),\n",
    "                 x='chr_pos',\n",
    "                 binwidth=10,\n",
    "                 # stat='percent',\n",
    "                 common_norm=False,\n",
    "                 ax=ax2,\n",
    "                 color='r',\n",
    "                 alpha=0.5)\n",
    "    ax2.set_ylabel('Adapted Reads')\n",
    "    \n",
    "    for stop_loc in stop_locations:\n",
    "        ax1.axvline(stop_loc, color='r', linestyle='--')\n",
    "        ax2.axvline(stop_loc, color='r', linestyle='--')\n",
    "    for start_site in start_sites:\n",
    "        ax1.axvline(start_site, color='g', linestyle='--')\n",
    "        ax2.axvline(start_site, color='g', linestyle='--')\n",
    "    \n",
    "    ax1.set_xlim(start-flanking_dist, end+flanking_dist)\n",
    "    ax2.set_xlim(start-flanking_dist, end+flanking_dist)\n",
    "    \n",
    "    plt.title(test_gene)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "genes = [\n",
    "    'rps-22',\n",
    "    'ubl-1',\n",
    "    'rpl-30',\n",
    "    # 'eef-1A.1',  # rev  # Something weird is going on...\n",
    "    'rla-1',  # rev\n",
    "    'rps-25',  # rev\n",
    "    'Y37E3.8',  # rev\n",
    "    'rpl-10',\n",
    "    'rla-0',\n",
    "    'rpl-18',\n",
    "    'rps-3',\n",
    "    'rpl-38',\n",
    "    'rpl-24.1',\n",
    "    # NMD Targets:\n",
    "    # 'rpl-30',\n",
    "    # 'ubl-1',\n",
    "    # 'rps-22',\n",
    "    'hel-1',\n",
    "    'rpl-7A',\n",
    "    'ZK228.4',\n",
    "    'R06C1.4',\n",
    "    'rpl-12',\n",
    "    'pqn-70',\n",
    "    'rsp-5',\n",
    "    'rpl-1',\n",
    "    'C53H9.2',\n",
    "    'rpl-3',\n",
    "    'rsp-6',\n",
    "    'rpl-26',\n",
    "    'K08D12.3',\n",
    "    'aly-3',\n",
    "]\n",
    "for gene_to_plot in genes:\n",
    "    # histo_plot_ends(gene_to_plot, gtf_df, obj_dict, flanking_dist=100)\n",
    "    print(gene_to_plot)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bade5eb128de5f4a",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "UNANNOTATED_TSS_DICT = {  # these are manual annotations of stops that are not in the GTF file\n",
    "    'rpl-7A': [4_390_579, 4_390_464],\n",
    "}\n",
    "\n",
    "UNANNOTATED_STOPS_DICT = {  # these are manual annotations of stops that are not in the GTF file\n",
    "    'rps-22': [1_950_853],\n",
    "    'rpl-30': [10_436_332],\n",
    "}\n",
    "\n",
    "def get_other_read_end(row):\n",
    "    cigar = row['cigar']\n",
    "    read_pos = row['chr_pos']\n",
    "    if cigar is None:\n",
    "        raise ValueError(\"Cigar is None!\")\n",
    "    if read_pos is None:\n",
    "        raise ValueError(\"read_pos is None!\")\n",
    "    pattern = r\"(\\d*)([MDISN])\"\n",
    "    matches = re.findall(pattern, cigar)\n",
    "    for match in matches:\n",
    "        count, key = match\n",
    "        if key in ['M', 'D', 'N']:\n",
    "            read_pos += int(count)\n",
    "    return read_pos\n",
    "\n",
    "def near_a_tss_or_stop(row, _strand, tss_positions, stop_positions, col_for_pos='chr_pos', full_window_size=50, window_bias_downstream=0.50, make_window_correction=True) -> Tuple[bool, bool]:\n",
    "    chr_pos = row[col_for_pos]\n",
    "    if _strand == '-' and make_window_correction:\n",
    "        window_bias_downstream = 1 - window_bias_downstream\n",
    "    elif _strand != '+' and make_window_correction:\n",
    "        raise ValueError(f\"Strand must be '+' or '-' not {_strand}\")\n",
    "    \n",
    "    near_tss = False\n",
    "    near_stop = False\n",
    "    chr_pos_upstream = chr_pos - full_window_size * (1 - window_bias_downstream)\n",
    "    chr_pos_downstream = chr_pos + full_window_size * window_bias_downstream\n",
    "    for tss in set(tss_positions):\n",
    "        if chr_pos_upstream <= tss <= chr_pos_downstream:\n",
    "            near_tss = True\n",
    "            break\n",
    "    for stop in set(stop_positions):\n",
    "        if chr_pos_upstream <= stop <= chr_pos_downstream:\n",
    "            near_stop = True\n",
    "            break\n",
    "    return near_tss, near_stop\n",
    "\n",
    "def convert_near_tss_and_stop_to_hue(row, tss_color='g', stop_color='r', neither_color='k', both_color='y'):\n",
    "    near_tss, near_stop = row['near_tss'], row['near_stop']\n",
    "    if near_tss and near_stop:\n",
    "        return both_color\n",
    "    elif near_tss:\n",
    "        return tss_color\n",
    "    elif near_stop:\n",
    "        return stop_color\n",
    "    else:\n",
    "        return neither_color\n",
    "\n",
    "def plot_gene_data(gtf_df, test_gene, plot_lib_objs_dict, window_size=60,\n",
    "                   up_v_down_fraction=0.25, number_of_bins=40,\n",
    "                   flanking_gene_fraction=0.05, current_dir=None, print_stats=False,\n",
    "                   title_override=None, save_name_override=None):\n",
    "    \n",
    "    extra_stops_dict = UNANNOTATED_STOPS_DICT\n",
    "    \n",
    "    extra_tss_dict = UNANNOTATED_TSS_DICT\n",
    "    \n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    gene_start, gene_end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    gene_length = gene_end - gene_start\n",
    "    bin_width = gene_length / number_of_bins\n",
    "    flanking_dist = int(gene_length * flanking_gene_fraction)\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    if print_stats:\n",
    "        print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    if strand == '-':\n",
    "        if print_stats:\n",
    "            print(\"Because this gene is on the reverse strand, we will have to calculate out the other end of the reads with the CIGAR strings. This will take slightly longer.\")\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'end'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'end'].values\n",
    "        up_v_down_fraction = 1 - up_v_down_fraction\n",
    "        flipped = True\n",
    "    else:\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "        flipped = False\n",
    "        \n",
    "    stop_locations = list(stop_locations)\n",
    "    start_sites = list(start_sites)\n",
    "    \n",
    "    if test_gene in extra_stops_dict:\n",
    "        for stop_location in extra_stops_dict[test_gene]:\n",
    "            stop_locations.append(stop_location)\n",
    "    if test_gene in extra_tss_dict:\n",
    "        for tss_location in extra_tss_dict[test_gene]:\n",
    "            start_sites.append(tss_location)\n",
    "    \n",
    "    fig = plt.figure(figsize=(11, 1.5 * len(plot_lib_objs_dict)))\n",
    "    \n",
    "    gs = fig.add_gridspec(len(plot_lib_objs_dict), 5,\n",
    "                          width_ratios=[4, 4, 1, 1, 1.5],\n",
    "                          ) \n",
    "    \n",
    "    for i, (lib, lib_obj) in enumerate(plot_lib_objs_dict.items()):\n",
    "\n",
    "        if print_stats:\n",
    "            print(f\"\\t{lib}:\")\n",
    "        ax1, ax2, ax3, ax4, ax5 = [fig.add_subplot(gs[i, j]) for j in range(5)]\n",
    "        \n",
    "        if i == 0:\n",
    "            ax1.set_title(\"Unadapted 5' Ends\")\n",
    "            ax2.set_title(\"Adapted 5' Ends\")\n",
    "            ax3.set_title(\"Near TSS\")\n",
    "            ax4.set_title(\"Near Stop\")\n",
    "            ax5.set_title(\"Lib Stats\")\n",
    "        \n",
    "        gene_df = lib_obj.mergedOnReads_df.query('gene_name == @test_gene').copy()\n",
    "        gene_df.replace({'t5': {0: '-', 1: '+', '+': '+', '-': '-'}}, inplace=True)\n",
    "        if flipped:\n",
    "            gene_df.loc[:, 'chr_pos'] = gene_df.apply(get_other_read_end, axis=1)\n",
    "        gene_df.loc[:, 'near_tss'], gene_df.loc[:, 'near_stop'] = zip(*gene_df.apply(near_a_tss_or_stop, args=(strand, start_sites, stop_locations),\n",
    "            full_window_size=window_size, window_bias_downstream=up_v_down_fraction, axis=1))\n",
    "        gene_df.loc[:, 'hue'] = gene_df.apply(convert_near_tss_and_stop_to_hue, axis=1)\n",
    "        \n",
    "        unadapted_df = gene_df.query('t5 == 0 or t5 == \"-\"')\n",
    "        unadapted_value_counts = unadapted_df['near_tss'].value_counts(normalize=True)\n",
    "        if print_stats:\n",
    "            print(f\"\\t\\tUnadapted Read Distribution (near a TSS):\")\n",
    "            print(f\"\\t\\t\\t True: {unadapted_value_counts.get(True, 0.0):%}\")\n",
    "            print(f\"\\t\\t\\tFalse: {unadapted_value_counts.get(False, 0.0):%}\")\n",
    "        if not unadapted_df.empty:\n",
    "            sea.histplot(data=unadapted_df,\n",
    "                         x='chr_pos',\n",
    "                         hue='hue',\n",
    "                         palette={'r': 'r', 'g': 'g', 'k': 'k', 'y': 'y'},\n",
    "                         binwidth=bin_width,\n",
    "                         ax=ax1,\n",
    "                         edgecolor='k', linewidth=1.0,\n",
    "                         )\n",
    "            ax1.get_legend().remove()\n",
    "            data_ct = pd.crosstab(index=gene_df['t5'], columns=gene_df['near_tss'])\n",
    "            data_ct = data_ct.div(data_ct.sum(axis=1), axis=0) * 100\n",
    "            data_ct.sort_index(ascending=False, inplace=True)\n",
    "            \n",
    "            color_dict = {True: 'g', False: 'k', None: 'y'}\n",
    "            try:\n",
    "                data_ct.plot(kind='bar',\n",
    "                             stacked=True,\n",
    "                             ax=ax3,\n",
    "                             color=[color_dict[col] for col in data_ct.columns],\n",
    "                             )\n",
    "            except ValueError as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                print(data_ct)\n",
    "            ax3.set_ylabel (\"\") # (\"% Near TSS\")\n",
    "            ax3.get_legend().remove()\n",
    "        else:\n",
    "            ax1.text(0.5, 0.5, \"No Data\", horizontalalignment='center', verticalalignment='center', transform=ax1.transAxes)\n",
    "            print(lib_obj.mergedOnReads_df.query('gene_name == @test_gene').head())\n",
    "        ax1.set_ylabel(f\"-- {lib} --\")\n",
    "        \n",
    "        adapted_df = gene_df.query('t5 == 1 or t5 == \"+\"')\n",
    "        adapted_value_counts = adapted_df['near_stop'].value_counts(normalize=True)\n",
    "        if print_stats:\n",
    "            print(f\"\\t\\tAdapted Read Distribution (near a STOP):\")\n",
    "            print(f\"\\t\\t\\t True: {adapted_value_counts.get(True, 0.0):%}\")\n",
    "            print(f\"\\t\\t\\tFalse: {adapted_value_counts.get(False, 0.0):%}\")\n",
    "        if not adapted_df.empty:\n",
    "            sea.histplot(data=adapted_df,\n",
    "                         x='chr_pos',\n",
    "                         hue='hue',\n",
    "                         palette={'r': 'r', 'g': 'g', 'k': 'k', 'y': 'y'},\n",
    "                         binwidth=bin_width,\n",
    "                         ax=ax2,\n",
    "                         edgecolor='k', linewidth=1.0,\n",
    "                         )\n",
    "            ax2.get_legend().remove()\n",
    "            data_ct = pd.crosstab(index=gene_df['t5'], columns=gene_df['near_stop'])\n",
    "            data_ct = data_ct.div(data_ct.sum(axis=1), axis=0) * 100\n",
    "            data_ct.sort_index(ascending=False, inplace=True)\n",
    "            \n",
    "            color_dict = {True: 'r', False: 'k', None: 'y'}\n",
    "            try:\n",
    "                data_ct.plot(kind='bar',\n",
    "                             stacked=True,\n",
    "                             ax=ax4,\n",
    "                             color=[color_dict[col] for col in data_ct.columns],\n",
    "                             )\n",
    "            except ValueError as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                print(data_ct)\n",
    "            ax4.set_ylabel(\"\")  #(\"% Near Stop\")\n",
    "            ax4.get_legend().remove()\n",
    "        else:\n",
    "            ax2.text(0.5, 0.5, \"No Data\", horizontalalignment='center', verticalalignment='center', transform=ax2.transAxes)\n",
    "        ax2.set_ylabel(f\"\")\n",
    "        \n",
    "        ax1.set_xlim(gene_start-flanking_dist, gene_end+flanking_dist)\n",
    "        ax2.set_xlim(gene_start-flanking_dist, gene_end+flanking_dist)\n",
    "        ax1.set_xlabel(f\"\")\n",
    "        ax2.set_xlabel(f\"\")\n",
    "        ax1.set_xticklabels([])\n",
    "        ax2.set_xticklabels([])\n",
    "        \n",
    "        ax3.set_xlabel(f\"\")\n",
    "        ax4.set_xlabel(f\"\")\n",
    "        ax3.set_xticklabels(['Unad', 'Ad'], rotation=0)\n",
    "        ax4.set_xticklabels(['Unad', 'Ad'], rotation=0)\n",
    "        \n",
    "        precompute_transforms = {ax1: ax1.get_xaxis_transform(),\n",
    "                                 ax2: ax2.get_xaxis_transform()}\n",
    "        for stop_loc in set(stop_locations):\n",
    "            for ax in [ax1, ax2]:\n",
    "                ax.axvline(stop_loc, color='r', linestyle='--', alpha=0.5)\n",
    "                ax.axvspan(stop_loc-window_size*up_v_down_fraction,\n",
    "                           stop_loc+window_size*(1-up_v_down_fraction),\n",
    "                           color='r', alpha=0.1)\n",
    "                blend_transform = precompute_transforms[ax]\n",
    "                ax.text(stop_loc, -0.1, \"*\", rotation=0,\n",
    "                        verticalalignment='top', horizontalalignment='center',\n",
    "                        transform=blend_transform,\n",
    "                        fontsize=14)\n",
    "        for start_site in set(start_sites):\n",
    "            for ax in [ax1, ax2]:\n",
    "                ax.axvline(start_site, color='g', linestyle='--', alpha=0.5)\n",
    "                ax.axvspan(start_site-window_size*up_v_down_fraction,\n",
    "                           start_site+window_size*(1-up_v_down_fraction),\n",
    "                           color='g', alpha=0.1)\n",
    "                blend_transform = precompute_transforms[ax]\n",
    "                if flipped:\n",
    "                    tss_icon = \"<\"\n",
    "                else:\n",
    "                    tss_icon = \">\"\n",
    "                ax.text(start_site, -0.05, tss_icon, rotation=0,\n",
    "                        verticalalignment='top', horizontalalignment='center',\n",
    "                        transform=blend_transform,\n",
    "                        fontsize=12)\n",
    "        \n",
    "        \n",
    "                # Create annotations to put in ax5\n",
    "        ada_near_stop_count = adapted_df.query('near_stop').shape[0]\n",
    "        ada_near_tss_count = adapted_df.query('near_tss').shape[0]\n",
    "        unada_near_stop_count = unadapted_df.query('near_stop').shape[0]\n",
    "        unada_near_tss_count = unadapted_df.query('near_tss').shape[0]\n",
    "        ada_total_count = adapted_df.shape[0]\n",
    "        unada_total_count = unadapted_df.shape[0]\n",
    "        ada_stop_over_FL_string = f\"{ada_near_stop_count}/{unada_near_tss_count} ({ada_near_stop_count/unada_near_tss_count:.2%})\"\n",
    "        ada_tss_over_FL_string = f\"{ada_near_tss_count}/{unada_near_tss_count} ({ada_near_tss_count/unada_near_tss_count:.2%})\"\n",
    "        ax5.axis('off')\n",
    "        plot_string = (\n",
    "            f\"Adapted Near Stop:   {ada_near_stop_count}\\n\"\n",
    "            f\"Adapted Near TSS:    {ada_near_tss_count}\\n\"\n",
    "            f\"Total Adapted:       {ada_total_count}\\n\"\n",
    "            f\"Unadapted Near Stop: {unada_near_stop_count}\\n\"\n",
    "            f\"Unadapted Near TSS:  {unada_near_tss_count}\\n\"\n",
    "            f\"Total Unadapted:     {unada_total_count}\\n\"\n",
    "            f\"Ada Stop/Unada FL: {ada_stop_over_FL_string}\\n\"\n",
    "            f\"Ada  TSS/Unada FL: {ada_tss_over_FL_string}\"\n",
    "        )\n",
    "        ax5.text(-0.5, 0.4, plot_string,\n",
    "                 fontsize=8,\n",
    "                 fontfamily='monospace',\n",
    "                 horizontalalignment='left',\n",
    "                 verticalalignment='center',\n",
    "                 # transform=ax5.transAxes,\n",
    "                 )\n",
    "    if isinstance(title_override, str):\n",
    "        plt.suptitle(title_override)\n",
    "    else:\n",
    "        plt.suptitle(f\"{test_gene} ({strand} strand)\")\n",
    "    plt.tight_layout()\n",
    "    if current_dir is not None:\n",
    "        if isinstance(save_name_override, str):\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{save_name_override}.png\", dpi=300)\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{save_name_override}.svg\")\n",
    "        else:\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{test_gene}_5PrimeEnds.png\", dpi=300)\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{test_gene}_5PrimeEnds.svg\")\n",
    "    plt.show()\n",
    "\n",
    "# plot_libs = [\n",
    "#     'oldN2',\n",
    "#     'newerN2',\n",
    "#     'newerS6',\n",
    "#     'newerS5',\n",
    "#     'thirdN2',\n",
    "#     'thirdS5',\n",
    "#     'thirdS6',\n",
    "#     'temp25cN2',\n",
    "#     'temp25cS5',\n",
    "#     'temp25cS6',\n",
    "#     'temp25cS7',\n",
    "# ]\n",
    "\n",
    "plot_libs = libs_to_run\n",
    "\n",
    "plot_lib_objs_dict = {lib: obj_dict[lib] for lib in plot_libs}\n",
    "\n",
    "plot_genes = genes[:]\n",
    "\n",
    "for test_gene in plot_genes[:]:\n",
    "    print(f\"{test_gene:#^50}\")\n",
    "    try:\n",
    "        # plot_gene_data(gtf_df, test_gene, plot_lib_objs_dict, current_dir=current_dir, print_stats=False)\n",
    "        pass\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        continue"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3f48398c359a5e8e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "gene_ids_to_names_dict = gtf_df.query('feature == \"gene\"').set_index('gene_id')['gene_name'].to_dict()\n",
    "gene_names_to_ids_dict = gtf_df.query('feature == \"gene\"').set_index('gene_name')['gene_id'].to_dict()\n",
    "for gene_name in plot_genes:\n",
    "    gene_id = gene_names_to_ids_dict[gene_name]\n",
    "    print(f\"{gene_name:>10} -> {gene_id}\")\n",
    "gene_ids = [gene_names_to_ids_dict[gene_name] for gene_name in plot_genes]"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ff7d58655d3aafc3",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import pysam\n",
    "\n",
    "def load_simple_nmd_targets(nmd_targets_path, target_gene_ids_to_names_dict) -> pd.DataFrame:\n",
    "    target_gene_names_to_ids_dict = {v: k for k, v in target_gene_ids_to_names_dict.items()}\n",
    "    reads_dict = {}\n",
    "    with pysam.AlignmentFile(nmd_targets_path, 'rb') as test_bam:\n",
    "        for read in test_bam.fetch():\n",
    "            read_assigned_gene_id = read.get_tag('gA')\n",
    "            # print(read.query_name, read_assigned_gene_id, target_gene_ids_to_names_dict[read_assigned_gene_id], read.get_tag('t5'), read.get_tag('nC'))\n",
    "            reads_dict[read.query_name] = {\n",
    "                'gene_id': read_assigned_gene_id,\n",
    "                'gene_name': target_gene_ids_to_names_dict[read_assigned_gene_id],\n",
    "                't5': read.get_tag('t5'),\n",
    "                'nmd_assignment': read.get_tag('nC'),\n",
    "                'cigar': read.cigarstring,\n",
    "                'chr_pos': read.reference_start,\n",
    "                'chr': read.reference_name,\n",
    "                'seq': read.query_sequence,\n",
    "            }\n",
    "    df = pd.DataFrame(reads_dict).T\n",
    "    return df\n",
    "\n",
    "nmd_genes = []\n",
    "\n",
    "for lib, lib_obj in obj_dict.items():\n",
    "    print(f\"Loading {lib} NMD Targets from {lib_obj.nmd_targets_bam_path}...\", end=\"\")\n",
    "    lib_obj.load_nmd_targets(return_sambamobj=False)\n",
    "    print(\" Done!\")\n",
    "    lib_obj.nmd_targets_df = load_simple_nmd_targets(lib_obj.nmd_targets_bam_path, gene_ids_to_names_dict)\n",
    "    # print(lib_obj.nmd_targets_df.gene_name.unique())\n",
    "    nmd_genes.extend(lib_obj.nmd_targets_df.gene_name.unique())\n",
    "nmd_genes = list(set(nmd_genes))\n",
    "print(nmd_genes)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ec61ebb1416400d4",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "UNANNOTATED_TSS_DICT = {  # these are manual annotations of transcription start sites that are not in the GTF file\n",
    "    'rpl-7A': [4_390_579, 4_390_464],\n",
    "}\n",
    "\n",
    "UNANNOTATED_STOPS_DICT = {  # these are manual annotations of stops that are not in the GTF file\n",
    "    'rps-22': [1_950_853],  # cassette exon contains PTC\n",
    "    'rpl-30': [10_436_332],  # extension of first exon contains PTC\n",
    "    'rpl-7A': [4_389_880, 4_389_745],  # unannotated extension of 3rd exon contains PTC (second loc due to potential FS)\n",
    "    'rsp-6': [7_790_576],  # longer unannotated form of cassette exon causes frame shift\n",
    "    'C53H9.2': [1_833_467],  # shorter first exon leaves frame shift\n",
    "    'pqn-70': [11_226_879],  # skipped exon leads to a frame change\n",
    "    'rpl-1': [2_875_908],  # new orf started in the 3' extension of the first exon, causing a frame change\n",
    "    'R06C1.4': [11_931_081],  # unannotated cassette exon contains a frame shift\n",
    "    'rpl-3': [3_868_335],  # incorrectly annotated first intron 3' truncation (2nd exon extended) contains stop\n",
    "    'hel-2': [8_327_970],  # later 5' SS adds to 3rd exon and contains PTC\n",
    "    'rpl-12': [13_240_078],  # later 5' SS of second intron adds extension to 2nd exon with a PTC\n",
    "    'aly-3': [12_123_859], # huge extension of second exon contains as PTC\n",
    "}\n",
    "\n",
    "def plot_gene_data3(gtf_df, test_gene, plot_lib_df_dict, window_size=100,\n",
    "                   up_v_down_fraction=0.10, number_of_bins=40,\n",
    "                   flanking_gene_fraction=0.05, current_dir=None, print_stats=False,\n",
    "                   title_override=None, save_name_override=None, show_plot=True):\n",
    "    \n",
    "    min_reads_cutoff = 3\n",
    "    \n",
    "    extra_stops_dict = UNANNOTATED_STOPS_DICT\n",
    "    extra_tss_dict = UNANNOTATED_TSS_DICT\n",
    "    \n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    gene_start, gene_end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    gene_length = gene_end - gene_start\n",
    "    bin_width = gene_length / number_of_bins\n",
    "    flanking_dist = int(gene_length * flanking_gene_fraction)\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    if print_stats:\n",
    "        print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    if strand == '-':\n",
    "        if print_stats:\n",
    "            print(\"Because this gene is on the reverse strand, we will have to calculate out the other end of the reads with the CIGAR strings. This will take slightly longer.\")\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'end'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'end'].values\n",
    "        up_v_down_fraction = 1 - up_v_down_fraction\n",
    "        flipped = True\n",
    "    elif strand == '+':\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "        flipped = False\n",
    "    else:\n",
    "        raise ValueError(f\"Strand must be '+' or '-' not '{strand}'\")\n",
    "        \n",
    "    stop_locations = list(stop_locations)\n",
    "    start_sites = list(start_sites)\n",
    "    \n",
    "    if test_gene in extra_stops_dict:\n",
    "        for stop_location in extra_stops_dict[test_gene]:\n",
    "            stop_locations.append(stop_location)\n",
    "    if test_gene in extra_tss_dict:\n",
    "        for tss_location in extra_tss_dict[test_gene]:\n",
    "            start_sites.append(tss_location)\n",
    "    \n",
    "    fig = plt.figure(figsize=(11, 1.5 * len(plot_lib_df_dict)))\n",
    "    \n",
    "    gs = fig.add_gridspec(len(plot_lib_df_dict), 5,\n",
    "                          width_ratios=[4, 4, 1, 1, 1.5],\n",
    "                          )\n",
    "    summary_dict = {}\n",
    "    for i, (lib, lib_df) in enumerate(plot_lib_df_dict.items()):\n",
    "\n",
    "        if print_stats:\n",
    "            print(f\"\\t{lib}:\")\n",
    "        ax1, ax2, ax3, ax4, ax5 = [fig.add_subplot(gs[i, j]) for j in range(5)]\n",
    "        \n",
    "        if i == 0:\n",
    "            ax1.set_title(\"Unadapted 5' Ends\")\n",
    "            ax2.set_title(\"Adapted 5' Ends\")\n",
    "            ax3.set_title(\"Near TSS\")\n",
    "            ax4.set_title(\"Near Stop\")\n",
    "            ax5.set_title(\"Lib Stats\")\n",
    "        \n",
    "        gene_df = lib_df.query('gene_name == @test_gene').copy()\n",
    "        gene_df.replace({'t5': {0: '-', 1: '+', '+': '+', '-': '-'}}, inplace=True)\n",
    "        if flipped:\n",
    "            gene_df.loc[:, 'chr_pos'] = gene_df.apply(get_other_read_end, axis=1)\n",
    "        gene_df.loc[:, 'near_tss'], gene_df.loc[:, 'near_stop'] = zip(*gene_df.apply(near_a_tss_or_stop, args=(strand, start_sites, stop_locations),\n",
    "            full_window_size=window_size,window_bias_downstream=up_v_down_fraction, make_window_correction=False, axis=1))\n",
    "        gene_df.loc[:, 'hue'] = gene_df.apply(convert_near_tss_and_stop_to_hue, axis=1)\n",
    "        \n",
    "        unadapted_df = gene_df.query('t5 == \"-\"').copy()\n",
    "        adapted_df = gene_df.query('t5 == \"+\"').copy()\n",
    "        \n",
    "        # Create annotations to put in ax5\n",
    "        ada_near_stop_count = adapted_df.query('near_stop').shape[0]\n",
    "        ada_near_tss_count = adapted_df.query('near_tss').shape[0]\n",
    "        unada_near_stop_count = unadapted_df.query('near_stop').shape[0]\n",
    "        unada_near_tss_count = unadapted_df.query('near_tss').shape[0]\n",
    "        ada_total_count = adapted_df.shape[0]\n",
    "        unada_total_count = unadapted_df.shape[0]\n",
    "        try:\n",
    "            ada_stop_over_FL_string = f\"{ada_near_stop_count}/{unada_near_tss_count} ({ada_near_stop_count/unada_near_tss_count:.2%})\"\n",
    "            ada_tss_over_FL_string = f\"{ada_near_tss_count}/{unada_near_tss_count} ({ada_near_tss_count/unada_near_tss_count:.2%})\"\n",
    "        except ZeroDivisionError:\n",
    "            ada_stop_over_FL_string = f\"{ada_near_stop_count}/{unada_near_tss_count} (X.XX%)\"\n",
    "            ada_tss_over_FL_string = f\"{ada_near_tss_count}/{unada_near_tss_count} (X.XX%)\"\n",
    "        ax5.axis('off')\n",
    "        plot_string = (\n",
    "            f\"Adapted Near Stop:   {ada_near_stop_count}\\n\"\n",
    "            f\"Adapted Near TSS:    {ada_near_tss_count}\\n\"\n",
    "            f\"Total Adapted:       {ada_total_count}\\n\"\n",
    "            f\"Unadapted Near Stop: {unada_near_stop_count}\\n\"\n",
    "            f\"Unadapted Near TSS:  {unada_near_tss_count}\\n\"\n",
    "            f\"Total Unadapted:     {unada_total_count}\\n\"\n",
    "            f\"Ada Stop/Unada FL: {ada_stop_over_FL_string}\\n\"\n",
    "            f\"Ada  TSS/Unada FL: {ada_tss_over_FL_string}\"\n",
    "        )\n",
    "        summary_dict[lib] = {\n",
    "            'Adapted Near Stop': ada_near_stop_count,\n",
    "            'Adapted Near TSS': ada_near_tss_count,\n",
    "            'Total Adapted': ada_total_count,\n",
    "            'Unadapted Near Stop': unada_near_stop_count,\n",
    "            'Unadapted Near TSS': unada_near_tss_count,\n",
    "            'Total Unadapted': unada_total_count,\n",
    "        }\n",
    "        try:\n",
    "            unadapted_value_counts = unadapted_df['near_tss'].value_counts(normalize=True)\n",
    "            if print_stats:\n",
    "                print(f\"\\t\\tUnadapted Read Distribution (near a TSS):\")\n",
    "                print(f\"\\t\\t\\t True: {unadapted_value_counts.get(True, 0.0):%}\")\n",
    "                print(f\"\\t\\t\\tFalse: {unadapted_value_counts.get(False, 0.0):%}\")\n",
    "            if unadapted_df.shape[0] > min_reads_cutoff:\n",
    "                sea.histplot(data=unadapted_df,\n",
    "                             x='chr_pos',\n",
    "                             hue='hue',\n",
    "                             palette={'r': 'r', 'g': 'g', 'k': 'k', 'y': 'y'},\n",
    "                             binwidth=bin_width,\n",
    "                             ax=ax1,\n",
    "                             edgecolor='k', linewidth=1.0,\n",
    "                             )\n",
    "                ax1.get_legend().remove()\n",
    "            else:\n",
    "                ax1.text(0.5, 0.5, \"Insufficient Reads\\nFor Plotting\", horizontalalignment='center', verticalalignment='center', transform=ax1.transAxes)\n",
    "                print(lib_obj.mergedOnReads_df.query('gene_name == @test_gene').head())\n",
    "            ax1.set_ylabel(f\"-- {lib} --\")\n",
    "            \n",
    "            data_ct = pd.crosstab(index=gene_df['t5'], columns=gene_df['near_tss'])\n",
    "            data_ct = data_ct.div(data_ct.sum(axis=1), axis=0) * 100\n",
    "            data_ct.sort_index(ascending=False, inplace=True)\n",
    "            \n",
    "            color_dict = {True: 'g', False: 'k', None: 'y'}\n",
    "            try:\n",
    "                data_ct.plot(kind='bar',\n",
    "                             stacked=True,\n",
    "                             ax=ax3,\n",
    "                             color=[color_dict[col] for col in data_ct.columns],\n",
    "                             )\n",
    "            except ValueError as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                print(data_ct)\n",
    "            ax3.set_ylabel (\"\") # (\"% Near TSS\")\n",
    "            ax3.get_legend().remove()\n",
    "            \n",
    "            adapted_value_counts = adapted_df['near_stop'].value_counts(normalize=True)\n",
    "            if print_stats:\n",
    "                print(f\"\\t\\tAdapted Read Distribution (near a STOP):\")\n",
    "                print(f\"\\t\\t\\t True: {adapted_value_counts.get(True, 0.0):%}\")\n",
    "                print(f\"\\t\\t\\tFalse: {adapted_value_counts.get(False, 0.0):%}\")\n",
    "            if adapted_df.shape[0] > min_reads_cutoff:\n",
    "                sea.histplot(data=adapted_df,\n",
    "                             x='chr_pos',\n",
    "                             hue='hue',\n",
    "                             palette={'r': 'r', 'g': 'g', 'k': 'k', 'y': 'y'},\n",
    "                             binwidth=bin_width,\n",
    "                             ax=ax2,\n",
    "                             edgecolor='k', linewidth=1.0,\n",
    "                             )\n",
    "                ax2.get_legend().remove()\n",
    "            else:\n",
    "                ax2.text(0.5, 0.5, \"Insufficient Reads\\nFor Plotting\", horizontalalignment='center', verticalalignment='center', transform=ax2.transAxes)\n",
    "            \n",
    "            ax2.set_ylabel(f\"\")\n",
    "            \n",
    "            data_ct = pd.crosstab(index=gene_df['t5'], columns=gene_df['near_stop'])\n",
    "            data_ct = data_ct.div(data_ct.sum(axis=1), axis=0) * 100\n",
    "            data_ct.sort_index(ascending=False, inplace=True)\n",
    "            \n",
    "            color_dict = {True: 'r', False: 'k', None: 'y'}\n",
    "            try:\n",
    "                data_ct.plot(kind='bar',\n",
    "                             stacked=True,\n",
    "                             ax=ax4,\n",
    "                             color=[color_dict[col] for col in data_ct.columns],\n",
    "                             )\n",
    "            except ValueError as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                print(data_ct)\n",
    "            ax4.set_ylabel(\"\")\n",
    "            ax4.get_legend().remove()\n",
    "            \n",
    "            ax1.set_xlim(gene_start-flanking_dist, gene_end+flanking_dist)\n",
    "            ax2.set_xlim(gene_start-flanking_dist, gene_end+flanking_dist)\n",
    "            ax1.set_xlabel(f\"\")\n",
    "            ax2.set_xlabel(f\"\")\n",
    "            ax1.set_xticklabels([])\n",
    "            ax2.set_xticklabels([])\n",
    "            \n",
    "            ax3.set_xlabel(f\"\")\n",
    "            ax4.set_xlabel(f\"\")\n",
    "            ax3.set_xticks([0 ,1], labels=['Unad', 'Ad'], rotation=0)\n",
    "            ax4.set_xticks([0 ,1], labels=['Unad', 'Ad'], rotation=0)\n",
    "            \n",
    "            precompute_transforms = {ax1: ax1.get_xaxis_transform(),\n",
    "                                     ax2: ax2.get_xaxis_transform()}\n",
    "            for stop_loc in set(stop_locations):\n",
    "                for ax in [ax1, ax2]:\n",
    "                    ax.axvline(stop_loc, color='r', linestyle='--', alpha=0.5)\n",
    "                    ax.axvspan(stop_loc-window_size*up_v_down_fraction,\n",
    "                               stop_loc+window_size*(1-up_v_down_fraction),\n",
    "                               color='r', alpha=0.1)\n",
    "                    blend_transform = precompute_transforms[ax]\n",
    "                    ax.text(stop_loc, -0.1, \"*\", rotation=0,\n",
    "                            verticalalignment='top', horizontalalignment='center',\n",
    "                            transform=blend_transform,\n",
    "                            fontsize=14)\n",
    "            for start_site in set(start_sites):\n",
    "                for ax in [ax1, ax2]:\n",
    "                    ax.axvline(start_site, color='g', linestyle='--', alpha=0.5)\n",
    "                    ax.axvspan(start_site-window_size*up_v_down_fraction,\n",
    "                               start_site+window_size*(1-up_v_down_fraction),\n",
    "                               color='g', alpha=0.1)\n",
    "                    blend_transform = precompute_transforms[ax]\n",
    "                    if flipped:\n",
    "                        tss_icon = \"<\"\n",
    "                    else:\n",
    "                        tss_icon = \">\"\n",
    "                    ax.text(start_site, -0.05, tss_icon, rotation=0,\n",
    "                            verticalalignment='top', horizontalalignment='center',\n",
    "                            transform=blend_transform,\n",
    "                            fontsize=12)\n",
    "            ax5.text(-0.5, 0.4, plot_string,\n",
    "                     fontsize=8,\n",
    "                     fontfamily='monospace',\n",
    "                     horizontalalignment='left',\n",
    "                     verticalalignment='center',\n",
    "                     )\n",
    "        except Exception as e:\n",
    "            print(f\"Error: {e}\")\n",
    "    if isinstance(title_override, str):\n",
    "        plt.suptitle(title_override)\n",
    "    else:\n",
    "        plt.suptitle(f\"{test_gene} ({strand} strand)\")\n",
    "    plt.tight_layout()\n",
    "    if current_dir is not None:\n",
    "        if isinstance(save_name_override, str):\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{save_name_override}.png\", dpi=300)\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{save_name_override}.svg\")\n",
    "        else:\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{test_gene}_5PrimeEnds.png\", dpi=300)\n",
    "            plt.savefig(current_dir / 'plots' / '5PrimeEndsPerGene' / f\"{test_gene}_5PrimeEnds.svg\")\n",
    "    if show_plot:\n",
    "        plt.show()\n",
    "    return summary_dict\n",
    "\n",
    "\n",
    "def just_extract_gene_data(gtf_df, test_gene, plot_lib_df_dict, window_size=100,\n",
    "                           up_v_down_fraction=0.25, number_of_bins=40,\n",
    "                            flanking_gene_fraction=0.05, print_stats=False):    \n",
    "    extra_stops_dict = UNANNOTATED_STOPS_DICT\n",
    "    extra_tss_dict = UNANNOTATED_TSS_DICT\n",
    "    \n",
    "    gtf_for_gene = gtf_df.query(\"gene_name == @test_gene\")\n",
    "    gene_start, gene_end = gtf_for_gene.query(\"feature == 'gene'\").loc[:, ['start', 'end']].values[0]\n",
    "    gene_length = gene_end - gene_start\n",
    "    \n",
    "    strand = gtf_for_gene.query(\"feature == 'gene'\").loc[:, 'strand'].values[0]\n",
    "    if print_stats:\n",
    "        print(f\"Gene target ({test_gene}) on strand: {strand}\")\n",
    "    if strand == '-':\n",
    "        if print_stats:\n",
    "            print(\"Because this gene is on the reverse strand, we will have to calculate out the other end of the reads with the CIGAR strings. This will take slightly longer.\")\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'end'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'end'].values\n",
    "        up_v_down_fraction = 1 - up_v_down_fraction\n",
    "        flipped = True\n",
    "    else:\n",
    "        stop_locations = gtf_for_gene.query(\"feature == 'stop_codon'\").loc[:, 'start'].values\n",
    "        start_sites = gtf_for_gene.query(\"feature == 'transcript'\").loc[:, 'start'].values\n",
    "        flipped = False\n",
    "        \n",
    "    stop_locations = list(stop_locations)\n",
    "    start_sites = list(start_sites)\n",
    "    \n",
    "    if test_gene in extra_stops_dict:\n",
    "        for stop_location in extra_stops_dict[test_gene]:\n",
    "            stop_locations.append(stop_location)\n",
    "    if test_gene in extra_tss_dict:\n",
    "        for tss_location in extra_tss_dict[test_gene]:\n",
    "            start_sites.append(tss_location)\n",
    "    summary_dict_for_gene = {}\n",
    "    for i, (lib, lib_df) in enumerate(plot_lib_df_dict.items()):\n",
    "\n",
    "        if print_stats:\n",
    "            print(f\"\\t{lib}:\")\n",
    "        \n",
    "        gene_df = lib_df.query('gene_name == @test_gene').copy()\n",
    "        gene_df.replace({'t5': {0: '-', 1: '+', '+': '+', '-': '-'}}, inplace=True)\n",
    "        if flipped:\n",
    "            gene_df.loc[:, 'chr_pos'] = gene_df.apply(get_other_read_end, axis=1)\n",
    "        gene_df.loc[:, 'near_tss'], gene_df.loc[:, 'near_stop'] = zip(*gene_df.apply(near_a_tss_or_stop, args=(strand, start_sites, stop_locations),\n",
    "            full_window_size=window_size,window_bias_downstream=up_v_down_fraction, axis=1))\n",
    "        gene_df.loc[:, 'hue'] = gene_df.apply(convert_near_tss_and_stop_to_hue, axis=1)\n",
    "        \n",
    "        unadapted_df = gene_df.query('t5 == \"-\"').copy()\n",
    "        adapted_df = gene_df.query('t5 == \"+\"').copy()\n",
    "        \n",
    "        # Create annotations to put in ax5\n",
    "        ada_near_stop_count = adapted_df.query('near_stop').shape[0]\n",
    "        ada_near_tss_count = adapted_df.query('near_tss').shape[0]\n",
    "        unada_near_stop_count = unadapted_df.query('near_stop').shape[0]\n",
    "        unada_near_tss_count = unadapted_df.query('near_tss').shape[0]\n",
    "        ada_total_count = adapted_df.shape[0]\n",
    "        unada_total_count = unadapted_df.shape[0]\n",
    "        summary_dict_for_gene[lib] = {\n",
    "            'Adapted Near Stop': ada_near_stop_count,\n",
    "            'Adapted Near TSS': ada_near_tss_count,\n",
    "            'Total Adapted': ada_total_count,\n",
    "            'Unadapted Near Stop': unada_near_stop_count,\n",
    "            'Unadapted Near TSS': unada_near_tss_count,\n",
    "            'Total Unadapted': unada_total_count,\n",
    "        }\n",
    "    plt.show()\n",
    "    return summary_dict_for_gene\n",
    "\n",
    "# plot_libs = [\n",
    "#     # 'oldN2',\n",
    "#     'newerN2',\n",
    "#     'newerS6',\n",
    "#     'newerS5',\n",
    "#     'thirdN2',\n",
    "#     'thirdS5',\n",
    "#     'thirdS6',\n",
    "# ]\n",
    "\n",
    "plot_libs = libs_to_run\n",
    "\n",
    "plot_lib_objs_dict = {lib: obj_dict[lib] for lib in plot_libs}\n",
    "overall_summaries_nested_dict = {}\n",
    "for gene in nmd_genes[::-1]:\n",
    "    # if gene != \"rpl-7A\":\n",
    "    #     continue\n",
    "    print(f\"{gene:#^50}\")\n",
    "    gene_summary_dicts = {}\n",
    "    target_nmd_types = ('nmd_target', 'ambiguous', 'non_nmd_target')\n",
    "    subset_nmd_types_list = [target_nmd_types[0], target_nmd_types[:2], target_nmd_types[2], target_nmd_types[:]]\n",
    "    subset_nmd_types_names = ['NMD_Targets', 'NMD_TargetsAndAmbiguous', 'NMD_NonTargets', 'All_NMD_Types']\n",
    "    \n",
    "    for subset_nmd_types, subset_nmd_types_name in zip(subset_nmd_types_list, subset_nmd_types_names):\n",
    "        try:\n",
    "            lib_df_dict_for_gene = {lib: lib_obj.nmd_targets_df.query('gene_name == @gene & nmd_assignment in @subset_nmd_types')\n",
    "                                    for lib, lib_obj in plot_lib_objs_dict.items()}\n",
    "            # summary_dict = plot_gene_data3(gtf_df, gene, lib_df_dict_for_gene, print_stats=False,\n",
    "            #                                title_override=f\"{gene}\\nNMD Target Types: {subset_nmd_types}\",\n",
    "            #                                save_name_override=f\"{gene}_{subset_nmd_types_name}\",\n",
    "            #                                current_dir=current_dir,)\n",
    "            summary_dict = just_extract_gene_data(gtf_df, gene, lib_df_dict_for_gene, print_stats=False)\n",
    "            gene_summary_dicts[subset_nmd_types_name] = summary_dict\n",
    "        except Exception as e:\n",
    "            print(f\"Error: {e}\")\n",
    "            print(f\"Gene: {gene}\")\n",
    "            print(f\"NMD Types: {subset_nmd_types}\")\n",
    "            print(f\"Continuing...\")\n",
    "    overall_summaries_nested_dict[gene] = gene_summary_dicts\n",
    "overall_summaries_nested_dict"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "550481b6553cb58",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "from pprint import pprint\n",
    "plotting_list_for_df = []\n",
    "\n",
    "dict_index = 0\n",
    "for gene, gene_dict in overall_summaries_nested_dict.items():\n",
    "    for nmd_target_type, target_type_dict in gene_dict.items():\n",
    "        print(f\"{gene} ({nmd_target_type}):\")\n",
    "        for lib, lib_dict in target_type_dict.items():\n",
    "            print(f\"\\t{lib}:\")\n",
    "            print(f\"\\t\\tAdapted Near Stop:   {lib_dict['Adapted Near Stop']}\")\n",
    "            print(f\"\\t\\tAdapted Near TSS:    {lib_dict['Adapted Near TSS']}\")\n",
    "            print(f\"\\t\\tTotal Adapted:       {lib_dict['Total Adapted']}\")\n",
    "            print(f\"\\t\\tUnadapted Near Stop: {lib_dict['Unadapted Near Stop']}\")\n",
    "            print(f\"\\t\\tUnadapted Near TSS:  {lib_dict['Unadapted Near TSS']}\")\n",
    "            print(f\"\\t\\tTotal Unadapted:     {lib_dict['Total Unadapted']}\")\n",
    "            try:\n",
    "                print(f\"\\t\\tAda Stop/Unada FL: {lib_dict['Adapted Near Stop']/lib_dict['Unadapted Near TSS']:.2%}\")\n",
    "            except ZeroDivisionError:\n",
    "                print(f\"\\t\\tAda Stop/Unada FL: -- NO UNADAPTED FL --\")\n",
    "            plotting_list_for_df.append(\n",
    "                {\n",
    "                    'gene_name': gene,\n",
    "                    'nmd_target_type': nmd_target_type,\n",
    "                    'lib': lib,\n",
    "                    'adapted_near_stop': lib_dict['Adapted Near Stop'],\n",
    "                    'adapted_near_tss': lib_dict['Adapted Near TSS'],\n",
    "                    'total_adapted': lib_dict['Total Adapted'],\n",
    "                    'unadapted_near_stop': lib_dict['Unadapted Near Stop'],\n",
    "                    'unadapted_near_tss': lib_dict['Unadapted Near TSS'],\n",
    "                    'total_unadapted': lib_dict['Total Unadapted'],\n",
    "                }\n",
    "            )"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "e5eb7af05556de14",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plotting_summary_df = pd.DataFrame.from_records(plotting_list_for_df)\n",
    "plotting_summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bbfabd1ac284d67e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plotting_summary_df['nmd_active'] = plotting_summary_df.lib.str.contains('N2')\n",
    "plotting_summary_df['ada_tss/unada_tss'] = plotting_summary_df['adapted_near_tss'] / plotting_summary_df['unadapted_near_tss']\n",
    "plotting_summary_df['ada_stop/unada_tss'] = plotting_summary_df['adapted_near_stop'] / plotting_summary_df['unadapted_near_tss']\n",
    "plotting_summary_df.nmd_target_type.unique()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c2bc67d86159dbfd",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "round_inf_to = 4\n",
    "plotting_summary_df['ada_tss/unada_tss'] = plotting_summary_df['ada_tss/unada_tss'].replace({np.inf: round_inf_to, -np.inf: -round_inf_to})\n",
    "plotting_summary_df['ada_stop/unada_tss'] = plotting_summary_df['ada_stop/unada_tss'].replace({np.inf: round_inf_to, -np.inf: -round_inf_to})\n",
    "\n",
    "sea.histplot(data=plotting_summary_df.query(\"nmd_target_type == 'NMD_Targets'\"),\n",
    "             x='ada_stop/unada_tss',\n",
    "             hue='lib',\n",
    "             # element='step',\n",
    "             bins=20,\n",
    "             multiple='stack',\n",
    "             fill=True,\n",
    "             # discrete=True,\n",
    "             log_scale=(True, False),\n",
    "             )\n",
    "plt.xscale('symlog')\n",
    "plt.show()\n",
    "sea.histplot(data=plotting_summary_df.query(\"nmd_target_type == 'NMD_Targets'\"),\n",
    "             x='ada_tss/unada_tss',\n",
    "             hue='lib',\n",
    "             # element='step',\n",
    "             binwidth=0.0001,\n",
    "             multiple='stack',\n",
    "             fill=True,\n",
    "             # discrete=True,\n",
    "             # log_scale=(True, False),\n",
    "             )\n",
    "plt.xscale('symlog', linthresh=0.001)\n",
    "plt.show()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3294a6828d8c2330",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import seaborn.objects as so\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "linthresh = 0.0001\n",
    "jitter_denominator = 4\n",
    "\n",
    "# xcols = ['ada_tss/unada_tss', 'ada_stop/unada_tss']\n",
    "xcols = ['ada_tss/total_tss', 'ada_stop/total_tss']\n",
    "ycol = 'total_unadapted'\n",
    "\n",
    "# lib_colors = {\n",
    "#     'newerN2': 'red',\n",
    "#     'newerS6': 'grey',\n",
    "#     'newerS5': 'black',\n",
    "#     'thirdN2': 'orange',\n",
    "#     'thirdS6': 'grey',\n",
    "#     'thirdS5': 'black',\n",
    "# }\n",
    "lib_colors = {\n",
    "    'newerN2': 'grey',\n",
    "    'newerS6': 'firebrick',\n",
    "    'newerS5': 'darkorange',\n",
    "    'thirdN2': 'black',\n",
    "    'thirdS6': 'tomato',\n",
    "    'thirdS5': 'orangered',\n",
    "}\n",
    "\n",
    "# fig, axes = plt.subplots(2, 2, figsize=(8, 8))\n",
    "# ((ax1, ax2), (ax3, ax4)) = axes\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 4))\n",
    "ax1, ax2 = axes\n",
    "\n",
    "nmd_types_to_plot = [\n",
    "    'NMD_Targets',\n",
    "    # 'NMD_NonTargets',\n",
    "]\n",
    "plot_df = plotting_summary_df.query(\"nmd_target_type in @nmd_types_to_plot\").copy()\n",
    "\n",
    "plot_df['total_tss'] = plot_df['adapted_near_tss'] + plot_df['unadapted_near_tss']\n",
    "plot_df['total_stop'] = plot_df['adapted_near_stop'] + plot_df['unadapted_near_tss']\n",
    "\n",
    "plot_df['ada_tss/total_tss'] = plot_df['adapted_near_tss'] / plot_df['total_tss']\n",
    "plot_df['ada_stop/total_tss'] = plot_df['adapted_near_stop'] / plot_df['total_tss']\n",
    "\n",
    "plot_df['lib_colors'] = plot_df.lib.map(lib_colors)\n",
    "plot_df[xcols[0]] = plot_df[xcols[0]].transform(lambda x: np.random.uniform(\n",
    "    -linthresh/jitter_denominator,\n",
    "    linthresh/jitter_denominator,\n",
    ") if x == 0 else x)\n",
    "plot_df[xcols[1]] = plot_df[xcols[1]].transform(lambda x: np.random.uniform(\n",
    "    -linthresh/jitter_denominator,\n",
    "    linthresh/jitter_denominator,\n",
    ") if x == 0 else x)\n",
    "\n",
    "for gene in plot_df.gene_name.unique():\n",
    "    sub_gene_plot_df = plot_df.query(\"gene_name == @gene\")\n",
    "    # sub_gene_plot_df.plot(x='ada_tss/unada_tss',\n",
    "    #                       y='total_unadapted',\n",
    "    #                       kind='line',\n",
    "    #                       linestyle=':',\n",
    "    #                       ax=ax1,\n",
    "    #                       c='black',\n",
    "    #                       alpha=0.25)\n",
    "    # sub_gene_plot_df.plot(x='ada_stop/unada_tss',\n",
    "    #                       y='total_unadapted',\n",
    "    #                       kind='line',\n",
    "    #                       linestyle=':',\n",
    "    #                       ax=ax2,\n",
    "    #                       c='black',\n",
    "    #                       alpha=0.25)\n",
    "    for lib in sub_gene_plot_df.lib.unique():\n",
    "        sub_lib_plot_df = sub_gene_plot_df.query(\"lib == @lib\")\n",
    "        marker_types = ['o', '+']\n",
    "        for i, nmd_target_type in enumerate(sub_lib_plot_df.nmd_target_type.unique()):\n",
    "            sub_lib_type_plot_df = sub_gene_plot_df.query(\"lib == @lib & nmd_target_type == @nmd_target_type\")\n",
    "            sub_lib_type_plot_df.plot(x=xcols[0],\n",
    "                                 y=ycol,\n",
    "                                 kind='scatter',\n",
    "                                 marker=marker_types[i],\n",
    "                                 ax=ax1,\n",
    "                                 c=sub_lib_type_plot_df['lib_colors'],\n",
    "                                 alpha=0.5,\n",
    "                                 zorder=100)\n",
    "            sub_lib_type_plot_df.plot(x=xcols[1],\n",
    "                                 y=ycol,\n",
    "                                 kind='scatter',\n",
    "                                 marker=marker_types[i],\n",
    "                                 ax=ax2,\n",
    "                                 c=sub_lib_type_plot_df['lib_colors'],\n",
    "                                 alpha=0.5,\n",
    "                                 zorder=100)\n",
    "for ax in axes.flat:\n",
    "    ax.set_xscale('symlog', linthresh=linthresh)\n",
    "    ax.set_xlim(-linthresh, 10)\n",
    "    ax.set_yscale('symlog', linthresh=10)\n",
    "    legend = ax.get_legend()\n",
    "    if legend is not None:\n",
    "        legend.remove()\n",
    "    ax.axvline(-linthresh/jitter_denominator,\n",
    "               color='k', linestyle='--',\n",
    "               alpha=0.25)\n",
    "    ax.axvline(linthresh/jitter_denominator,\n",
    "               color='k', linestyle='--',\n",
    "               alpha=0.25)\n",
    "    ax.annotate(' Points at zero jittered',\n",
    "                xy=(-linthresh/jitter_denominator, 0),\n",
    "                xytext=(-linthresh/jitter_denominator, 0),\n",
    "                rotation=90,\n",
    "                alpha=0.75,\n",
    "                fontsize=8,\n",
    "                horizontalalignment='right',\n",
    "                verticalalignment='bottom',\n",
    "                )\n",
    "\n",
    "ax1.set_xlabel(\"decapped / total full length\")\n",
    "ax1.set_ylabel(\"Total Unadapted\")\n",
    "ax2.set_xlabel(\"cleaved / total full length\")\n",
    "ax2.set_ylabel(\"Total Unadapted\")\n",
    "\n",
    "ax1.set_ylim(0, 1e4)\n",
    "ax2.set_ylim(0, 1e4)\n",
    "\n",
    "legend_items = [mlines.Line2D([],\n",
    "                              [],\n",
    "                              color=lib_colors[lib],\n",
    "                              marker='o',\n",
    "                              linestyle='',\n",
    "                              label=lib) for lib in plot_df.lib.unique()]\n",
    "if len(plot_df.nmd_target_type.unique()) > 1:\n",
    "    legend_items.extend([mlines.Line2D([],\n",
    "                                       [],\n",
    "                                       color='black',\n",
    "                                       marker=marker,\n",
    "                                       linestyle='',\n",
    "                                       label=nmd_target_type) for nmd_target_type, marker in zip(plot_df.nmd_target_type.unique(), marker_types)])\n",
    "    fig.legend(handles=legend_items, title='Library and NMD Type',\n",
    "               loc='upper center',\n",
    "               bbox_to_anchor=(0.0, 1.06, 1.0, 0.102),\n",
    "               ncols=4,\n",
    "               mode='expand',\n",
    "               borderaxespad=0.0,\n",
    "               facecolor='w',\n",
    "               framealpha=1,\n",
    "               )\n",
    "else:\n",
    "    fig.legend(handles=legend_items, title='Library',\n",
    "               loc='upper center',\n",
    "               bbox_to_anchor=(0.0, 1.00, 1.0, 0.102),\n",
    "               ncols=6,\n",
    "               mode='expand',\n",
    "               borderaxespad=0.0,\n",
    "               facecolor='w',\n",
    "               framealpha=1,\n",
    "               )\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.draw()\n",
    "save_dir = current_dir / 'plots' / 'decappingAndCleavage'\n",
    "save_dir.mkdir(exist_ok=True, parents=True)\n",
    "plt.savefig(save_dir / f\"SummaryPlot{len(plot_df.nmd_target_type.unique())}.png\",\n",
    "            dpi=300, bbox_inches='tight')\n",
    "plt.savefig(save_dir / f\"SummaryPlot{len(plot_df.nmd_target_type.unique())}.svg\",\n",
    "            bbox_inches='tight')\n",
    "plt.show()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "570251f37665e446",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import seaborn.objects as so\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "linthresh = 1\n",
    "jitter_denominator = 10\n",
    "\n",
    "# xcols = ['ada_tss/unada_tss', 'ada_stop/unada_tss']\n",
    "xcols = ['adapted_near_tss', 'adapted_near_stop']\n",
    "ycol = 'total_tss'\n",
    "\n",
    "lib_colors = {\n",
    "    'newerN2': 'grey',\n",
    "    'newerS6': 'firebrick',\n",
    "    'newerS5': 'darkorange',\n",
    "    'thirdN2': 'black',\n",
    "    'thirdS6': 'tomato',\n",
    "    'thirdS5': 'orangered',\n",
    "}\n",
    "\n",
    "# fig, axes = plt.subplots(2, 2, figsize=(8, 8))\n",
    "# ((ax1, ax2), (ax3, ax4)) = axes\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 4))\n",
    "ax1, ax2 = axes\n",
    "\n",
    "nmd_types_to_plot = [\n",
    "    'NMD_Targets',\n",
    "    'NMD_NonTargets',\n",
    "]\n",
    "plot_df = plotting_summary_df.query(\"nmd_target_type in @nmd_types_to_plot\").copy()\n",
    "\n",
    "plot_df['total_tss'] = plot_df['adapted_near_tss'] + plot_df['unadapted_near_tss']\n",
    "plot_df['total_stop'] = plot_df['adapted_near_stop'] + plot_df['unadapted_near_tss']\n",
    "\n",
    "plot_df['ada_tss/total_tss'] = plot_df['adapted_near_tss'] / plot_df['total_tss']\n",
    "plot_df['ada_stop/total_tss'] = plot_df['adapted_near_stop'] / plot_df['total_tss']\n",
    "\n",
    "plot_df['lib_colors'] = plot_df.lib.map(lib_colors)\n",
    "plot_df[xcols[0]] = plot_df[xcols[0]].transform(lambda x: np.random.uniform(\n",
    "    -linthresh/jitter_denominator,\n",
    "    linthresh/jitter_denominator,\n",
    ") if x == 0 else x)\n",
    "plot_df[xcols[1]] = plot_df[xcols[1]].transform(lambda x: np.random.uniform(\n",
    "    -linthresh/jitter_denominator,\n",
    "    linthresh/jitter_denominator,\n",
    ") if x == 0 else x)\n",
    "\n",
    "for gene in plot_df.gene_name.unique():\n",
    "    sub_gene_plot_df = plot_df.query(\"gene_name == @gene\")\n",
    "    for lib in sub_gene_plot_df.lib.unique():\n",
    "        sub_lib_plot_df = sub_gene_plot_df.query(\"lib == @lib\")\n",
    "        marker_types = ['o', '+']\n",
    "        for i, nmd_target_type in enumerate(sub_lib_plot_df.nmd_target_type.unique()):\n",
    "            sub_lib_type_plot_df = sub_gene_plot_df.query(\"lib == @lib & nmd_target_type == @nmd_target_type\")\n",
    "            sub_lib_type_plot_df.plot(x=xcols[0],\n",
    "                                 y=ycol,\n",
    "                                 kind='scatter',\n",
    "                                 marker=marker_types[i],\n",
    "                                 ax=ax1,\n",
    "                                 c=sub_lib_type_plot_df['lib_colors'],\n",
    "                                 alpha=0.5,\n",
    "                                 zorder=100)\n",
    "            sub_lib_type_plot_df.plot(x=xcols[1],\n",
    "                                 y=ycol,\n",
    "                                 kind='scatter',\n",
    "                                 marker=marker_types[i],\n",
    "                                 ax=ax2,\n",
    "                                 c=sub_lib_type_plot_df['lib_colors'],\n",
    "                                 alpha=0.5,\n",
    "                                 zorder=100)\n",
    "for ax in axes.flat:\n",
    "    ax.set_xscale('symlog', linthresh=linthresh)\n",
    "    ax.set_xlim(-linthresh, 1e2)\n",
    "    ax.set_yscale('symlog', linthresh=1)\n",
    "    legend = ax.get_legend()\n",
    "    if legend is not None:\n",
    "        legend.remove()\n",
    "    ax.axvline(-linthresh/jitter_denominator,\n",
    "               color='k', linestyle='--',\n",
    "               alpha=0.25)\n",
    "    ax.axvline(linthresh/jitter_denominator,\n",
    "               color='k', linestyle='--',\n",
    "               alpha=0.25)\n",
    "    ax.annotate(' Points at zero jittered',\n",
    "                xy=(-linthresh/jitter_denominator, 0),\n",
    "                xytext=(-linthresh/jitter_denominator, 0),\n",
    "                rotation=90,\n",
    "                alpha=0.75,\n",
    "                fontsize=8,\n",
    "                horizontalalignment='right',\n",
    "                verticalalignment='bottom',\n",
    "                )\n",
    "\n",
    "ax1.set_xlabel(\"decapped\")\n",
    "ax1.set_ylabel(\"Total Near TSS\")\n",
    "ax2.set_xlabel(\"cleaved\")\n",
    "ax2.set_ylabel(\"Total Near TSS\")\n",
    "\n",
    "ax1.set_ylim(-1e0, 1e4)\n",
    "ax2.set_ylim(-1e0, 1e4)\n",
    "\n",
    "legend_items = [mlines.Line2D([],\n",
    "                              [],\n",
    "                              color=lib_colors[lib],\n",
    "                              marker='o',\n",
    "                              linestyle='',\n",
    "                              label=lib) for lib in plot_df.lib.unique()]\n",
    "if len(plot_df.nmd_target_type.unique()) > 1:\n",
    "    legend_items.extend([mlines.Line2D([],\n",
    "                                       [],\n",
    "                                       color='black',\n",
    "                                       marker=marker,\n",
    "                                       linestyle='',\n",
    "                                       label=nmd_target_type) for nmd_target_type, marker in zip(plot_df.nmd_target_type.unique(), marker_types)])\n",
    "    fig.legend(handles=legend_items, title='Library and NMD Type',\n",
    "               loc='upper center',\n",
    "               bbox_to_anchor=(0.0, 1.06, 1.0, 0.102),\n",
    "               ncols=4,\n",
    "               mode='expand',\n",
    "               borderaxespad=0.0,\n",
    "               facecolor='w',\n",
    "               framealpha=1,\n",
    "               )\n",
    "else:\n",
    "    fig.legend(handles=legend_items, title='Library',\n",
    "               loc='upper center',\n",
    "               bbox_to_anchor=(0.0, 1.00, 1.0, 0.102),\n",
    "               ncols=6,\n",
    "               mode='expand',\n",
    "               borderaxespad=0.0,\n",
    "               facecolor='w',\n",
    "               framealpha=1,\n",
    "               )\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.draw()\n",
    "save_dir = current_dir / 'plots' / 'decappingAndCleavage'\n",
    "save_dir.mkdir(exist_ok=True, parents=True)\n",
    "plt.savefig(save_dir / f\"SummaryPlot{len(plot_df.nmd_target_type.unique())}.png\",\n",
    "            dpi=300, bbox_inches='tight')\n",
    "plt.savefig(save_dir / f\"SummaryPlot{len(plot_df.nmd_target_type.unique())}.svg\",\n",
    "            bbox_inches='tight')\n",
    "plt.show()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "8ec508c2a7e704c7",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "xcols = ['adapted_near_tss', 'adapted_near_stop']\n",
    "ycol = 'total_tss'\n",
    "\n",
    "lib_colors = {\n",
    "    'newerN2': 'grey',\n",
    "    'newerS6': 'firebrick',\n",
    "    'newerS5': 'darkorange',\n",
    "    'thirdN2': 'black',\n",
    "    'thirdS6': 'tomato',\n",
    "    'thirdS5': 'orangered',\n",
    "}\n",
    "\n",
    "nmd_types_to_plot = [\n",
    "    'NMD_Targets',\n",
    "    'NMD_NonTargets',\n",
    "]\n",
    "plot_df = plotting_summary_df.query(\"nmd_target_type in @nmd_types_to_plot\").copy()\n",
    "\n",
    "plot_df['total_tss'] = plot_df['adapted_near_tss'] + plot_df['unadapted_near_tss']\n",
    "plot_df['total_stop'] = plot_df['adapted_near_stop'] + plot_df['unadapted_near_stop']\n",
    "\n",
    "plot_df['ada_tss/total_tss'] = plot_df['adapted_near_tss'] / plot_df['total_tss']\n",
    "plot_df['ada_stop/total_tss'] = plot_df['adapted_near_stop'] / plot_df['total_tss']\n",
    "\n",
    "plot_df['lib_colors'] = plot_df.lib.map(lib_colors)\n",
    "\n",
    "figures = []\n",
    "# for xcol in xcols:\n",
    "#     figures.append(px.scatter(plot_df,\n",
    "#                               x=xcol,\n",
    "#                               y=ycol,\n",
    "#                               color='lib',\n",
    "#                               symbol='nmd_target_type',\n",
    "#                               color_discrete_map=lib_colors,\n",
    "#                               symbol_sequence=['circle', 'cross'],\n",
    "#                               log_x=True,\n",
    "#                               log_y=True,\n",
    "#                               title=f\"{xcol} vs {ycol}\",\n",
    "#                               hover_name='gene_name',\n",
    "#                               hover_data=[\n",
    "#                                   'lib',\n",
    "#                                   'nmd_target_type',\n",
    "#                                   'total_unadapted',\n",
    "#                                   'total_adapted',\n",
    "#                                   'total_tss',\n",
    "#                                   'total_stop',\n",
    "#                                           ],\n",
    "#                               \n",
    "#                               )\n",
    "#                      )\n",
    "for i, xcol in enumerate(xcols):\n",
    "    for gene_name in plot_df['gene_name'].unique():\n",
    "        gene_df = plot_df[plot_df['gene_name'] == gene_name]\n",
    "        figures.append(go.Scatter(\n",
    "            x=gene_df[xcol],\n",
    "            y=gene_df[ycol],\n",
    "            mode='markers',\n",
    "            marker=dict(\n",
    "                color=gene_df['lib_colors'],\n",
    "                symbol=gene_df['nmd_target_type'].map({'NMD_Targets': 'circle', 'NMD_NonTargets': 'cross'}),\n",
    "                size=10\n",
    "            ),\n",
    "            name=gene_name,\n",
    "            legendgroup=gene_name,\n",
    "            showlegend=True if i == 0 else False,\n",
    "        ))\n",
    "fig = make_subplots(rows=1, cols=(len(xcols)))\n",
    "for i, figure in enumerate(figures):\n",
    "    fig.add_trace(figure, row=1, col=(i % len(xcols))+1)\n",
    "\n",
    "# We need to add axis labels:\n",
    "fig.update_xaxes(title_text=\"Decapped\", row=1, col=1)\n",
    "fig.update_xaxes(title_text=\"Cleaved\", row=1, col=2)\n",
    "fig.update_yaxes(title_text=\"Total Near TSS\", row=1, col=1)\n",
    "fig.update_yaxes(title_text=\"Total Near TSS\", row=1, col=2)\n",
    "\n",
    "\n",
    "fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "660fd53f7d65770c",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plotting_summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f56585d354d70ce7",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Work on 7/8/2024\n",
    "\n",
    "Trying to get a more simple summary metric working. Going to be more aggressive with throwing out low depth libs.\n",
    "\n",
    "The general idea is that the increase in reads near the TSS when we break NMD with smg-5 or smg-6 is due to the vast increase of the precursor NMD species. This would be expected if NMD targets were undergoing typical end of life cycle deadenylation and decapping.\n",
    "\n",
    "If the decapping was somehow an artifact of the NMD knockouts, we'd expect the ratio between reads and TSS reads to differ between NMD targets and non-NMD targets.\n",
    "\n",
    "An issue here, is that our ability to capture adapted near-TSS species is entirely reliant the knockdown of xrn-1 and the adapter ligation reaction efficacy.\n",
    "\n",
    "One idea to avoid these sources of noise could be only looking within a library, comparing the ratio of decapped + adapted species on the NMD targets and the non-targets... ? Let's try:"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "efe1427e4c8e1409"
  },
  {
   "cell_type": "code",
   "source": [
    "plot_libs = libs_to_run\n",
    "\n",
    "window_size_for_search = 100\n",
    "\n",
    "target_nmd_types = ('nmd_target', 'ambiguous', 'non_nmd_target')\n",
    "subset_nmd_types_list = [target_nmd_types[0], target_nmd_types[1], target_nmd_types[2], target_nmd_types[:2], target_nmd_types[:]]\n",
    "subset_nmd_types_names = ['NMD_Targets', 'Ambiguous_Reads', 'NMD_NonTargets', 'NMD_TargetsAndAmbiguous', 'All_Types']\n",
    "\n",
    "plot_lib_objs_dict = {lib: obj_dict[lib] for lib in plot_libs}\n",
    "overall_summaries_nested_dict = {}\n",
    "for gene in nmd_genes[::-1]:\n",
    "    # if gene != \"rpl-7A\":\n",
    "    #     continue\n",
    "    print(f\"{gene:#^50}\")\n",
    "    gene_summary_dicts = {}\n",
    "    \n",
    "    for subset_nmd_types, subset_nmd_types_name in zip(subset_nmd_types_list, subset_nmd_types_names):\n",
    "        try:\n",
    "            lib_df_dict_for_gene = {lib: lib_obj.nmd_targets_df.query('gene_name == @gene & nmd_assignment in @subset_nmd_types')\n",
    "                                    for lib, lib_obj in plot_lib_objs_dict.items()}\n",
    "            # summary_dict = plot_gene_data3(gtf_df, gene, lib_df_dict_for_gene, print_stats=False,\n",
    "            #                                title_override=f\"{gene}\\nNMD Target Types: {subset_nmd_types}\",\n",
    "            #                                save_name_override=f\"{gene}_{subset_nmd_types_name}\",\n",
    "            #                                current_dir=current_dir,)\n",
    "            summary_dict = just_extract_gene_data(gtf_df, gene, lib_df_dict_for_gene, print_stats=False, window_size=window_size_for_search)\n",
    "            gene_summary_dicts[subset_nmd_types_name] = summary_dict\n",
    "        except Exception as e:\n",
    "            print(f\"Error: {e}\")\n",
    "            print(f\"Gene: {gene}\")\n",
    "            print(f\"NMD Types: {subset_nmd_types}\")\n",
    "            print(f\"Continuing...\")\n",
    "    overall_summaries_nested_dict[gene] = gene_summary_dicts\n",
    "\n",
    "plotting_list_for_df = []\n",
    "\n",
    "dict_index = 0\n",
    "for gene, gene_dict in overall_summaries_nested_dict.items():\n",
    "    for nmd_target_type, target_type_dict in gene_dict.items():\n",
    "        print(f\"{gene} ({nmd_target_type}):\")\n",
    "        for lib, lib_dict in target_type_dict.items():\n",
    "            print(f\"\\t{lib}:\")\n",
    "            print(f\"\\t\\tAdapted Near Stop:   {lib_dict['Adapted Near Stop']}\")\n",
    "            print(f\"\\t\\tAdapted Near TSS:    {lib_dict['Adapted Near TSS']}\")\n",
    "            print(f\"\\t\\tTotal Adapted:       {lib_dict['Total Adapted']}\")\n",
    "            print(f\"\\t\\tUnadapted Near Stop: {lib_dict['Unadapted Near Stop']}\")\n",
    "            print(f\"\\t\\tUnadapted Near TSS:  {lib_dict['Unadapted Near TSS']}\")\n",
    "            print(f\"\\t\\tTotal Unadapted:     {lib_dict['Total Unadapted']}\")\n",
    "            try:\n",
    "                print(f\"\\t\\tAda Stop/Unada FL: {lib_dict['Adapted Near Stop']/lib_dict['Unadapted Near TSS']:.2%}\")\n",
    "            except ZeroDivisionError:\n",
    "                print(f\"\\t\\tAda Stop/Unada FL: -- NO UNADAPTED FL --\")\n",
    "            plotting_list_for_df.append(\n",
    "                {\n",
    "                    'gene_name': gene,\n",
    "                    'nmd_target_type': nmd_target_type,\n",
    "                    'lib': lib,\n",
    "                    'adapted_near_stop': lib_dict['Adapted Near Stop'],\n",
    "                    'adapted_near_tss': lib_dict['Adapted Near TSS'],\n",
    "                    'total_adapted': lib_dict['Total Adapted'],\n",
    "                    'unadapted_near_stop': lib_dict['Unadapted Near Stop'],\n",
    "                    'unadapted_near_tss': lib_dict['Unadapted Near TSS'],\n",
    "                    'total_unadapted': lib_dict['Total Unadapted'],\n",
    "                }\n",
    "            )\n",
    "\n",
    "plotting_summary_df = pd.DataFrame.from_records(plotting_list_for_df)\n",
    "plotting_summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4148ece0d9447f27",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "summary_df = plotting_summary_df.copy()\n",
    "\n",
    "lib_strain_dict = {'N2': 'wildtype',\n",
    "                   'S5': 'smg-5',\n",
    "                   'S6': 'smg-6',\n",
    "                   'S7': 'smg-7'}\n",
    "\n",
    "summary_df['strain'] = summary_df.lib.str[-2:].map(lib_strain_dict)\n",
    "\n",
    "summary_df['total_near_tss'] = summary_df['adapted_near_tss'] + summary_df['unadapted_near_tss']\n",
    "summary_df['total_near_stop'] = summary_df['adapted_near_stop'] + summary_df['unadapted_near_stop']\n",
    "summary_df['total_reads'] = summary_df['total_adapted'] + summary_df['total_unadapted']\n",
    "\n",
    "summary_df['frac_near_tss'] = summary_df['total_near_tss'] / summary_df['total_reads']\n",
    "summary_df['frac_near_stop'] = summary_df['total_near_stop'] / summary_df['total_reads']\n",
    "\n",
    "summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2682c07a11d0ccd",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "summary_gb  = summary_df.groupby(['gene_name', 'nmd_target_type', 'strain'])\n",
    "df = summary_gb['frac_near_tss'].agg(['mean', 'std', 'sem'])\n",
    "\n",
    "strain_dict = {}\n",
    "for strain_val in summary_df['strain'].unique():\n",
    "    strain_dict[strain_val] = df.loc[:, :, strain_val].rename(columns={'mean': f'{strain_val}_mean', 'std': f'{strain_val}_std', 'sem': f'{strain_val}_sem'})\n",
    "remerge_df = pd.concat(strain_dict.values(), axis=1)\n",
    "\n",
    "for strain_val in summary_df['strain'].unique():\n",
    "    if strain_val == 'wildtype':\n",
    "        continue\n",
    "    remerge_df[f'{strain_val}/wildtype'] = remerge_df[f'{strain_val}_mean'] / remerge_df['wildtype_mean']\n",
    "remerge_df[[col for col in remerge_df.columns if 'mean' in col]]  # This is very close, but now I am visually comparing each mean in NMD targets to the non-NMD targets. I want to do this via and actual comparison (maybe plot fractions or do a paired t-test between"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "a3f28679d669dbf6",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plot_df = remerge_df[[col for col in remerge_df.columns if 'mean' in col and 'smg-7' not in col]].loc[:, ['NMD_Targets', 'NMD_NonTargets'], :].reset_index()\n",
    "\n",
    "fig = px.bar(plot_df,\n",
    "             x='gene_name',\n",
    "             y=[col for col in plot_df.columns if 'mean' in col and 'smg-7' not in col],\n",
    "             color='nmd_target_type',\n",
    "             barmode='group',\n",
    "             )\n",
    "fig.show(renderer='firefox')\n",
    "plot_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ae0959fb4b186aed",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "wider_plot_dict = {}\n",
    "for target_type in plot_df['nmd_target_type'].unique():\n",
    "    target_type_df = plot_df.query(\"nmd_target_type == @target_type\").drop(columns='nmd_target_type')\n",
    "    wider_plot_dict[target_type] = target_type_df.rename(columns={col: f\"{target_type}_{col}\" for col in target_type_df.columns if 'mean' in col}).set_index('gene_name')\n",
    "wider_plot_df = pd.concat(wider_plot_dict.values(), axis=1)\n",
    "wider_plot_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "a47b95f782f33c76",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "for lib_mean in ['wildtype_mean', 'smg-5_mean', 'smg-6_mean']:\n",
    "    wider_plot_df[f\"{lib_mean}_diff\"] = wider_plot_df[f'NMD_Targets_{lib_mean}'] - wider_plot_df[f'NMD_NonTargets_{lib_mean}']\n",
    "    wider_plot_df[f\"{lib_mean}_diff_frac\"] = wider_plot_df[f'NMD_Targets_{lib_mean}'] / wider_plot_df[f'NMD_NonTargets_{lib_mean}']\n",
    "wider_plot_df[[col for col in wider_plot_df.columns if 'diff' in col]]\n",
    "\n",
    "# THE PROBLEM: I merged on lib types (N2, S5, S6) before I made the actual comparisons between NMD targets and non-NMD targets. I need to do this first, then merge. This will allow me to target the issue I noted above regarding variable depth, ligation, and xrn-1 knockdown!!"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "8d11fb821c60569d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# THE PROBLEM:\n",
    "I merged on lib types (N2, S5, S6) before I made the actual comparisons between NMD targets and non-NMD targets. I need to do this first, then merge. This will allow me to target the issue I noted above regarding variable depth, ligation, and xrn-1 knockdown!!\n",
    "\n",
    "1. We want to compare the TSS ratio for NMD vs non_NMD targets in each library\n",
    "2. We can take a mean of those ratios for each strain (ie. wt or smg-5)\n",
    "3. Finally we can compare the mean ratios between strains, WT should look similar to smg-5 and smg-6 (if nothing is changing on the decapping front)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "9d3d401b08e11738"
  },
  {
   "cell_type": "code",
   "source": [
    "# Before anything let's copy down code so we can regenerate the input df\n",
    "summary_df = plotting_summary_df.copy()\n",
    "\n",
    "lib_strain_dict = {'N2': 'wildtype',\n",
    "                   'S5': 'smg-5',\n",
    "                   'S6': 'smg-6',\n",
    "                   'S7': 'smg-7'}\n",
    "\n",
    "summary_df['strain'] = summary_df.lib.str[-2:].map(lib_strain_dict)\n",
    "\n",
    "summary_df['total_near_tss'] = summary_df['adapted_near_tss'] + summary_df['unadapted_near_tss']\n",
    "summary_df['total_near_stop'] = summary_df['adapted_near_stop'] + summary_df['unadapted_near_stop']\n",
    "summary_df['total_reads'] = summary_df['total_adapted'] + summary_df['total_unadapted']\n",
    "\n",
    "# There are a few different ways to do this!\n",
    "summary_df['frac_near_tss'] = summary_df['total_near_tss'] / summary_df['total_reads']\n",
    "\n",
    "summary_df['frac_near_tss'] = summary_df['adapted_near_tss'] / summary_df['total_reads']\n",
    "\n",
    "summary_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "61be3fe44c2b95d3",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# First we want to compare the TSS ratio for NMD vs non_NMD targets in each library\n",
    "in_df = summary_df.copy()\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.set_index(['gene_name', 'nmd_target_type', 'strain', 'rep','lib'])[['frac_near_tss', 'total_reads']].sort_index()\n",
    "inter_wide_df = in_df.loc[:, ['NMD_Targets', 'NMD_NonTargets'], :].unstack('nmd_target_type')\n",
    "inter_wide_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4ae520f60e9e6713",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Second we can take a mean of those ratios for each strain (ie. wt or smg-5)\n",
    "# First lets drop anything that doesn't have at least X reads for each:\n",
    "min_reads = 50\n",
    "semi_wide_df = inter_wide_df.copy()\n",
    "semi_wide_df = semi_wide_df[semi_wide_df['total_reads'] >= min_reads].dropna(axis=0)\n",
    "# So we want to divide the NMD_Targets by the NMD_NonTargets for each row of semi_wide_df\n",
    "semi_wide_df['near_tss(nmd / non_nmd)'] = semi_wide_df['frac_near_tss']['NMD_Targets'] / semi_wide_df['frac_near_tss']['NMD_NonTargets']\n",
    "semi_wide_df['total_total_reads'] = semi_wide_df['total_reads']['NMD_Targets'] + semi_wide_df['total_reads']['NMD_NonTargets']"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "171d5b3530aaf019",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "comp_box_df = semi_wide_df.copy().reset_index()\n",
    "comp_box_df = comp_box_df[comp_box_df['near_tss(nmd / non_nmd)'] > 0]\n",
    "comp_box_df['log2_near_tss(nmd / non_nmd)'] = np.log2(comp_box_df['near_tss(nmd / non_nmd)'])\n",
    "comp_box_df.columns = [\"_\".join(a).strip('_') for a in comp_box_df.columns.to_flat_index()]\n",
    "fig = px.box(comp_box_df,\n",
    "             x='strain',\n",
    "             y='near_tss(nmd / non_nmd)',\n",
    "             log_y=True,\n",
    "             # color='gene_name',\n",
    "             points='all',\n",
    "             hover_name='gene_name',\n",
    "             hover_data=['rep',\n",
    "                         'strain',\n",
    "                         'lib',\n",
    "                         'near_tss(nmd / non_nmd)',\n",
    "                         'frac_near_tss_NMD_Targets',\n",
    "                         'frac_near_tss_NMD_NonTargets',\n",
    "                         ],\n",
    "                )\n",
    "fig.show()\n",
    "comp_box_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "d5cb2689992895a2",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Finally we can compare the mean ratios between strains, WT should look similar to smg-5 and smg-6\n",
    "final_df = semi_wide_df.copy()\n",
    "final_df.columns = [\"_\".join(a).strip('_') for a in final_df.columns.to_flat_index()]\n",
    "final_df = final_df[['frac_near_tss_NMD_Targets', 'frac_near_tss_NMD_NonTargets', 'NMD_Targets_frac', 'total_total_reads']]\n",
    "final_df.sort_index()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "8bae4134198b5f0c",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plot_df = final_df.reset_index()\n",
    "window_size_for_search = 100\n",
    "limits = [-0.1, 1.1] # [-0.0025, 0.03] # [-0.1, 1.1]\n",
    "read_type = \"Any Read\" # \"Adapted Reads\" # Any Read\n",
    "\n",
    "plot_df['name'] = plot_df['gene_name'] + ' ' + plot_df['strain'] + ' rep' + plot_df['rep'].astype(str)\n",
    "fig = px.scatter(plot_df,\n",
    "                 x='frac_near_tss_NMD_Targets',\n",
    "                 y='frac_near_tss_NMD_NonTargets',\n",
    "                 color='gene_name',\n",
    "                 hover_name='name',\n",
    "                 size='total_total_reads',\n",
    "                 )\n",
    "fig.update_layout(shapes = [{'type': 'line', 'line': {'dash': 'dot'}, 'yref': 'y', 'xref': 'x', 'y0': -1, 'y1': 2, 'x0': -1, 'x1': 2}],\n",
    "                  title=f'Fraction of Total Reads Near TSS<br>(Counting {read_type} within {window_size_for_search}nt)',\n",
    "                  )\n",
    "fig.update_xaxes(constrain='domain', range=limits)  \n",
    "fig.update_yaxes(scaleanchor= 'x', scaleratio= 1, range=limits)\n",
    "fig.write_html(current_dir / 'plots' / 'decappingAndCleavage' / f\"FractionOfTotalReadsNearTSS_{read_type.replace(' ', '')}.html\")\n",
    "fig.write_image(current_dir / 'plots' / 'decappingAndCleavage' / f\"FractionOfTotalReadsNearTSS_{read_type.replace(' ', '')}.png\")\n",
    "fig.show()\n",
    "plot_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6da8bb2a10b73cd8",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "final_df.unstack(\"strain\").mean()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "d9ed4d83cd40e175",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "final_df.unstack(\"strain\").describe()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "713ec96ce27064e4",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "fig = px.box(final_df.reset_index(),\n",
    "             x='strain',\n",
    "             y='NMD_Targets_frac',\n",
    "             # color='gene_name',\n",
    "             points='all',\n",
    "             )\n",
    "fig.show(renderer='firefox')\n",
    "final_df.reset_index()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "9b9b0ea86760333f",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "semi_wide_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5a6a2b661b2617d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "min_reads = 50\n",
    "\n",
    "box_df = in_df.loc[:, :, :, :, ['NMD_Targets', 'NMD_NonTargets'], :].reset_index()\n",
    "box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'] + ' rep' + box_df['rep'].astype(str)\n",
    "box_df = box_df[box_df['total_reads'] >= min_reads].dropna(axis=0)\n",
    "box_df = box_df[box_df['frac_near_tss'] > 0]\n",
    "\n",
    "fig = px.box(box_df,\n",
    "             x='strain',\n",
    "             y='frac_near_tss',\n",
    "             color='nmd_target_type',\n",
    "             points='all',\n",
    "             hover_name='name',\n",
    "             log_y=True,\n",
    "             )\n",
    "fig.show(renderer='firefox')\n",
    "box_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "1a665307dd1b1b54",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import scipy.stats as stats\n",
    "from colorama import Fore, Style\n",
    "\n",
    "def run_stats_intralib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       strains=('wildtype', 'smg-5', 'smg-6'),\n",
    "                       output_file: Path = None):\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for strain in strains:\n",
    "        nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_Targets'\")\n",
    "        non_nmd_df = df_in.query(\"strain == @strain & nmd_target_type == 'NMD_NonTargets'\")\n",
    "        merge_df = nmd_df.merge(non_nmd_df,\n",
    "                                on=('gene_name', 'strain', 'rep', 'lib', 'name'),\n",
    "                                suffixes=('_nmd', '_non_nmd'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_nmd'], merge_df[f'{compare_col}_non_nmd'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[strain] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    output_string += f\"{'Strain':^8} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-' * 50 + '\\n'\n",
    "    for strain, test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{strain:>8} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string\n",
    "\n",
    "def run_stats_interlib(df_in, compare_col, sig_cutoff=0.05, steps=[100, 10, 1],\n",
    "                       comparisons=(('wildtype', 'smg-5'), ('wildtype', 'smg-6'), ('smg-5', 'smg-6')),\n",
    "                       nmd_type='NMD_NonTargets',\n",
    "                       output_file=None):\n",
    "    # Now let's do a similar comparison but between non-NMD-targets between libs\n",
    "    test_results = {}\n",
    "    output_string = \"\"\n",
    "    for comp in comparisons:\n",
    "        nmd_df = df_in.query(\"strain == @comp[0] & nmd_target_type == @nmd_type\")\n",
    "        non_nmd_df = df_in.query(\"strain == @comp[1] & nmd_target_type == @nmd_type\")\n",
    "        drop_cols = ['nmd_target_type', 'strain', 'lib', 'name']\n",
    "        merge_df = nmd_df.drop(columns=drop_cols).merge(non_nmd_df.drop(columns=drop_cols), on=('gene_name', 'rep'), suffixes=('_lib1', '_lib2'))\n",
    "        paired_test = stats.ttest_rel(merge_df[f'{compare_col}_lib1'], merge_df[f'{compare_col}_lib2'])\n",
    "        unpaired_test = stats.ttest_ind(nmd_df[compare_col], non_nmd_df[compare_col])\n",
    "        test_results[comp] = {'paired': paired_test, 'unpaired': unpaired_test}\n",
    "    \n",
    "    output_string += f\"{'Comp':^14} | {'Paired p-val (df)':^15} | {'Unpaired p-val (df)':^15}\\n\"\n",
    "    output_string += '-'*56 + '\\n'\n",
    "    \n",
    "    for (lib1, lib2), test_dict in test_results.items():\n",
    "        paired_str = f\" {test_dict['paired'].pvalue:^1.3e}   ({test_dict['paired'].df:>2})\"\n",
    "        unpaired_str = f\"   {test_dict['unpaired'].pvalue:^1.3e}   ({int(test_dict['unpaired'].df):>2})\"\n",
    "        if lib1 == 'wildtype':\n",
    "            lib1 = 'wt'\n",
    "        if test_dict['paired'].pvalue < sig_cutoff / steps[0]:\n",
    "            paired_str = Fore.MAGENTA + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[1]:\n",
    "            paired_str = Fore.RED + paired_str + Style.RESET_ALL\n",
    "        elif test_dict['paired'].pvalue < sig_cutoff / steps[2]:\n",
    "            paired_str = Fore.YELLOW + paired_str + Style.RESET_ALL\n",
    "        if test_dict['unpaired'].pvalue < sig_cutoff / steps[0]:\n",
    "            unpaired_str = Fore.MAGENTA + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[1]:\n",
    "            unpaired_str = Fore.RED + unpaired_str + Style.RESET_ALL\n",
    "        elif test_dict['unpaired'].pvalue < sig_cutoff / steps[2]:\n",
    "            unpaired_str = Fore.YELLOW + unpaired_str + Style.RESET_ALL\n",
    "        output_string += f\"{lib1:>5} vs {lib2:<5} | {paired_str} | {unpaired_str}\\n\"\n",
    "    key_string = (f\"Key: \"\n",
    "                  f\"{Fore.MAGENTA}p < {sig_cutoff / steps[0]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.RED}p < {sig_cutoff / steps[1]}{Style.RESET_ALL} \"\n",
    "                  f\"{Fore.YELLOW}p < {sig_cutoff / steps[2]}{Style.RESET_ALL}\\n\")\n",
    "    print(output_string + key_string)\n",
    "    ansi_escape = re.compile(r'\\x1B(?:[@-Z\\\\-_]|\\[[0-?]*[ -/]*[@-~])')\n",
    "    output_string = ansi_escape.sub('', output_string)\n",
    "    if output_file is not None:\n",
    "        with open(output_file, 'a') as f:\n",
    "            f.write(output_string)\n",
    "    return output_string\n",
    "\n",
    "run_stats_intralib(box_df, 'frac_near_tss')\n",
    "print(\"\\n\\n\")\n",
    "run_stats_interlib(box_df, 'frac_near_tss')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bd35127c1586d307",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Maybe the NMD targets have higher adapted_TSS fractions in the WT not because they have high numbers of adapted_TSS reads, but b/c they have lower numbers of the denominator (all reads)\n",
    "# To remedy this, let's try comparing the adapted_TSS to the unadapted_TSS\n",
    "\n",
    "# Additionally, it would be nice to do an adapted RPM comparison against Adapt TSS reads. But this needs more reads!!\n",
    "\n",
    "# I need to circle WAY back to retain both adapt and unadapted TSSes.. UGH LETS DO IT\n",
    "in_df = summary_df.copy()\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type']).sort_index()\n",
    "better_df = in_df.loc[:, :, :, :, ['NMD_Targets', 'NMD_NonTargets'], :].drop(columns=['frac_near_tss'])\n",
    "better_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "21fbb9b7a3b7335e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "min_reads = 1\n",
    "\n",
    "# target_col = 'adapted_near_tss / unadapted_near_tss'\n",
    "# target_col = 'adapted_near_tss / total_reads'\n",
    "\n",
    "target_numerator = 'adapted_near_tss'\n",
    "\n",
    "# target_denominator = 'total_reads'\n",
    "# target_denominator = 'unadapted_near_tss'\n",
    "# target_denominator = 'total_near_tss'\n",
    "target_denominator = 'total_adapted'\n",
    "\n",
    "target_col = f\"{target_numerator} / {target_denominator}\"\n",
    "\n",
    "\n",
    "better_df['adapted_near_tss'] = better_df['adapted_near_tss'].fillna(0)\n",
    "better_df['unadapted_near_tss'] = better_df['unadapted_near_tss'].fillna(0)\n",
    "better_df[target_col] = better_df[target_numerator] / better_df[target_denominator]\n",
    "\n",
    "box_df = better_df.copy().reset_index()\n",
    "box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'] + ' rep' + box_df['rep'].astype(str)\n",
    "box_df = box_df[box_df['total_reads'] >= min_reads].dropna(axis=0)\n",
    "box_df = box_df[box_df[target_col] > 0]\n",
    "\n",
    "fig = px.box(box_df,\n",
    "             x='strain',\n",
    "             y=target_col,\n",
    "             color='nmd_target_type',\n",
    "             points='all',\n",
    "             hover_name='name',\n",
    "             log_y=True,\n",
    "             )\n",
    "save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "fig.write_html(current_dir / 'plots' / 'decapping' / f\"boxPlot_{save_friendly_target_col}.html\")\n",
    "fig.write_image(current_dir / 'plots' / 'decapping' / f\"boxPlot_{save_friendly_target_col}.png\")\n",
    "fig.show(renderer='firefox')\n",
    "\n",
    "output_file = current_dir / 'plots' / 'decapping' / f'boxPlotStats_{save_friendly_target_col}.txt'\n",
    "output_file.touch(exist_ok=True)\n",
    "with open(output_file, 'w') as f:\n",
    "    f.write(f\"Stats for {target_col}\\n\")\n",
    "    f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "print(\"Targets\")\n",
    "run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "with open(output_file, 'a') as f:\n",
    "    f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "print(\"NonTargets\")\n",
    "run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "\n",
    "with open(output_file, 'a') as f:\n",
    "    f.write(\"\\nInter Library Stats:\\n\")\n",
    "run_stats_intralib(box_df, target_col, output_file=output_file)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3e323169ecbd428e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def run_box_and_ttests(input_df: pd.DataFrame,\n",
    "                       target_num: str, target_denom: str,\n",
    "                       min_reads: int = 1,\n",
    "                       output_dir: Path = current_dir / 'plots' / 'decapping'):\n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    input_df\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'] + ' rep' + box_df['rep'].astype(str)\n",
    "    box_df = box_df[box_df['total_reads'] >= min_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=True,\n",
    "                 )\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    fig.write_html(output_dir / f\"boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"boxPlot_{save_friendly_target_col}.png\")\n",
    "    fig.show(renderer='firefox')\n",
    "    \n",
    "    output_file = output_dir / f'boxPlotStats_{save_friendly_target_col}.txt'\n",
    "    output_file.touch(exist_ok=True)\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(f\"Stats for {target_col}\\n\")\n",
    "        f.write(\"\\nIntra Library Stats w/ NMD Targets:\\n\")\n",
    "    print(\"Targets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_Targets', output_file=output_file)\n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nIntra Library Stats w/ Non Targets:\\n\")\n",
    "    print(\"NonTargets\")\n",
    "    run_stats_interlib(box_df, target_col, nmd_type='NMD_NonTargets', output_file=output_file)\n",
    "    \n",
    "    with open(output_file, 'a') as f:\n",
    "        f.write(\"\\nInter Library Stats:\\n\")\n",
    "    run_stats_intralib(box_df, target_col, output_file=output_file)\n",
    "\n",
    "for target_denominator in ['total_reads', 'unadapted_near_tss', 'total_adapted']:\n",
    "    print(f\"\\n\\nUsing {target_denominator} for denominator...\\n\")\n",
    "    run_box_and_ttests(better_df, 'adapted_near_tss', target_denominator)\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "77c569b70c986e1b",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# Let's see if we can make quick bar plots for individual gene species"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "d728e97bb590e48c",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
