{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# decappingQuantification4_finalVersions.ipynb\n",
    "## Marcus Viscardi,    August 19, 2024\\\n",
    "\n",
    "I need to finalize some of these plots and get them ready for publication.\n",
    "\n",
    "Generally, I want to take the pre-processed data from previous notebooks and create the final versions. I'll try to be good about using markdown blocks to label what I am doing in each section!"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2ae6280f65c025d8"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "from typing import Tuple\n",
    "\n",
    "import nanoporePipelineCommon as npCommon\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "from icecream import ic\n",
    "from datetime import datetime\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from scipy.stats import mannwhitneyu, ks_2samp\n",
    "\n",
    "import pickle as pkl\n",
    "\n",
    "import random\n",
    "\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "def __time_formatter__():\n",
    "    now = datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "    return f\"ic: {now} | > \"\n",
    "ic.configureOutput(prefix=__time_formatter__)\n",
    "\n",
    "_ = ic(\"Imports done!\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Load some of the pre-processed data\n",
    "\n",
    "### This pre-processed dataframe comes from `decappingQuantification2_metaPlots.ipynb` and is the data that has been cleaned up and had the near TSS data added to it."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "46903286f522c3e7"
  },
  {
   "cell_type": "code",
   "source": [
    "gtf_parquet_path = Path(\"/data16/marcus/genomes/plus_cerENO2_elegansRelease100/230327_allChrs_plus-cerENO2.gtf.parquet\")\n",
    "\n",
    "gtf_df = pd.read_parquet(gtf_parquet_path)\n",
    "\n",
    "current_dir = Path.cwd()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "945a88812b1e713f",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "preprocessed_df = pd.read_parquet(current_dir / \"preprocessed_data\" / \"decapping_data.parquet\")\n",
    "preprocessed_df.head(10)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "d9a3d12297364861",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Trying to make a better script/box plot:\n",
    "\n",
    "I want to connect genes with the same name across species and NMD_types. Going to lean on the following post: [Stack Overflow](https://stackoverflow.com/q/60164606/13316742)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "9ee48493f2aa20a4"
  },
  {
   "cell_type": "code",
   "source": [
    "df = preprocessed_df.copy()\n",
    "df.set_index(['gene_name', 'nmd_target_type', 'lib', 'strain'], inplace=True)\n",
    "# print(df.index.get_level_values('nmd_target_type').unique())\n",
    "in_df = df[['adapted_near_tss', 'total_adapted', 'total_reads', 'unadapted_near_tss', 'total_near_tss']].fillna(0).reset_index()\n",
    "\n",
    "set_dict = {'old': 1, 'newer': 2, 'third': 3, 'temp25c': 4}\n",
    "\n",
    "in_df['rep'] = in_df.lib.str[:-2].map(set_dict)\n",
    "in_df = in_df.query(\"strain != 'smg-7'\")\n",
    "in_df = in_df.set_index(['gene_name', 'strain', 'rep','lib', 'nmd_target_type']).sort_index()\n",
    "better_df = in_df.loc[:, :, :, :, ['NMD_Targets', 'NMD_NonTargets', 'Ambiguous_Reads'], :]\n",
    "better_df.head()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "90cbd3fabcf78b32",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "This is a working version of the boxplots from `decappingQuantification2_metaPlots.ipynb` I can use it as a starting point if I get confused..."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2f8ca78f232e5e45"
  },
  {
   "cell_type": "code",
   "source": [
    "def process_df_for_means(input_df: pd.DataFrame,\n",
    "                         target_num: str, target_denom: str,\n",
    "                         min_total_reads: int = 1,\n",
    "                         min_adapted_reads: int = 1,\n",
    "                         min_tss_ada_reads: int = 0,\n",
    "                         add_tss_ada_pseudo: float = -1.0,\n",
    "                         min_number_reps_observed: int = -1,\n",
    "                         log2_y: bool = False,\n",
    "                         log_y: bool = True,\n",
    "                         print_failed_genes: bool = False,\n",
    "                         ) -> pd.DataFrame:\n",
    "    input_df = input_df.copy()\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    input_df['total_near_tss'] = input_df['adapted_near_tss'] + input_df['unadapted_near_tss']\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str)\n",
    "    \n",
    "    box_df = box_df[box_df['total_reads'] >= min_total_reads].dropna(axis=0)\n",
    "    box_df = box_df[box_df['total_adapted'] >= min_adapted_reads].dropna(axis=0)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    index_fixed_box_df = box_df.reset_index().set_index(['strain', 'gene_name', 'nmd_target_type'])\n",
    "    groupby_obj = index_fixed_box_df.groupby(['strain', 'gene_name', 'nmd_target_type'], observed=True)\n",
    "    og_numerical_cols = index_fixed_box_df.select_dtypes(include=[np.number]).columns.to_list()\n",
    "    keep_cols = [target_col, target_num, target_denom, 'total_near_tss', 'total_reads', 'adapted_near_tss', 'unadapted_near_tss'] + og_numerical_cols\n",
    "    keep_cols = list(set(keep_cols))\n",
    "    groupby_df = groupby_obj[keep_cols].mean()\n",
    "    std_cols = [f'{col}_std' for col in keep_cols]\n",
    "    groupby_df[std_cols] = groupby_obj[keep_cols].std()\n",
    "    groupby_df['successful_reps'] = groupby_obj['rep'].apply(lambda x: ','.join(map(str, x)))\n",
    "    groupby_df['num_successful_reps'] = groupby_obj['rep'].count()\n",
    "    groupby_df.reset_index(inplace=True)\n",
    "    if min_number_reps_observed > 0:\n",
    "        if print_failed_genes:\n",
    "            print(f\"Failed Genes at `min_number_reps_observed >= {min_number_reps_observed}` level:\")\n",
    "            print(groupby_df.query(\"num_successful_reps < @min_number_reps_observed\")[['gene_name', 'strain', 'gene_name', 'nmd_target_type', 'num_successful_reps']])\n",
    "        groupby_df = groupby_df.query(\"num_successful_reps >= @min_number_reps_observed\")\n",
    "    return groupby_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "73b27ca56a716767",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now let's try to manually make a similar version to above, from there we can start building up the attached version!!"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "69b501e3a2ed0f7f"
  },
  {
   "cell_type": "code",
   "source": [
    "# random.seed(42)\n",
    "# \n",
    "# plot_num = 'adapted_near_tss'\n",
    "# plot_denum = 'total_adapted'\n",
    "# plot_col = f'{plot_num} / {plot_denum}'\n",
    "# \n",
    "# min_adapted = 3\n",
    "# min_total = 10\n",
    "# \n",
    "# plot_df = better_df.copy().query(f\"total_adapted >= {min_adapted} & total_reads >= {min_total}\")\n",
    "# plot_df[plot_col] = plot_df[plot_num] / df[plot_denum]\n",
    "# base_value_dict = {'wildtype': 1,\n",
    "#                    'smg-5': 2,\n",
    "#                    'smg-6': 3,\n",
    "#                    # 'smg-7': 4,\n",
    "#                    }\n",
    "# adjusted_value_dict = {'NMD_Targets': -0.15,\n",
    "#                        'NMD_NonTargets': 0.15}\n",
    "# # These colors are updated from the final plots!\n",
    "# plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "#                ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "#                ('#676767', '#414141'),  # ambiguous greys\n",
    "#                )\n",
    "# colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "#                'NMD_NonTargets': plot_colors[1][0],}\n",
    "# \n",
    "# def adjust_value(row: pd.Series) -> float:\n",
    "#     strain = row.name[1]\n",
    "#     nmd_target_type = row.name[4]\n",
    "#     return base_value_dict[strain] + adjusted_value_dict[nmd_target_type] + (random.random() - 0.5) * 0.1\n",
    "# \n",
    "# plot_df['adjusted_x'] = plot_df.apply(adjust_value, axis=1)\n",
    "# \n",
    "# fig, ax = plt.subplots(figsize=(8, 8))\n",
    "# for strain in plot_df.index.get_level_values('strain').unique():\n",
    "#     added_points_rows = []\n",
    "#     for gene in plot_df.index.get_level_values('gene_name').unique():\n",
    "#         for rep in plot_df.index.get_level_values('rep').unique():\n",
    "#             rows = plot_df.query(f\"gene_name == '{gene}' & strain == '{strain}' & rep == {rep}\")\n",
    "#             nmd_vals = rows.index.get_level_values('nmd_target_type').unique().to_list()\n",
    "#             if len(nmd_vals) == 2:\n",
    "#                 for nmd_val in nmd_vals:\n",
    "#                     ax.plot(rows.query(f\"nmd_target_type == '{nmd_val}'\")['adjusted_x'],\n",
    "#                             rows.query(f\"nmd_target_type == '{nmd_val}'\")[plot_col],\n",
    "#                             marker='o', linestyle='', markersize=5, color=colors_dict[nmd_val])\n",
    "#                 ax.plot(rows['adjusted_x'], rows[plot_col], color='black', alpha=0.5, linewidth=0.5, dashes=[6, 2])\n",
    "#                 added_points_rows.append(rows)\n",
    "#     added_points_df = pd.concat(added_points_rows)\n",
    "#     for nmd_val, color in colors_dict.items():\n",
    "#         subplot_df = added_points_df.query(f\"nmd_target_type == '{nmd_val}'\")\n",
    "#         ax.boxplot(subplot_df[plot_col],\n",
    "#                    vert=True,\n",
    "#                    positions=[2.5*adjusted_value_dict[nmd_val]+base_value_dict[strain]],\n",
    "#                    widths=(adjusted_value_dict[nmd_val]*1.5),\n",
    "#                    meanline=True,\n",
    "#                    patch_artist=True,\n",
    "#                    showmeans=True,\n",
    "#                    showfliers=False,\n",
    "#                    boxprops=dict(facecolor=color),\n",
    "#                    )\n",
    "# ax.set_xticks(list(base_value_dict.values()))\n",
    "# ax.set_xticklabels(list(base_value_dict.keys()))\n",
    "# ax.set_ylabel(f\"{plot_num} / {plot_denum}\")\n",
    "# lines = [plt.Line2D([0], [0], color=colors_dict[nmd_val], lw=2, label=nmd_val) for nmd_val in colors_dict.keys()]\n",
    "# ax.legend(lines, colors_dict.keys())\n",
    "# plt.show()\n",
    "# added_points_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "19e269d35d7a2cb4",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Now let's make that same plot but while using averages for each gene and connections between observations of the same gene"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6e316e13a8436528"
  },
  {
   "cell_type": "code",
   "source": [
    "from matplotlib import colormaps as cm\n",
    "from matplotlib.colors import Normalize\n",
    "from matplotlib.cm import ScalarMappable\n",
    "\n",
    "# def plot_box_and_strip_with_connections(in_df: pd.DataFrame,\n",
    "#                                         plot_num: str,  # 'adapted_near_tss'\n",
    "#                                         plot_denum: str,  # 'total_adapted' or 'total_reads'\n",
    "#                                         min_adapted: int = 3,\n",
    "#                                         min_total: int = 10,\n",
    "#                                         min_tss_ada: int = 0,\n",
    "#                                         add_tss_ada_pseudo: float = 0.0,\n",
    "#                                         min_number_reps_observed: int = -1,\n",
    "#                                         with_connections: bool = True,\n",
    "#                                         log2_y: bool = False,\n",
    "#                                         log_y: bool = False,\n",
    "#                                         with_gene_names=True,\n",
    "#                                         save_dir: Path = current_dir / 'plots' / 'decapping_connectedMeans',\n",
    "#                                         additional_save_text:str = '',\n",
    "#                                         ) -> Tuple[pd.DataFrame, pd.DataFrame, go.Figure]:\n",
    "#     random.seed(42)\n",
    "#     fig, ax = plt.subplots(figsize=(6, 4),)\n",
    "#     spacing_factor = 0.20  # 0.15\n",
    "#     box_pos_scaler = 3.25   # 3.5\n",
    "#     box_width_scaler = 3.0 # 2.5\n",
    "#     \n",
    "#     \n",
    "#     plot_col = f'{plot_num} / {plot_denum}'\n",
    "#     plot_df = in_df.copy()\n",
    "#     \n",
    "#     plot_df = plot_df.query(f\"total_adapted >= {min_adapted} & total_reads >= {min_total}\")\n",
    "#     plot_df[plot_col] = plot_df[plot_num] / plot_df[plot_denum]\n",
    "#     \n",
    "#     plot_df = process_df_for_means(plot_df,\n",
    "#                                    plot_num, plot_denum,\n",
    "#                                    min_total_reads=min_total,\n",
    "#                                    min_adapted_reads=min_adapted,\n",
    "#                                    min_tss_ada_reads=min_tss_ada,\n",
    "#                                    add_tss_ada_pseudo=add_tss_ada_pseudo,\n",
    "#                                    min_number_reps_observed=min_number_reps_observed,\n",
    "#                                    log2_y=log2_y,\n",
    "#                                    log_y=log_y,\n",
    "#                                    )\n",
    "#     \n",
    "#     base_value_dict = {'wildtype': 0,\n",
    "#                        'smg-5': 2,\n",
    "#                        'smg-6': 4,\n",
    "#                        # 'smg-7': 4,\n",
    "#                        }\n",
    "#     adjusted_value_dict = {'NMD_Targets': spacing_factor,\n",
    "#                            'NMD_NonTargets': -spacing_factor}\n",
    "#     # These colors are updated from the final plots!\n",
    "#     plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "#                    ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "#                    ('#676767', '#414141'),  # ambiguous greys\n",
    "#                    )\n",
    "#     colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "#                    'NMD_NonTargets': plot_colors[1][0],}\n",
    "#     \n",
    "#     def adjust_value2(row: pd.Series) -> float:\n",
    "#         strain = row.strain\n",
    "#         nmd_target_type = row.nmd_target_type\n",
    "#         return base_value_dict[strain] + adjusted_value_dict[nmd_target_type] + (random.random() - 0.5) * 0.1\n",
    "#     \n",
    "#     plot_df['adjusted_x'] = plot_df.apply(adjust_value2, axis=1)\n",
    "#     \n",
    "#     all_points_rows = []\n",
    "#     gene_counter_dict = {strain: dict(NMD_Targets=0, NMD_NonTargets=0) for strain in base_value_dict.keys()}\n",
    "#     for strain in plot_df.strain.unique():\n",
    "#         added_points_rows = []\n",
    "#         for gene in plot_df.gene_name.unique():\n",
    "#             rows = plot_df.query(f\"gene_name == '{gene}' & strain == '{strain}'\")\n",
    "#             nmd_vals = rows.nmd_target_type.unique()\n",
    "#             if len(nmd_vals) == 2 and with_connections:\n",
    "#                 ax.plot(rows['adjusted_x'], rows[plot_col], color='black', alpha=0.5, linewidth=0.5, dashes=[6, 2])\n",
    "#             for nmd_val in nmd_vals:\n",
    "#                 ax.plot(rows.query(f\"nmd_target_type == '{nmd_val}'\")['adjusted_x'],\n",
    "#                         rows.query(f\"nmd_target_type == '{nmd_val}'\")[plot_col],\n",
    "#                         marker='o', linestyle='', markersize=5, color=colors_dict[nmd_val],\n",
    "#                         markeredgecolor='black', markeredgewidth=0.5,)\n",
    "#                 if with_gene_names:\n",
    "#                     ax.text(rows.query(f\"nmd_target_type == '{nmd_val}'\")['adjusted_x'].mean(),\n",
    "#                             rows.query(f\"nmd_target_type == '{nmd_val}'\")[plot_col].mean(),\n",
    "#                             gene, fontsize=8, color='black', alpha=0.5)\n",
    "#                 gene_counter_dict[strain][nmd_val] += 1\n",
    "#             added_points_rows.append(rows)\n",
    "#             all_points_rows.append(rows)\n",
    "#         added_points_df = pd.concat(added_points_rows)\n",
    "#         for nmd_val, color in colors_dict.items():\n",
    "#             subplot_df = added_points_df.query(f\"nmd_target_type == '{nmd_val}'\")\n",
    "#             ax.boxplot(subplot_df[plot_col],\n",
    "#                        vert=True,\n",
    "#                        positions=[box_pos_scaler*adjusted_value_dict[nmd_val]+base_value_dict[strain]],\n",
    "#                        widths=(adjusted_value_dict[nmd_val]*box_width_scaler),\n",
    "#                        meanline=True,\n",
    "#                        patch_artist=True,\n",
    "#                        showmeans=True,\n",
    "#                        showfliers=False,\n",
    "#                        boxprops=dict(facecolor=color),\n",
    "#                        )\n",
    "#     better_tick_labels = {}\n",
    "#     for old_label, pos in base_value_dict.items():\n",
    "#         better_tick_labels[(f\"{old_label}\\n\"\n",
    "#                             f\"n = ({gene_counter_dict[old_label]['NMD_NonTargets']}, {gene_counter_dict[old_label]['NMD_Targets']})\")] = pos\n",
    "#     ax.set_xticks(list(better_tick_labels.values()))\n",
    "#     ax.set_xticklabels(list(better_tick_labels.keys()))\n",
    "#     ax.set_ylabel(f\"{plot_num} / {plot_denum}\")\n",
    "#     lines = [plt.Line2D([0], [0], color=colors_dict[nmd_val], lw=2, label=nmd_val) for nmd_val in colors_dict.keys()]\n",
    "#     ax.legend(lines, colors_dict.keys())\n",
    "#     if log_y:\n",
    "#         plt.yscale('log')\n",
    "#     plt.tight_layout()\n",
    "#     save_name = f\"{npCommon.get_dt(for_file=True)}_boxPlot_{plot_num}vs{plot_denum}\"\n",
    "#     if log_y:\n",
    "#         save_name += '_logY'\n",
    "#     else:\n",
    "#         save_name += '_linY'\n",
    "#     save_name += f\"_minAdapted{min_adapted}_minTotal{min_total}_minTSSAda{min_tss_ada}_minReps{min_number_reps_observed}_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "#     if with_connections:\n",
    "#         save_name += '_withConnections'\n",
    "#     else:\n",
    "#         save_name += '_noConnections'\n",
    "#     \n",
    "#     if additional_save_text:\n",
    "#         save_name += f\"_{additional_save_text}\"\n",
    "#     save_dir.mkdir(exist_ok=True, parents=False)\n",
    "#     plt.savefig(save_dir / (save_name+\".png\"), dpi=300)\n",
    "#     plt.savefig(save_dir / (save_name+\".svg\"))\n",
    "#     plt.grid(True, which=\"both\")\n",
    "#     plt.show()\n",
    "#     return (pd.concat(all_points_rows), plot_df,  fig)\n",
    "\n",
    "def merge_ambig_and_targets(input_df: pd.DataFrame, share_with_both=False, NMD_in_wildtype_only: bool = True) -> pd.DataFrame:\n",
    "    # Easy way to go about this is to break up all the different NMD_target_types into their own dataframes, then recombine them as needed w/ some renaming\n",
    "    target_df = input_df.query(\"nmd_target_type == 'NMD_Targets'\").copy()\n",
    "    non_target_df = input_df.query(\"nmd_target_type == 'NMD_NonTargets'\").copy()\n",
    "    \n",
    "    if NMD_in_wildtype_only:\n",
    "        ambig_df = input_df.query(\"nmd_target_type == 'Ambiguous_Reads' and strain == 'wildtype'\").copy()\n",
    "    else:\n",
    "        ambig_df = input_df.query(\"nmd_target_type == 'Ambiguous_Reads'\").copy()\n",
    "    \n",
    "    target_df = target_df.add(ambig_df.rename(index={\"Ambiguous_Reads\": \"NMD_Targets\"}),\n",
    "                              fill_value=0)\n",
    "    if share_with_both:\n",
    "        non_target_df = non_target_df.add(ambig_df.rename(index={\"Ambiguous_Reads\": \"NMD_NonTargets\"}),\n",
    "                                          fill_value=0)\n",
    "    output_df = pd.concat([target_df, non_target_df])\n",
    "    return output_df\n",
    "\n",
    "def process_transcript_lengths(meaned: pd.DataFrame):\n",
    "    NMD_ISO_REGIONS = {\n",
    "        'rpl-30': {'start': 10_436_292,\n",
    "                   'end': 10_436_409,\n",
    "                   'region_is_target': True},\n",
    "        'rps-22': {'start': 1_950_820,  # Now rps-15A\n",
    "                   'end': 1_950_996,\n",
    "                   'region_is_target': True},\n",
    "        'ubl-1': {'start': 3_068_573,  # Now rps-27A\n",
    "                  'end': 3_068_581,\n",
    "                  'region_is_target': False},  # The target is actually shorter than the non-target\n",
    "        'rpl-7A': {'start': 4_389_805,\n",
    "                   'end': 4_389_886,\n",
    "                   'region_is_target': True},\n",
    "        'rpl-3': {'start': 3_868_329,\n",
    "                  'end': 3_868_443,\n",
    "                  'region_is_target': True},\n",
    "        'rpl-1': {'start': 2_876_019,\n",
    "                  'end': 2_876_144,\n",
    "                  'region_is_target': True},\n",
    "        'rpl-12': {'start': 13_240_024,\n",
    "                   'end': 13_240_143,\n",
    "                   'region_is_target': True},\n",
    "        'hel-1': {'start': 8_327_767,\n",
    "                  'end': 8_327_885,\n",
    "                  'region_is_target': True},\n",
    "        'aly-3': {'start': 12_123_830,\n",
    "                  'end': 12_124_176,\n",
    "                  'region_is_target': True},\n",
    "        'rsp-6': {'start': 7_790_498,\n",
    "                  'end': 7_790_745,\n",
    "                  'region_is_target': True},\n",
    "        'K08D12.3': {'start': 1_710_371,\n",
    "                     'end': 1_710_405,\n",
    "                     'region_is_target': True},\n",
    "        'R06C1.4': {'start': 11_931_061,\n",
    "                    'end': 11_931_173,\n",
    "                    'region_is_target': True},\n",
    "        'C53H9.2': {'start': 1_833_384,\n",
    "                    'end': 1_833_408,\n",
    "                    'region_is_target': False},  # The target is actually shorter than the non-target\n",
    "        'rsp-5': {'start': 6_490_718,\n",
    "                  'end': 6_490_883,\n",
    "                  'region_is_target': True},\n",
    "        # These following examples are a little shakier:\n",
    "        'ZK228.4': {'start': 18_462_943,\n",
    "                    'end': 18_462_971,\n",
    "                    'region_is_target': True},  # Not sure about this one's NMD targeted region\n",
    "        'rpl-26': {'start': 8_603_277,\n",
    "                   'end': 8_603_324,\n",
    "                   'region_is_target': True},  # rpl-26 is weird because it doesn't seem to have cleavage buildup/change\n",
    "        'pqn-70': {'start': 11_227_418,\n",
    "                   'end': 11_227_453,\n",
    "                   'region_is_target': False},  # Somewhat unclear, but adding for posterity\n",
    "                                                # The target is actually shorter than the non-target\n",
    "        # 'examp': {'start': 3_068_573,\n",
    "        #           'end': 3_068_581,\n",
    "        #           'region_is_target': True},\n",
    "    }\n",
    "    \n",
    "    region_lengths = []\n",
    "    region_lengths_dict = {}\n",
    "    for gene, region_dict in NMD_ISO_REGIONS.items():\n",
    "        nmd_longer = 'longer target' if region_dict['region_is_target'] else 'longer non-target'\n",
    "        region_length = abs(region_dict['end'] - region_dict['start'])\n",
    "        region_length_adj = region_length if region_dict['region_is_target'] else -region_length\n",
    "        region_lengths.append(region_length_adj)\n",
    "        region_lengths_dict[gene] = region_length_adj\n",
    "    #     print(f\"{gene} NMD ID region: {region_length_adj}nt ({nmd_longer})\")\n",
    "    # print(f\"\\nRegion Lengths Mean: {np.mean(region_lengths):0.2f}; Median {np.median(region_lengths):0.2f}; Std: {np.std(region_lengths):0.2f}\")\n",
    "    \n",
    "    genes_df = pd.read_parquet(current_dir / 'preprocessed_data' / 'all_genes_avgTranscriptLength_and_nearTSS.parquet')\n",
    "    just_transcript_lengths = genes_df[['avg_transcript_length', 'std_transcript_length']].reset_index(\n",
    "        level=['strain', 'rep', 'lib', 'nmd_target_type'],\n",
    "        drop=True).reset_index().drop_duplicates()\n",
    "    \n",
    "    meaned = meaned.merge(just_transcript_lengths, on='gene_name', how='left')\n",
    "    \n",
    "    def adjust_length(row: pd.Series) -> float:\n",
    "        gene = row.gene_name\n",
    "        length_adjustment = region_lengths_dict.get(gene, 0) / 2\n",
    "        nmd_target_type = row.nmd_target_type\n",
    "        if nmd_target_type == 'NMD_Targets':\n",
    "            return row.avg_transcript_length + length_adjustment\n",
    "        else:\n",
    "            return row.avg_transcript_length - length_adjustment\n",
    "    \n",
    "    meaned['adj_transcript_length'] = meaned.apply(adjust_length, axis=1)\n",
    "    return meaned\n",
    "\n",
    "def plot_box_and_strip_with_connections2(in_df: pd.DataFrame,\n",
    "                                         plot_num: str,  # 'adapted_near_tss'\n",
    "                                         plot_denum: str,  # 'total_adapted' or 'total_reads'\n",
    "                                         min_adapted: int = 3,\n",
    "                                         min_total: int = 10,\n",
    "                                         add_tss_ada_pseudo: float = 0.0,\n",
    "                                         min_number_reps_observed: int = -1,\n",
    "                                         with_connections: str = '',  # 'length' or 'dashed' or ''/'none'\n",
    "                                         with_color_bar: bool = True,\n",
    "                                         log2_y: bool = False,\n",
    "                                         log_y: bool = False,\n",
    "                                         with_gene_names: bool = True,\n",
    "                                         limited_gene_names: list = [],  # if any genes are listed here, only those genes will be labeled\n",
    "                                         save_dir: Path = current_dir / 'plots' / 'decapping_connectedMeans',\n",
    "                                         additional_save_text:str = '',\n",
    "                                         print_failed_genes: bool = False,\n",
    "                                         remove_manual_outliers: bool = True,\n",
    "                                         print_mean_median_vals: bool = True,\n",
    "                                         figsize: Tuple[int, int] = (4, 4),\n",
    "                                         ) -> Tuple[pd.DataFrame, pd.DataFrame, go.Figure]:\n",
    "    random.seed(42)\n",
    "    fig, ax = plt.subplots(figsize=figsize,)\n",
    "    spacing_factor = 0.20  # 0.15\n",
    "    box_pos_scaler = 3.25   # 3.5\n",
    "    box_width_scaler = 3.0 # 2.5\n",
    "    \n",
    "    \n",
    "    plot_col = f'{plot_num} / {plot_denum}'\n",
    "    plot_df = in_df.copy()\n",
    "    \n",
    "    if print_failed_genes:\n",
    "        print(f\"Failed Genes at `total_adapted >= {min_adapted}` or `total_reads >= {min_total}` levels:\")\n",
    "        print(plot_df.query(f\"total_adapted < {min_adapted} | total_reads < {min_total}\"))\n",
    "    plot_df = plot_df.query(f\"total_adapted >= {min_adapted} & total_reads >= {min_total}\")\n",
    "    if with_connections == 'length':\n",
    "        plot_df = process_transcript_lengths(plot_df.reset_index())\n",
    "    else:\n",
    "        plot_df = plot_df.reset_index()\n",
    "    plot_df[plot_col] = plot_df[plot_num] / plot_df[plot_denum]\n",
    "    \n",
    "    plot_df = process_df_for_means(plot_df,\n",
    "                                   plot_num, plot_denum,\n",
    "                                   min_total_reads=min_total,\n",
    "                                   min_adapted_reads=min_adapted,\n",
    "                                   min_tss_ada_reads=-1,\n",
    "                                   add_tss_ada_pseudo=add_tss_ada_pseudo,\n",
    "                                   min_number_reps_observed=min_number_reps_observed,\n",
    "                                   log2_y=log2_y,\n",
    "                                   log_y=log_y,\n",
    "                                   print_failed_genes=print_failed_genes,\n",
    "                                   )\n",
    "    base_value_dict = {'wildtype': 0,\n",
    "                       'smg-5': 2,\n",
    "                       'smg-6': 4,\n",
    "                       # 'smg-7': 4,\n",
    "                       }\n",
    "    adjusted_value_dict = {'NMD_Targets': spacing_factor,\n",
    "                           'NMD_NonTargets': -spacing_factor}\n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    if 'adapted' in plot_col:\n",
    "        colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                       'NMD_NonTargets': plot_colors[1][0],}\n",
    "    else:\n",
    "        colors_dict = {'NMD_Targets': plot_colors[0][1],\n",
    "                       'NMD_NonTargets': plot_colors[1][1],}\n",
    "    \n",
    "    def adjust_value2(row: pd.Series) -> float:\n",
    "        strain = row.strain\n",
    "        nmd_target_type = row.nmd_target_type\n",
    "        return base_value_dict[strain] + adjusted_value_dict[nmd_target_type] + (random.random() - 0.5) * 0.1\n",
    "    \n",
    "    # Define the color map and normalization\n",
    "    cmap = cm.get_cmap('RdYlBu')\n",
    "    norm = Normalize(vmin=-50, vmax=150)\n",
    "    \n",
    "    # Function to convert a value to a color\n",
    "    def value_to_color(value: float) -> tuple:\n",
    "        normalized_value = norm(value)\n",
    "        return cmap(normalized_value)\n",
    "    \n",
    "    plot_df['adjusted_x'] = plot_df.apply(adjust_value2, axis=1)\n",
    "    \n",
    "    all_points_rows = []\n",
    "    gene_counter_dict = {strain: dict(NMD_Targets=0, NMD_NonTargets=0) for strain in base_value_dict.keys()}\n",
    "    for strain in plot_df.strain.unique():\n",
    "        added_points_rows = []\n",
    "        for gene in plot_df.gene_name.unique():\n",
    "            rows = plot_df.query(f\"gene_name == '{gene}' & strain == '{strain}'\")\n",
    "            nmd_vals = rows.nmd_target_type.unique()\n",
    "            \n",
    "            # Let's get rid of our extreme outliers:\n",
    "            if remove_manual_outliers:\n",
    "                if gene == 'hel-1' and strain in ['smg-5', 'smg-6']:\n",
    "                    continue\n",
    "                if gene == 'rpl-3' and strain == 'smg-5':\n",
    "                    continue\n",
    "                # elif gene == 'pqn-70' and strain == 'smg-6':\n",
    "                #     continue\n",
    "                if gene == 'aly-3' and strain == 'smg-6':\n",
    "                    continue\n",
    "                if gene == 'C53H9.2' and strain == 'smg-5':\n",
    "                    continue\n",
    "                if gene == 'rpl-1' and strain == 'smg-6':\n",
    "                    continue\n",
    "                if gene == 'rsp-6' and strain == 'smg-5':\n",
    "                    continue\n",
    "            \n",
    "            if len(nmd_vals) == 2 and with_connections.lower() == 'length':\n",
    "                # Okay, here is where we want to calculate the difference in NMD targets vs non target legnths and color the dashe3d line by that!!\n",
    "                # We will also want to add a colorbar to the plot eventually\n",
    "                difference_in_adj_transcript_len = rows.query(\"nmd_target_type == 'NMD_Targets'\")['adj_transcript_length'].mean() - rows.query(\"nmd_target_type == 'NMD_NonTargets'\")['adj_transcript_length'].mean()\n",
    "                color = value_to_color(difference_in_adj_transcript_len)\n",
    "                ax.plot(rows['adjusted_x'], rows[plot_col],\n",
    "                        color=color, # color='black',\n",
    "                        # alpha=0.5,\n",
    "                        linewidth=0.95,\n",
    "                        # dashes=[6, 2],\n",
    "                        )\n",
    "            elif len(nmd_vals) == 2 and with_connections.lower() != 'none' and with_connections:\n",
    "                # Okay, here is where we want to calculate the difference in NMD targets vs non target legnths and color the dashe3d line by that!!\n",
    "                # We will also want to add a colorbar to the plot eventually\n",
    "                ax.plot(rows['adjusted_x'], rows[plot_col],\n",
    "                        color='black',\n",
    "                        alpha=0.5,\n",
    "                        linewidth=0.5,\n",
    "                        dashes=[6, 2],\n",
    "                        )\n",
    "            for nmd_val in nmd_vals:\n",
    "                ax.plot(rows.query(f\"nmd_target_type == '{nmd_val}'\")['adjusted_x'],\n",
    "                        rows.query(f\"nmd_target_type == '{nmd_val}'\")[plot_col],\n",
    "                        marker='o', linestyle='', markersize=5, color=colors_dict[nmd_val],\n",
    "                        markeredgecolor='black', markeredgewidth=0.5,)\n",
    "                if with_gene_names:\n",
    "                    if not limited_gene_names or gene in limited_gene_names:\n",
    "                        ax.text(rows.query(f\"nmd_target_type == '{nmd_val}'\")['adjusted_x'].mean(),\n",
    "                                rows.query(f\"nmd_target_type == '{nmd_val}'\")[plot_col].mean(),\n",
    "                                gene, fontsize=8, color='black', alpha=0.5)\n",
    "                gene_counter_dict[strain][nmd_val] += 1\n",
    "            added_points_rows.append(rows)\n",
    "            all_points_rows.append(rows)\n",
    "        added_points_df = pd.concat(added_points_rows)\n",
    "        for nmd_val, color in colors_dict.items():\n",
    "            subplot_df = added_points_df.query(f\"nmd_target_type == '{nmd_val}'\")\n",
    "            ax.boxplot(subplot_df[plot_col],\n",
    "                       vert=True,\n",
    "                       positions=[box_pos_scaler*adjusted_value_dict[nmd_val]+base_value_dict[strain]],\n",
    "                       widths=(adjusted_value_dict[nmd_val]*box_width_scaler),\n",
    "                       # meanline=True,\n",
    "                       patch_artist=True,\n",
    "                       # showmeans=True,\n",
    "                       showfliers=False,\n",
    "                       boxprops=dict(facecolor=color),\n",
    "                       )\n",
    "            if print_mean_median_vals:\n",
    "                print(f\"{strain} {nmd_val} mean: {subplot_df[plot_col].mean()*100:0.6f}; median: {subplot_df[plot_col].median()*100:0.6f}\")\n",
    "    better_tick_labels = {}\n",
    "    for old_label, pos in base_value_dict.items():\n",
    "        better_tick_labels[(f\"{old_label}\\n\"\n",
    "                            f\"n = ({gene_counter_dict[old_label]['NMD_NonTargets']}, {gene_counter_dict[old_label]['NMD_Targets']})\")] = pos\n",
    "    ax.set_xticks(list(better_tick_labels.values()))\n",
    "    ax.set_xticklabels(list(better_tick_labels.keys()))\n",
    "    \n",
    "    # ax.set_yticks([0.0001, 0.001, 0.01, 0.1, 1, 10])\n",
    "    # ax.set_yticklabels(['0.0001', '0.001', '0.01', '0.1', '1', '10'])\n",
    "    \n",
    "    ax.set_ylabel(f\"{plot_num} / {plot_denum}\")\n",
    "    \n",
    "    lines = [plt.Line2D([0], [0], color=colors_dict[nmd_val], lw=2, label=nmd_val) for nmd_val in colors_dict.keys()]\n",
    "    ax.legend(lines, colors_dict.keys())\n",
    "    if log_y:\n",
    "        plt.yscale('log')\n",
    "    if with_connections.lower() == 'length' and with_color_bar:\n",
    "        cbar = fig.colorbar(ScalarMappable(norm=norm, cmap=cmap), ax=ax)\n",
    "        cbar.set_label('Difference in Adjusted Transcript Length')\n",
    "    \n",
    "    save_name = f\"{npCommon.get_dt(for_file=True)}_boxPlot_{plot_num}vs{plot_denum}\"\n",
    "    title_text = f\"{plot_num} vs {plot_denum}\\n\"\n",
    "    if log_y:\n",
    "        save_name += '_logY'\n",
    "    else:\n",
    "        save_name += '_linY'\n",
    "    save_name += f\"_minAdapted{min_adapted}_minTotal{min_total}_minReps{min_number_reps_observed}_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    title_text += f\"Min Adapt: {min_adapted}; Min Total: {min_total}; Min Reps: {min_number_reps_observed}; Pseudo TSS: {add_tss_ada_pseudo}\"\n",
    "    if with_connections.lower() == 'length':\n",
    "        save_name += '_withLengthConnections'\n",
    "    elif with_connections.lower() != 'none' and with_connections:\n",
    "        save_name += '_withConnections'\n",
    "    else:\n",
    "        save_name += '_noConnections'\n",
    "    \n",
    "    if additional_save_text:\n",
    "        save_name += f\"_{additional_save_text}\"\n",
    "        title_text += f\"\\n{additional_save_text.replace('_', ' ')}\"\n",
    "    save_dir.mkdir(exist_ok=True, parents=False)\n",
    "    \n",
    "    plt.title(title_text)\n",
    "    plt.grid(True, which=\"both\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(save_dir / (save_name+\".png\"), dpi=300)\n",
    "    plt.savefig(save_dir / (save_name+\".svg\"))\n",
    "    plt.show()\n",
    "    return pd.concat(all_points_rows), plot_df,  fig"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "a9f06d9ccd9bd82a",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "overall_save_dir = Path(f\"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/decappingFigure/plotsFromPython/{npCommon.get_dt(for_file=True)}_decappingBoxPlots\")\n",
    "overall_save_dir.mkdir(exist_ok=True, parents=False)\n",
    "\n",
    "# Okay let's run it:\n",
    "for merge_ambig in [\n",
    "    True,\n",
    "    # False,\n",
    "]:\n",
    "    # Note that we can only really predict that most of the ambiguous reads come from the NMD target species in the wildtype strains.\n",
    "    # Because this is the simplest assumption we can make, I'm not carrying any of the ambiguous reads through from the smg-5 and smg-6 strains.\n",
    "    run_df = better_df.copy()\n",
    "    if merge_ambig:\n",
    "        run_df = merge_ambig_and_targets(run_df,\n",
    "                                         share_with_both=False,\n",
    "                                         NMD_in_wildtype_only=True)\n",
    "    else:\n",
    "        run_df = run_df.loc[:, :, :, :, ['NMD_Targets', 'NMD_NonTargets'], :]\n",
    "    \n",
    "    for targ_num, targ_denom in [\n",
    "        ('adapted_near_tss', 'total_adapted'),\n",
    "        ('total_near_tss', 'total_reads'),\n",
    "        ('adapted_near_tss', 'total_reads'),\n",
    "        ('total_adapted', 'total_reads'),\n",
    "    ]:\n",
    "        for with_names in [\n",
    "            True,\n",
    "            # False,\n",
    "        ]:\n",
    "            for with_connections in [\n",
    "                # 'length',\n",
    "                'dashed',\n",
    "                # 'none',\n",
    "            ]:\n",
    "                add_save_text = '_withAmbiguousInWT' if merge_ambig else '_noAmbiguous'\n",
    "                add_save_text += '_withNames' if with_names else '_noNames'\n",
    "                plotted, meaned, fig = plot_box_and_strip_with_connections2(run_df,\n",
    "                                                    targ_num, targ_denom,\n",
    "                                                    min_adapted=1 if 'adapted' in targ_num else 0,\n",
    "                                                    min_total=1,\n",
    "                                                    add_tss_ada_pseudo=0.01,\n",
    "                                                    min_number_reps_observed=2,\n",
    "                                                    additional_save_text=add_save_text,\n",
    "                                                    with_gene_names=with_names,\n",
    "                                                    limited_gene_names=['rpl-30', 'rps-22', 'ubl-1'],\n",
    "                                                    with_connections=with_connections,\n",
    "                                                    with_color_bar=False,\n",
    "                                                    save_dir=overall_save_dir / 'withAmbigMergeForWT' if merge_ambig else overall_save_dir / 'noAmbigReads',\n",
    "                                                    print_failed_genes=False,\n",
    "                                                    remove_manual_outliers=True,\n",
    "                                                    log_y=True)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "1e5091a217430878",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "I am grabbing the below manual annotations of NMD target regions from `ambiguousReads.py`"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c17b5366715fd032"
  },
  {
   "cell_type": "code",
   "source": [
    "base_value_dict = {'wildtype': 0,\n",
    "                       'smg-5': 2,\n",
    "                       'smg-6': 4,\n",
    "                       # 'smg-7': 4,\n",
    "                       }\n",
    "# These colors are updated from the final plots!\n",
    "plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "               ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "               ('#676767', '#414141'),  # ambiguous greys\n",
    "               )\n",
    "colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "               'NMD_NonTargets': plot_colors[1][0],}"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2f9483f909ef868b",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "NMD_ISO_REGIONS = {\n",
    "    'rpl-30': {'start': 10_436_292,\n",
    "               'end': 10_436_409,\n",
    "               'region_is_target': True},\n",
    "    'rps-22': {'start': 1_950_820,\n",
    "               'end': 1_950_996,\n",
    "               'region_is_target': True},\n",
    "    'ubl-1': {'start': 3_068_573,\n",
    "              'end': 3_068_581,\n",
    "              'region_is_target': False},  # The target is actually shorter than the non-target\n",
    "    'rpl-7A': {'start': 4_389_805,\n",
    "               'end': 4_389_886,\n",
    "               'region_is_target': True},\n",
    "    'rpl-3': {'start': 3_868_329,\n",
    "              'end': 3_868_443,\n",
    "              'region_is_target': True},\n",
    "    'rpl-1': {'start': 2_876_019,\n",
    "              'end': 2_876_144,\n",
    "              'region_is_target': True},\n",
    "    'rpl-12': {'start': 13_240_024,\n",
    "               'end': 13_240_143,\n",
    "               'region_is_target': True},\n",
    "    'hel-1': {'start': 8_327_767,\n",
    "              'end': 8_327_885,\n",
    "              'region_is_target': True},\n",
    "    'aly-3': {'start': 12_123_830,\n",
    "              'end': 12_124_176,\n",
    "              'region_is_target': True},\n",
    "    'rsp-6': {'start': 7_790_498,\n",
    "              'end': 7_790_745,\n",
    "              'region_is_target': True},\n",
    "    'K08D12.3': {'start': 1_710_371,\n",
    "                 'end': 1_710_405,\n",
    "                 'region_is_target': True},\n",
    "    'R06C1.4': {'start': 11_931_061,\n",
    "                'end': 11_931_173,\n",
    "                'region_is_target': True},\n",
    "    'C53H9.2': {'start': 1_833_384,\n",
    "                'end': 1_833_408,\n",
    "                'region_is_target': False},  # The target is actually shorter than the non-target\n",
    "    'rsp-5': {'start': 6_490_718,\n",
    "              'end': 6_490_883,\n",
    "              'region_is_target': True},\n",
    "    # These following examples are a little shakier:\n",
    "    'ZK228.4': {'start': 18_462_943,\n",
    "                'end': 18_462_971,\n",
    "                'region_is_target': True},  # Not sure about this one's NMD targeted region\n",
    "    'rpl-26': {'start': 8_603_277,\n",
    "               'end': 8_603_324,\n",
    "               'region_is_target': True},  # rpl-26 is weird because it doesn't seem to have cleavage buildup/change\n",
    "    'pqn-70': {'start': 11_227_418,\n",
    "               'end': 11_227_453,\n",
    "               'region_is_target': False},  # Somewhat unclear, but adding for posterity\n",
    "                                            # The target is actually shorter than the non-target\n",
    "    # 'examp': {'start': 3_068_573,\n",
    "    #           'end': 3_068_581,\n",
    "    #           'region_is_target': True},\n",
    "}\n",
    "\n",
    "region_lengths = []\n",
    "region_lengths_for_df = []\n",
    "region_lengths_dict = {}\n",
    "for gene, region_dict in NMD_ISO_REGIONS.items():\n",
    "    nmd_longer = 'longer target' if region_dict['region_is_target'] else 'longer non-target'\n",
    "    region_length = abs(region_dict['end'] - region_dict['start'])\n",
    "    region_length_adj = region_length if region_dict['region_is_target'] else -region_length\n",
    "    region_lengths.append(region_length_adj)\n",
    "    region_lengths_dict[gene] = region_length_adj\n",
    "    region_lengths_for_df.append({'gene': gene,\n",
    "                                 'region_length': region_length,\n",
    "                                 'region_length_adj': region_length_adj,\n",
    "                                 'nmd_longer': nmd_longer})\n",
    "    print(f\"{gene} NMD ID region: {region_length_adj}nt ({nmd_longer})\")\n",
    "print(f\"\\nRegion Lengths Mean: {np.mean(region_lengths):0.2f}; Median {np.median(region_lengths):0.2f}; Std: {np.std(region_lengths):0.2f}\")\n",
    "region_lengths_df = pd.DataFrame(region_lengths_for_df)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "18ec2d82afe25ee5",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "sea.swarmplot(data=region_lengths_df, y='gene', x='region_length_adj', hue='nmd_longer')\n",
    "other_plots_dir = current_dir / 'plots' / 'other'\n",
    "other_plots_dir.mkdir(exist_ok=True, parents=True)\n",
    "plt.tight_layout()\n",
    "plt.savefig(other_plots_dir / f\"{npCommon.get_dt(for_file=True)}_NMD_region_lengths.png\", dpi=300)\n",
    "plt.show()\n",
    "region_lengths_df"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "bf54623ae7d11d87",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Now let's quickly try to produce the curve scatter plot of reads near the TSS / total reads vs. transcript length!\n",
    "\n",
    "Since I don't want to MANUALLY find the lengths of the MANUAL annotations I produced of the NMD targets and their unique regions:\n",
    "I am going to take the average from my calculations in `decappingQuantification3_allGenes.ippynb` and then use the length from above to calculate the adjusted length of the NMD targeted regions."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "eb48f34ec796b3fd"
  },
  {
   "cell_type": "code",
   "source": [
    "# First let's load the preprocessed data from the all_genes NB (THIS TOOK NEARLY 3 HOURS TO GENERATE)\n",
    "genes_df = pd.read_parquet(current_dir / 'preprocessed_data' / 'all_genes_avgTranscriptLength_and_nearTSS.parquet')\n",
    "genes_df.query(\"gene_name == '21ur-10296'\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "153fb161dbe52be",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "just_transcript_lengths = genes_df[['avg_transcript_length', 'std_transcript_length']].reset_index(level=['strain', 'rep', 'lib', 'nmd_target_type'], drop=True).reset_index().drop_duplicates()\n",
    "just_transcript_lengths.head()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "fb56c4cc03d67b23",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "genes_df['total_near_tss / total_reads'] = genes_df['total_near_tss'] / genes_df['total_reads']\n",
    "genes_df['nmd_active'] = genes_df.index.get_level_values('strain') == 'wildtype'\n",
    "genes_df.head(50)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "dca9f83531d25622",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "groupby_genes_obj = genes_df.query(\"total_reads >= 1\").reset_index().groupby(['gene_name', 'nmd_active'], observed=False)\n",
    "groupby_genes = groupby_genes_obj[['adapted_near_tss', 'total_adapted', 'total_reads', 'unadapted_near_tss', 'total_near_tss', 'total_near_tss / total_reads']].mean()\n",
    "groupby_genes['observed_reps'] = groupby_genes_obj['rep'].apply(lambda x: ','.join(map(str, x)))\n",
    "groupby_genes['observed_strains'] = groupby_genes_obj['strain'].apply(lambda x: ','.join(x))\n",
    "groupby_genes['num_observed_reps'] = groupby_genes_obj['rep'].count()\n",
    "groupby_genes = groupby_genes.reset_index().merge(just_transcript_lengths, on='gene_name', how='left')\n",
    "groupby_genes.query(\"num_observed_reps >= 3\").query(\"total_adapted >= 1\").head(25)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5d392f4e7e168692",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "sea.scatterplot(data=groupby_genes.query(\"num_observed_reps >= 2\").query(\"total_adapted >= 1\"), x='avg_transcript_length', y='total_near_tss / total_reads', hue='nmd_active')\n",
    "plt.xscale('log')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "7f9ec8c440afe65f",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "fig = px.scatter(groupby_genes.query(\"num_observed_reps >= 2\").query(\"total_adapted >= 1\").query(\"nmd_active == False\"),\n",
    "                 x='avg_transcript_length', y='total_near_tss / total_reads',\n",
    "                 color='nmd_active',\n",
    "                 log_x=True,\n",
    "                 hover_name='gene_name',\n",
    "                 hover_data=['total_adapted', 'total_reads', 'observed_reps', 'num_observed_reps'],)\n",
    "fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "86338454f1f88fad",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "genes_df = pd.read_parquet(current_dir / 'preprocessed_data' / 'all_genes_avgTranscriptLength_and_nearTSS.parquet')\n",
    "just_transcript_lengths = genes_df[['avg_transcript_length', 'std_transcript_length']].reset_index(\n",
    "    level=['strain', 'rep', 'lib', 'nmd_target_type'],\n",
    "    drop=True).reset_index().drop_duplicates()\n",
    "\n",
    "meaned = meaned.merge(just_transcript_lengths, on='gene_name', how='left')\n",
    "\n",
    "def adjust_length(row: pd.Series) -> float:\n",
    "    print(row)\n",
    "    gene = row.gene_name\n",
    "    length_adjustment = region_lengths_dict.get(gene, 0) / 2\n",
    "    nmd_target_type = row.nmd_target_type\n",
    "    if nmd_target_type == 'NMD_Targets':\n",
    "        return row.avg_transcript_length + length_adjustment\n",
    "    else:\n",
    "        return row.avg_transcript_length - length_adjustment\n",
    "\n",
    "meaned['adj_transcript_length'] = meaned.apply(adjust_length, axis=1)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "49dedf1aae5f0e00",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "x_col = 'adj_transcript_length'\n",
    "# y_num = 'total_near_tss'  # 'adapted_near_tss' or 'total_near_tss'\n",
    "# y_denom = 'total_reads'  # 'total_adapted' or 'total_reads'\n",
    "# y_col = f'{y_num} / {y_denom}'\n",
    "# only_wt = True\n",
    "\n",
    "def plot_scatter_with_connections(input_df: pd.DataFrame,\n",
    "                                  groupby_genes: pd.DataFrame,\n",
    "                                  x_col: str,\n",
    "                                  y_num: str,\n",
    "                                  y_denom: str,\n",
    "                                  output_dir: Path = current_dir / 'plots' / 'other',\n",
    "                                  only_wt: bool = True,\n",
    "                                  ) -> None:\n",
    "    y_col = f'{y_num} / {y_denom}'\n",
    "    plotter_df = (input_df.query(\"total_adapted >= 1\")\n",
    "                  .copy())\n",
    "    if only_wt:\n",
    "        plotter_df = plotter_df.query(\"strain == 'wildtype'\")\n",
    "    else:\n",
    "        plotter_df = plotter_df.query(\"strain != 'wildtype'\")\n",
    "    \n",
    "    plotter_df[y_col] = plotter_df[y_num] / plotter_df[y_denom]\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    \n",
    "    groupby_genes[y_col] = groupby_genes[y_num] / groupby_genes[y_denom]\n",
    "    sea.scatterplot(data=groupby_genes.query(\"num_observed_reps >= 2\").query(\"nmd_active == @only_wt\").query(\"total_reads >= 10\"),\n",
    "                    x='avg_transcript_length', y=y_col,\n",
    "                    alpha=0.4)\n",
    "    \n",
    "    \n",
    "    for gene in plotter_df.gene_name.unique():\n",
    "        for strain in plotter_df.strain.unique():\n",
    "            rows = plotter_df.query(f\"gene_name == '{gene}' & strain == '{strain}'\")\n",
    "            nmd_vals = rows.nmd_target_type.unique()\n",
    "            if len(nmd_vals) == 2:\n",
    "                for nmd_val in nmd_vals:\n",
    "                    ax.plot(rows.query(f\"nmd_target_type == '{nmd_val}'\")[x_col],\n",
    "                            rows.query(f\"nmd_target_type == '{nmd_val}'\")[y_col],\n",
    "                            marker='o', linestyle='', markersize=5, color=colors_dict[nmd_val],\n",
    "                            # I want to adda black outline to these points\n",
    "                            markeredgecolor='black', markeredgewidth=0.5,\n",
    "                            )\n",
    "                    ax.annotate(gene, (rows[x_col].mean(), rows[y_col].mean()), fontsize=8, color='black', alpha=0.5)\n",
    "                    if nmd_val == 'NMD_Targets':\n",
    "                        x1 = rows.query(f\"nmd_target_type == 'NMD_Targets'\")[x_col].values[0]\n",
    "                        y1 = rows.query(f\"nmd_target_type == 'NMD_Targets'\")[y_col].values[0]\n",
    "                    else:\n",
    "                        x2 = rows.query(f\"nmd_target_type == 'NMD_NonTargets'\")[x_col].values[0]\n",
    "                        y2 = rows.query(f\"nmd_target_type == 'NMD_NonTargets'\")[y_col].values[0]\n",
    "                # ax.plot(rows[x_col], rows[y_col], color='black', alpha=0.8, linewidth=1, dashes=[6, 2])\n",
    "                prop = dict(arrowstyle=\"-|>,head_width=0.2,head_length=0.4\",\n",
    "                            shrinkA=0,shrinkB=0, lw=1, color='black')\n",
    "                plt.annotate(\"\",\n",
    "                             arrowprops=prop,\n",
    "                             xy=(x1, y1), xycoords='data',\n",
    "                             xytext=(x2, y2), textcoords='data',\n",
    "                             )\n",
    "    plt.xscale('log')\n",
    "    ax.set_ylabel(y_col)\n",
    "    ax.set_xlabel('Transcript Length')\n",
    "    title = f\"{y_col} vs. Transcript Length\"\n",
    "    save_name = f\"{npCommon.get_dt(for_file=True)}_NMDRegionLengths\"\n",
    "    save_name += f\"_{y_num}vs{y_denom}\"\n",
    "    if only_wt:\n",
    "        title += \"\\n(Wildtype Only)\"\n",
    "        save_name += \"_wildtype\"\n",
    "    else:\n",
    "        title += \"\\n(SMG Mutants Only)\"\n",
    "        save_name += \"_smgMutants\"\n",
    "    ax.set_title(title)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(output_dir / (save_name+\".png\"), dpi=300)\n",
    "    plt.savefig(output_dir / (save_name+\".svg\"))\n",
    "    plt.show()\n",
    "    return ax\n",
    "\n",
    "def plot_scatter_with_connections_and_fit(input_df: pd.DataFrame,\n",
    "                                  groupby_genes: pd.DataFrame,\n",
    "                                  x_col: str,\n",
    "                                  y_num: str,\n",
    "                                  y_denom: str,\n",
    "                                  output_dir: Path = current_dir / 'plots' / 'other',\n",
    "                                  only_wt: bool = True,\n",
    "                                  ) -> None:\n",
    "    y_col = f'{y_num} / {y_denom}'\n",
    "    plotter_df = (input_df.query(\"total_adapted >= 1\")\n",
    "                  .copy())\n",
    "    if only_wt:\n",
    "        plotter_df = plotter_df.query(\"strain == 'wildtype'\")\n",
    "    else:\n",
    "        plotter_df = plotter_df.query(\"strain != 'wildtype'\")\n",
    "    \n",
    "    plotter_df[y_col] = plotter_df[y_num] / plotter_df[y_denom]\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    \n",
    "    groupby_genes[y_col] = groupby_genes[y_num] / groupby_genes[y_denom]\n",
    "    data = groupby_genes.query(\"num_observed_reps >= 2\").query(\"nmd_active == @only_wt\").query(\"total_reads >= 50\")\n",
    "    from scipy.optimize import curve_fit\n",
    "\n",
    "    def arctan_func(x: float, a: float, b: float, z: float, w: float) -> float:\n",
    "        # z is around 1/3 or 1/2 and w is between 1/2 and 3/5? maybe\n",
    "        # a is likely near 1/1000, and b is 1000 or more\n",
    "        return z * np.arctan(a * (-x + b)) + w\n",
    "    \n",
    "    def bounded_arctan_func(x: float, a: float, b: float, z: float, w: float) -> float:\n",
    "        return z * np.arctan(a * (-x + b)) + w\n",
    "    \n",
    "    data_x = data['avg_transcript_length']\n",
    "    data_y = data[y_col]\n",
    "    # Combine data_x and data_y into a DataFrame\n",
    "    data_simple = pd.DataFrame({'x': data_x, 'y': data_y})\n",
    "    \n",
    "    # Remove rows with NaN or infinite values\n",
    "    data_simple = data_simple[np.isfinite(data_simple['x']) & np.isfinite(data_simple['y'])]\n",
    "    \n",
    "    # Separate the cleaned data_x and data_y\n",
    "    data_x = data_simple['x']\n",
    "    data_y = data_simple['y']\n",
    "    \n",
    "    # Initial guesses for the parameters\n",
    "    initial_guesses = [1/1000, 1000, 1/3, 1/2]\n",
    "    \n",
    "    # Bounds for the parameters\n",
    "    param_bounds = (0, [np.inf, np.inf, 1, 1])\n",
    "    \n",
    "    # Fit the arctan function to the data\n",
    "    popt, pcov = curve_fit(bounded_arctan_func, data_x, data_y, p0=initial_guesses, bounds=param_bounds)\n",
    "    \n",
    "    # Print the fitted parameters for debugging\n",
    "    print(\"Fitted parameters:\", popt)\n",
    "    \n",
    "    # Generate x values for the fitted curve\n",
    "    x_fit = np.linspace(min(data_x), max(data_x), 500)\n",
    "    y_fit = bounded_arctan_func(x_fit, *popt)\n",
    "    \n",
    "    ax.scatter(data_x, data_y, alpha=0.4, label='Data')\n",
    "    ax.plot(x_fit, y_fit, color='red', linestyle='--', label='Arctan Fit')\n",
    "    \n",
    "    sea.scatterplot(data=data,\n",
    "                    x='avg_transcript_length', y=y_col,\n",
    "                    alpha=0.4)\n",
    "    \n",
    "    \n",
    "    for gene in plotter_df.gene_name.unique():\n",
    "        for strain in plotter_df.strain.unique():\n",
    "            rows = plotter_df.query(f\"gene_name == '{gene}' & strain == '{strain}'\")\n",
    "            nmd_vals = rows.nmd_target_type.unique()\n",
    "            if len(nmd_vals) == 2:\n",
    "                for nmd_val in nmd_vals:\n",
    "                    ax.plot(rows.query(f\"nmd_target_type == '{nmd_val}'\")[x_col],\n",
    "                            rows.query(f\"nmd_target_type == '{nmd_val}'\")[y_col],\n",
    "                            marker='o', linestyle='', markersize=5, color=colors_dict[nmd_val],\n",
    "                            # I want to adda black outline to these points\n",
    "                            markeredgecolor='black', markeredgewidth=0.5,\n",
    "                            )\n",
    "                    ax.annotate(gene, (rows[x_col].mean(), rows[y_col].mean()), fontsize=8, color='black', alpha=0.5)\n",
    "                    if nmd_val == 'NMD_Targets':\n",
    "                        x1 = rows.query(f\"nmd_target_type == 'NMD_Targets'\")[x_col].values[0]\n",
    "                        y1 = rows.query(f\"nmd_target_type == 'NMD_Targets'\")[y_col].values[0]\n",
    "                    else:\n",
    "                        x2 = rows.query(f\"nmd_target_type == 'NMD_NonTargets'\")[x_col].values[0]\n",
    "                        y2 = rows.query(f\"nmd_target_type == 'NMD_NonTargets'\")[y_col].values[0]\n",
    "                # ax.plot(rows[x_col], rows[y_col], color='black', alpha=0.8, linewidth=1, dashes=[6, 2])\n",
    "                prop = dict(arrowstyle=\"-|>,head_width=0.2,head_length=0.4\",\n",
    "                            shrinkA=0,shrinkB=0, lw=1, color='black')\n",
    "                plt.annotate(\"\",\n",
    "                             arrowprops=prop,\n",
    "                             xy=(x1, y1), xycoords='data',\n",
    "                             xytext=(x2, y2), textcoords='data',\n",
    "                             )\n",
    "    plt.xscale('log')\n",
    "    ax.set_ylabel(y_col)\n",
    "    ax.set_xlabel('Transcript Length')\n",
    "    title = f\"{y_col} vs. Transcript Length\"\n",
    "    save_name = f\"{npCommon.get_dt(for_file=True)}_NMDRegionLengths\"\n",
    "    save_name += f\"_{y_num}vs{y_denom}\"\n",
    "    if only_wt:\n",
    "        title += \"\\n(Wildtype Only)\"\n",
    "        save_name += \"_wildtype\"\n",
    "    else:\n",
    "        title += \"\\n(SMG Mutants Only)\"\n",
    "        save_name += \"_smgMutants\"\n",
    "    ax.set_title(title)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(output_dir / (save_name+\".png\"), dpi=300)\n",
    "    plt.savefig(output_dir / (save_name+\".svg\"))\n",
    "    plt.show()\n",
    "    return ax\n",
    "\n",
    "num_and_denoms = [('adapted_near_tss', 'total_adapted'),\n",
    "                   ('total_near_tss', 'total_reads')]\n",
    "for num, denom in num_and_denoms:\n",
    "    for wildtype_usage in [True, False]:\n",
    "        plot_scatter_with_connections(meaned, groupby_genes, x_col, num, denom, only_wt=wildtype_usage)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "81b9294729ea0366",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Finalized bootstrap and means plot:"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5aa0c70d86592a8f"
  },
  {
   "cell_type": "code",
   "source": [
    "def bootstrap_row_optimized(index: tuple, row: pd.Series, num_samples: int, sample_size: int) -> pd.Series:\n",
    "\n",
    "    num_ada_tss = row['adapted_near_tss']\n",
    "    num_unada_tss = row['unadapted_near_tss']\n",
    "    num_ada_not_tss = row['total_adapted'] - row['adapted_near_tss']\n",
    "    num_unada_not_tss = row['total_unadapted'] - row['unadapted_near_tss']\n",
    "\n",
    "    set_up_array = np.array([\n",
    "        [1, 1],  # adapted, near TSS\n",
    "        [1, 0],  # adapted, not TSS\n",
    "        [0, 1],  # unadapted, near TSS\n",
    "        [0, 0],  # unadapted, not TSS\n",
    "    ])\n",
    "    bootstrap_array = np.repeat(set_up_array, [num_ada_tss, num_ada_not_tss, num_unada_tss, num_unada_not_tss], axis=0)\n",
    "\n",
    "    bootstrap_results = []\n",
    "    for cycle in range(num_samples):\n",
    "        sample_indices = np.random.choice(bootstrap_array.shape[0], size=sample_size, replace=True)\n",
    "        sample = bootstrap_array[sample_indices]\n",
    "        unique, counts = np.unique(sample, axis=0, return_counts=True)\n",
    "        bootstrap_results.append(dict(zip(map(tuple, unique), counts)))\n",
    "\n",
    "    col_conversion_dict = {\n",
    "        (1, 1): 'ada_near_tss',\n",
    "        (1, 0): 'ada_not_tss',\n",
    "        (0, 1): 'unada_near_tss',\n",
    "        (0, 0): 'unada_not_tss',\n",
    "    }\n",
    "\n",
    "    # Convert list of dicts to DataFrame\n",
    "    bootstrap_df = pd.DataFrame(bootstrap_results).fillna(0).astype(int)\n",
    "    bootstrap_df.columns = [col_conversion_dict[col] for col in bootstrap_df.columns]\n",
    "    \n",
    "    means = bootstrap_df.mean().rename(index)\n",
    "    return means\n",
    "\n",
    "def bootstrap_row_for_apply(row, bootstrap_cycles: int = 1000, sample_size: int = 10):\n",
    "    index = row.name\n",
    "    return bootstrap_row_optimized(index, row, num_samples=bootstrap_cycles, sample_size=sample_size)\n",
    "\n",
    "def run_box_bootstrapped(input_df: pd.DataFrame,\n",
    "                         target_num: str, target_denom: str,\n",
    "                         min_total_reads: int = 1,\n",
    "                         min_adapted_reads: int = -1,\n",
    "                         min_tss_ada_reads: int = -1,\n",
    "                         add_tss_ada_pseudo: float = -1.0,\n",
    "                         output_dir: Path = current_dir / 'plots' / 'decapping_bootstrapped',\n",
    "                         bootstrap_cycles: int = 1000,\n",
    "                         bootstrap_sample_size: int = 10,\n",
    "                         log_y: bool = False,\n",
    "                         log2_y: bool = False,\n",
    "                         template: str = 'none') -> None:\n",
    "    input_df = input_df.copy()\n",
    "    if min_total_reads > 0:\n",
    "        input_df = input_df.query(\"total_reads >= @min_total_reads\")\n",
    "    if min_adapted_reads > 0:\n",
    "        input_df = input_df.query(\"total_adapted >= @min_adapted_reads\")\n",
    "    if min_tss_ada_reads > 0:\n",
    "        input_df = input_df.query(\"adapted_near_tss >= @min_tss_ada_reads\")\n",
    "    \n",
    "    input_df['total_unadapted'] = input_df['total_reads'] - input_df['total_adapted']\n",
    "    \n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    tqdm.pandas(desc=\"Bootstrapping\")\n",
    "    boot_df = input_df.progress_apply(bootstrap_row_for_apply,\n",
    "                                      bootstrap_cycles=bootstrap_cycles,\n",
    "                                      sample_size=bootstrap_sample_size,\n",
    "                                      axis=1).fillna(0).astype(float)\n",
    "    boot_df.rename(columns={'ada_near_tss': 'adapted_near_tss',\n",
    "                            'ada_not_tss': 'ada_not_tss',\n",
    "                            'unada_near_tss': 'unadapted_near_tss',\n",
    "                            'unada_not_tss': 'unada_not_tss'},\n",
    "                   inplace=True)\n",
    "    # print(boot_df.columns)\n",
    "    \n",
    "    boot_df['adapted_near_tss'] = boot_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        boot_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    boot_df['unadapted_near_tss'] = boot_df['unadapted_near_tss'].fillna(0)\n",
    "    \n",
    "    boot_df['total_reads'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss'] + boot_df['unadapted_near_tss'] + boot_df['unada_not_tss']\n",
    "    boot_df['total_adapted'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss']\n",
    "    boot_df['total_near_tss'] = boot_df['adapted_near_tss'] + boot_df['unadapted_near_tss']\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    boot_df[target_col] = boot_df[target_num] / boot_df[target_denom]\n",
    "    \n",
    "    box_df = boot_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = (box_df['gene_name'] + ' ' +\n",
    "                      box_df['strain'].astype(str) +\n",
    "                      ' rep' + box_df['rep'].astype(str))\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>\"\n",
    "                       f\"Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}<br>\"\n",
    "                       f\"Bootstrapped {bootstrap_cycles} cycles of {bootstrap_sample_size} samples\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    \n",
    "    # # Can we add a line for the means?\n",
    "    # means = box_df.groupby(['strain', 'nmd_target_type']).mean().reset_index()\n",
    "    # fig.add_trace(go.Scatter(x=means['strain'], y=means[target_col], mode='lines+markers', marker=dict(size=10, color='black'), name='Mean'))\n",
    "    \n",
    "    fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=16,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    save_friendly_target_col += f\"_bootstrapped{bootstrap_cycles}cycleX{bootstrap_sample_size}samples\"\n",
    "    print(f\"Saving file to {output_dir / f'{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png/html'}\")\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "\n",
    "def run_box_mean_and_boot(input_df: pd.DataFrame,\n",
    "                         target_num: str, target_denom: str,\n",
    "                         min_total_reads: int = 1,\n",
    "                         min_adapted_reads: int = 1,\n",
    "                         min_tss_ada_reads: int = 0,\n",
    "                         add_tss_ada_pseudo: float = -1.0,\n",
    "                         output_dir: Path = current_dir / 'plots' / 'decapping_bootstrappedAndMeaned',\n",
    "                         bootstrap_cycles: int = 1000,\n",
    "                         bootstrap_sample_size: int = 10,\n",
    "                         min_number_reps_observed=1,\n",
    "                         log_y: bool = True,\n",
    "                         log2_y: bool = False,\n",
    "                         template: str = 'none') -> None:\n",
    "    input_df = input_df.copy()\n",
    "    input_df = input_df.query(\"adapted_near_tss >= @min_tss_ada_reads\")\n",
    "    input_df = input_df.query(\"total_reads >= @min_total_reads\")\n",
    "    input_df = input_df.query(\"total_adapted >= @min_adapted_reads\")\n",
    "    \n",
    "    input_df['total_unadapted'] = input_df['total_reads'] - input_df['total_adapted']\n",
    "    \n",
    "    \n",
    "    if not output_dir.exists():\n",
    "        output_dir.mkdir(parents=True)\n",
    "        print(f\"Made {output_dir}\")\n",
    "    \n",
    "    tqdm.pandas(desc=\"Bootstrapping\")\n",
    "    boot_df = input_df.query(\"total_adapted >= @min_adapted_reads\").progress_apply(bootstrap_row_for_apply, bootstrap_cycles=bootstrap_cycles, sample_size=bootstrap_sample_size, axis=1).fillna(0).astype(float)\n",
    "    boot_df.rename(columns={'ada_near_tss': 'adapted_near_tss',\n",
    "                            'ada_not_tss': 'ada_not_tss',\n",
    "                            'unada_near_tss': 'unadapted_near_tss',\n",
    "                            'unada_not_tss': 'unada_not_tss'},\n",
    "                        inplace=True)\n",
    "    print(boot_df.index.get_level_values('strain').unique())\n",
    "    boot_df['total_reads'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss'] + boot_df['unadapted_near_tss'] + boot_df['unada_not_tss']\n",
    "    boot_df['total_adapted'] = boot_df['adapted_near_tss'] + boot_df['ada_not_tss']\n",
    "    boot_df['total_near_tss'] = boot_df['adapted_near_tss'] + boot_df['unadapted_near_tss']\n",
    "    input_df = boot_df\n",
    "    \n",
    "    target_col = f\"{target_num} / {target_denom}\"\n",
    "    \n",
    "    input_df['adapted_near_tss'] = input_df['adapted_near_tss'].fillna(0)\n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        input_df['adapted_near_tss'] += add_tss_ada_pseudo\n",
    "    input_df['unadapted_near_tss'] = input_df['unadapted_near_tss'].fillna(0)\n",
    "    \n",
    "    input_df = process_df_for_means(input_df,\n",
    "                                    plot_num, plot_denum,\n",
    "                                    min_total_reads=-1,\n",
    "                                    min_adapted_reads=-1,\n",
    "                                    min_tss_ada_reads=-1,\n",
    "                                    add_tss_ada_pseudo=-1,\n",
    "                                    min_number_reps_observed=min_number_reps_observed,\n",
    "                                    log2_y=log2_y,\n",
    "                                    log_y=log_y,\n",
    "                                    )\n",
    "    print(input_df.columns)\n",
    "    input_df[target_col] = input_df[target_num] / input_df[target_denom]\n",
    "    \n",
    "    box_df = input_df.copy().reset_index()\n",
    "    \n",
    "    box_df['strain'] = pd.Categorical(box_df['strain'], ['wildtype', 'smg-5', 'smg-6', 'smg-7'])\n",
    "    box_df.sort_values(by=['strain', 'nmd_target_type'], inplace=True)\n",
    "    \n",
    "    box_df = box_df[box_df['strain'] != 'smg-7']\n",
    "    box_df['name'] = box_df['gene_name'] + ' ' + box_df['strain'].astype(str)\n",
    "    \n",
    "    if log2_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "        box_df[f\"log2({target_col})\"] = np.log2(box_df[target_col])\n",
    "        target_col = f\"log2({target_col})\"\n",
    "        log_y = False\n",
    "    \n",
    "    if log_y:\n",
    "        box_df = box_df[box_df[target_col] > 0]\n",
    "    \n",
    "    if min_tss_ada_reads > 0:\n",
    "        box_df = box_df[box_df['adapted_near_tss'] >= min_tss_ada_reads]\n",
    "    \n",
    "    # These colors are updated from the final plots!\n",
    "    plot_colors = (('#37c871', '#16502d'),  # NMD greens\n",
    "                   ('#64a7ff', '#00008b'),  # Non NMD blues\n",
    "                   ('#676767', '#414141'),  # ambiguous greys\n",
    "                   )\n",
    "    colors_dict = {'NMD_Targets': plot_colors[0][0],\n",
    "                   'NMD_NonTargets': plot_colors[1][0],}\n",
    "    \n",
    "    print(box_df.columns)\n",
    "    fig = px.box(box_df,\n",
    "                 x='strain',\n",
    "                 y=target_col,\n",
    "                 color='nmd_target_type',\n",
    "                 color_discrete_map=colors_dict,\n",
    "                 points='all',\n",
    "                 hover_name='name',\n",
    "                 log_y=log_y,\n",
    "                 hover_data=[target_num, target_denom, 'num_successful_reps', 'successful_reps'],\n",
    "                 template=template,\n",
    "                 title=f\"{target_col} by Strain<br>\"\n",
    "                       f\"Min Total Reads: {min_total_reads}, Min Adapted Reads: {min_adapted_reads}, Y Scale: {'Log' if log_y else 'Linear'}<br>\"\n",
    "                       f\"Bootstrapped {bootstrap_cycles} cycles of {bootstrap_sample_size} samples & Mean across {min_number_reps_observed}+ replicates\",\n",
    "                 height=500,\n",
    "                 width=1000,\n",
    "                 )\n",
    "    \n",
    "    # # Can we add a line for the means?\n",
    "    # means = box_df.groupby(['strain', 'nmd_target_type']).mean().reset_index()\n",
    "    # fig.add_trace(go.Scatter(x=means['strain'], y=means[target_col], mode='lines+markers', marker=dict(size=10, color='black'), name='Mean'))\n",
    "    \n",
    "    if target_denom == 'total_reads':\n",
    "        fig.update_yaxes(title_text=f'{target_num} / <br>{target_denom}')\n",
    "    else:\n",
    "        fig.update_yaxes(# range=[-0.1, 1.1],\n",
    "                         title_text=f'{target_num} / <br>{target_denom}')\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            # family=\"Courier New, monospace\",\n",
    "            size=16,\n",
    "            # color=\"RebeccaPurple\",\n",
    "        ),\n",
    "                 )\n",
    "    fig.update_yaxes(nticks=7, ticklabelposition=\"inside\")\n",
    "    save_friendly_target_col = target_col.replace(' ', '_').replace('/', '_vs_')\n",
    "    if log_y:\n",
    "        save_friendly_target_col += '_logY'\n",
    "    else:\n",
    "        save_friendly_target_col += '_linY'\n",
    "        \n",
    "    if add_tss_ada_pseudo > 0:\n",
    "        save_friendly_target_col += f\"_pseudoTSS{add_tss_ada_pseudo}\"\n",
    "    save_friendly_target_col += f\"_minTotal{min_total_reads}_minAdapted{min_adapted_reads}_minTSSAdapted{min_tss_ada_reads}\"\n",
    "    save_friendly_target_col += f\"_bootstrapped{bootstrap_cycles}cycleX{bootstrap_sample_size}samples\"\n",
    "    save_friendly_target_col += f\"_meaned{min_number_reps_observed}plusReps\"\n",
    "    print(f\"Saving file to {output_dir / f'{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png/html'}\")\n",
    "    output_dir.mkdir(exist_ok=True, parents=False)\n",
    "    fig.write_html(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.html\")\n",
    "    fig.write_image(output_dir / f\"{npCommon.get_dt(for_file=True)}_boxPlot_{save_friendly_target_col}.png\", scale=5)\n",
    "    fig.show(renderer='firefox')\n",
    "\n",
    "\n",
    "min_adapteds = [\n",
    "    # 1,\n",
    "    # 2,\n",
    "    3,\n",
    "    # 4,\n",
    "    # 5,\n",
    "    # 10,\n",
    "]\n",
    "\n",
    "target_numerators = [\n",
    "    'adapted_near_tss',\n",
    "    'total_near_tss',\n",
    "]\n",
    "\n",
    "target_denominators = [\n",
    "    'total_reads',\n",
    "    # 'unadapted_near_tss',\n",
    "    'total_adapted',\n",
    "    # 'total_unadapted',\n",
    "]\n",
    "\n",
    "target_num_denom_pairs = [\n",
    "    ('total_near_tss', 'total_reads'),\n",
    "    ('adapted_near_tss', 'total_adapted'),\n",
    "]\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "# for target_numerator in target_numerators:\n",
    "#     for target_denominator in target_denominators:\n",
    "for target_numerator, target_denominator in target_num_denom_pairs:\n",
    "        for min_adapted in min_adapteds:\n",
    "            print(f\"\\n\\nUsing {target_denominator=} for denominator... And {min_adapted=}...\\n\")\n",
    "            # run_box_mean_and_boot(\n",
    "            #     better_df,\n",
    "            #     target_numerator,\n",
    "            #     target_denominator,\n",
    "            #     min_adapted_reads=min_adapted,\n",
    "            #     min_tss_ada_reads=0,\n",
    "            #     bootstrap_cycles=100,\n",
    "            #     bootstrap_sample_size=10,\n",
    "            #     min_number_reps_observed=2,\n",
    "            #     log_y=True,\n",
    "            #     add_tss_ada_pseudo=0.001)\n",
    "            for repeat in range(repeats):\n",
    "                run_box_bootstrapped(\n",
    "                    better_df,\n",
    "                    target_numerator,\n",
    "                    target_denominator,\n",
    "                    min_adapted_reads=min_adapted,\n",
    "                    min_tss_ada_reads=-1,\n",
    "                    bootstrap_cycles=1,\n",
    "                    bootstrap_sample_size=10,\n",
    "                    # log_y=True,\n",
    "                    # add_tss_ada_pseudo=0.001,\n",
    "                )\n",
    "print(\"Done!\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f082c628be07e391",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "from scipy.optimize import curve_fit\n",
    "\n",
    "def arctan_func(x: float, a: float, b: float, z: float, w: float) -> float:\n",
    "    # z is around 1/3 or 1/2 and w is between 1/2 and 3/5? maybe\n",
    "    # a is likely near 1/1000, and b is 1000 or more\n",
    "    return z * np.arctan(a * (-x + b)) + w\n",
    "\n",
    "def bounded_arctan_func(x: float, a: float, b: float, z: float, w: float) -> float:\n",
    "    return z * np.arctan(a * (-x + b)) + w\n",
    "\n",
    "data_x = groupby_genes.query(\"num_observed_reps >= 2\").query(\"total_adapted >= 1\").query(\"nmd_active == False\")['avg_transcript_length']\n",
    "data_y = groupby_genes.query(\"num_observed_reps >= 2\").query(\"total_adapted >= 1\").query(\"nmd_active == False\")['total_near_tss / total_reads']\n",
    "\n",
    "\n",
    "# Remove NaN and infinite values\n",
    "data_x = data_x[np.isfinite(data_x)]\n",
    "data_y = data_y[np.isfinite(data_y)]\n",
    "\n",
    "# Initial guesses for the parameters\n",
    "initial_guesses = [1/1000, 1000, 1/3, 1/2]\n",
    "\n",
    "# Bounds for the parameters\n",
    "param_bounds = (0, [np.inf, np.inf, 1, 1])\n",
    "\n",
    "# Fit the arctan function to the data\n",
    "popt, pcov = curve_fit(bounded_arctan_func, data_x, data_y, p0=initial_guesses, bounds=param_bounds)\n",
    "\n",
    "# Print the fitted parameters for debugging\n",
    "print(\"Fitted parameters:\", popt)\n",
    "\n",
    "# Generate x values for the fitted curve\n",
    "x_fit = np.linspace(min(data_x), max(data_x), 500)\n",
    "y_fit = bounded_arctan_func(x_fit, *popt)\n",
    "\n",
    "# Plot the data and the fitted curve\n",
    "fig, ax = plt.subplots(figsize=(10, 6))\n",
    "ax.scatter(data_x, data_y, alpha=0.4, label='Data')\n",
    "ax.plot(x_fit, y_fit, color='red', linestyle='--', label='Arctan Fit')\n",
    "plt.xscale('log')\n",
    "ax.set_xlabel('Transcript Length')\n",
    "ax.set_ylabel('Total Near TSS / Total Reads')\n",
    "ax.set_title('Arctan Fit to Data')\n",
    "ax.legend()\n",
    "plt.show()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "9e47fc15d5a450cd",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "# TODO: What about plotting the % adapted for each gene across from the % decapped?\n",
    "#       Idea here is that the % decapped gets towards HOW RNAs are being degraded, while the % adapted is more about HOW MUCH the RNA is being degraded\n",
    "#       This could be a good way to show both (that NMD targets are more likely to be degraded by non-decapping means in WT while they have the same rate for decay + decapping in the mutants)\n",
    "plot_df = meaned.copy()\n",
    "plot_df['adapted_near_tss / total_adapted'] = plot_df['adapted_near_tss'] / plot_df['total_adapted']\n",
    "plot_df['total_adapted / total_reads'] = plot_df['total_adapted'] / plot_df['total_reads']\n",
    "plot_df['label'] = plot_df['gene_name'] + ' ' + plot_df['strain'].astype(str) + ' ' + plot_df['nmd_target_type']\n",
    "plot_df['color'] = plot_df['strain'].astype(str) + '' + plot_df['nmd_target_type']\n",
    "fig = px.scatter(plot_df.query(\"total_adapted >= 2\"),\n",
    "                 x='total_adapted / total_reads',\n",
    "                 y='adapted_near_tss / total_adapted',\n",
    "                 # color='strain', symbol='nmd_target_type',\n",
    "                 symbol='strain', color='nmd_target_type',\n",
    "                 symbol_map={'NMD_Targets': 'x',\n",
    "                             'NMD_NonTargets': 'circle',\n",
    "                             'NMD_Ambiguous': 'diamond',\n",
    "                             'wildtype': 'circle',\n",
    "                             'smg-5': 'x',\n",
    "                             'smg-6': 'cross',},\n",
    "                 hover_name='label',\n",
    "                 # hover_data=['total_adapted', 'total_reads', 'observed_reps', 'num_observed_reps'],\n",
    "                 template='none',\n",
    "                 title='Decapping vs Adaptation',\n",
    "                 log_x=True, log_y=True\n",
    "                 )\n",
    "fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "a48134997ade8222",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "all_genes_df = pd.read_parquet(current_dir / 'preprocessed_data' / 'all_genes_avgTranscriptLength_and_nearTSS.parquet')\n",
    "all_genes_df = all_genes_df.query(\"total_reads >= 10\")\n",
    "meaned_all_genes = process_df_for_means(all_genes_df, 'total_adapted', 'total_reads',\n",
    "                                        min_total_reads=-1, min_adapted_reads=-1, min_tss_ada_reads=-1,\n",
    "                                        add_tss_ada_pseudo=0.001,\n",
    "                                        min_number_reps_observed=2,\n",
    "                                        log_y=False, log2_y=False)\n",
    "meaned_all_genes.head()"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "96da91cea5c42f5d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plot_df = pd.concat([meaned_all_genes.query(\"total_adapted >= 2\").query(\"total_reads >= 1\").query(\"avg_transcript_length <= 850\").copy(), meaned.copy()])\n",
    "plot_df.query(\"gene_name == 'ubl-1'\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "a595ef0eece8f210",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "plot_df['adapted_near_tss / total_adapted'] = plot_df['adapted_near_tss'] / plot_df['total_adapted']\n",
    "plot_df['total_adapted / total_reads'] = plot_df['total_adapted'] / plot_df['total_reads']\n",
    "plot_df['label'] = plot_df['gene_name'] + ' ' + plot_df['strain'].astype(str) + ' ' + plot_df['nmd_target_type']\n",
    "plot_df['color'] = plot_df['strain'].astype(str) + '' + plot_df['nmd_target_type']\n",
    "fig = px.scatter(plot_df,\n",
    "                 x='total_adapted / total_reads',\n",
    "                 y='adapted_near_tss / total_adapted',\n",
    "                 # color='strain', symbol='nmd_target_type',\n",
    "                 symbol='strain', color='nmd_target_type',\n",
    "                 symbol_map={'NMD_Targets': 'x',\n",
    "                             'NMD_NonTargets': 'circle',\n",
    "                             'NMD_Ambiguous': 'diamond',\n",
    "                             'wildtype': 'circle',\n",
    "                             'smg-5': 'x',\n",
    "                             'smg-6': 'cross',},\n",
    "                 hover_name='label',\n",
    "                 hover_data=['total_adapted', 'total_reads', 'avg_transcript_length'],\n",
    "                 template='none',\n",
    "                 title='Decapping vs Adaptation',\n",
    "                 log_x=True, log_y=True\n",
    "                 )\n",
    "fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6943157287bda55e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [
    "fig = px.scatter(plot_df,\n",
    "                 x='total_adapted / total_reads',\n",
    "                 y='adapted_near_tss / total_adapted',\n",
    "                 # color='strain', symbol='nmd_target_type',\n",
    "                 symbol='strain', color='nmd_target_type',\n",
    "                 symbol_map={'NMD_Targets': 'x',\n",
    "                             'NMD_NonTargets': 'circle',\n",
    "                             'NMD_Ambiguous': 'diamond',\n",
    "                             'wildtype': 'circle',\n",
    "                             'smg-5': 'x',\n",
    "                             'smg-6': 'cross',},\n",
    "                 hover_name='label',\n",
    "                 hover_data=['total_adapted', 'total_reads', 'avg_transcript_length'],\n",
    "                 template='none',\n",
    "                 title='Decapping vs Adaptation',\n",
    "                 log_x=True, log_y=True\n",
    "                 )\n",
    "plot_df = plot_df[[col for col in plot_df.columns if '_std' not in col]]\n",
    "for gene in plot_df.gene_name.unique():\n",
    "    gene_rows = plot_df.query(f\"gene_name == '{gene}'\")\n",
    "    for nmd_targ_type in gene_rows.nmd_target_type.unique():\n",
    "        if nmd_targ_type == 'indifferent':\n",
    "            # continue\n",
    "            pass\n",
    "        nmd_rows = gene_rows.query(f\"nmd_target_type == '{nmd_targ_type}'\")\n",
    "        if not 'wildtype' in nmd_rows.strain.unique():\n",
    "            continue\n",
    "        # print(nmd_rows[['gene_name', 'nmd_target_type', 'strain', 'total_adapted / total_reads', 'adapted_near_tss / total_adapted']])\n",
    "        wt_x = nmd_rows.query(\"strain == 'wildtype'\")['total_adapted / total_reads'].values[0]\n",
    "        wt_y = nmd_rows.query(\"strain == 'wildtype'\")['adapted_near_tss / total_adapted'].values[0]\n",
    "        for smg in ['smg-5', 'smg-6']:\n",
    "            if smg in nmd_rows.strain.unique():\n",
    "                smg_x = nmd_rows.query(f\"strain == '{smg}'\")['total_adapted / total_reads'].values[0]\n",
    "                smg_y = nmd_rows.query(f\"strain == '{smg}'\")['adapted_near_tss / total_adapted'].values[0]\n",
    "                # fig.add_trace(go.Scatter(x=[wt_x, smg_x], y=[wt_y, smg_y], mode='lines',\n",
    "                #                          line=dict(color='black', width=2, dash='dash'),\n",
    "                #                          opacity=0.5,\n",
    "                #                          name=None,\n",
    "                #                          showlegend=False,\n",
    "                #                          ))\n",
    "                fig.add_trace(go.Scatter(x=[wt_x, smg_x], y=[wt_y, smg_y],\n",
    "                                         marker= dict(size=10,symbol= \"arrow-bar-up\", angleref=\"previous\", color='black'),\n",
    "                                         line=dict(color='black', width=2, dash='dash'),\n",
    "                                         opacity=0.5,\n",
    "                                         name=None,\n",
    "                                         showlegend=False,\n",
    "                                         ))\n",
    "fig.show(renderer='firefox')"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6f7182c0d23f5e74",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "430fbd7caf215cc0",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
