{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# simplerReadPlotting.ipynb\n",
    "## Marcus Viscardi,    July 17, 2023\n",
    "\n",
    "I want to quickly take some tricks I learned from fourLibrary5TERA/ambiguousReads.py and apply them here.\n",
    "\n",
    "The goal is to be able to plot reads from a locus based on their overlap of that locus and directionality. This will be a little messier, but important for all those genes that have a had time getting reads assigned to them! (*ets-4*)\n",
    "\n",
    "Plan is to feed this a **target_gene** and a **bam file** then use pysam to rip all the relevant reads from the bam, and plot those!"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import pysam\n",
    "import sys\n",
    "from tqdm import tqdm\n",
    "sys.path.insert(0, '/data16/marcus/scripts/nanoporePipelineScripts')\n",
    "import nanoporePipelineCommon as npC\n",
    "from finalizingReadAndCoveragePlotting_matplotlib import plot_reads, coverage_plotting_5tera\n",
    "\n",
    "MIN_COVERAGE_OF_GENE = 50\n",
    "print(f\"Imports Complete @ {npC.get_dt(for_print=True)}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "def create_cutdown_bam(target_genes: list, target_lib:str, save_dir='/tmp') -> (Path, pd.DataFrame):\n",
    "    bam_file = Path(npC.pick_lib_return_path(npC.REV_CONVERSION_DICT[target_lib], output_dir_folder='cat_files', file_midfix='cat.sorted.mappedAndPrimary', file_suffix='.bam'))\n",
    "    gtf_df = pd.read_parquet('/data16/marcus/genomes/'\n",
    "                         'plus_cerENO2_elegansRelease100/'\n",
    "                         '230327_allChrs_plus-cerENO2.gtf.parquet')\n",
    "    short_gtf_df = gtf_df[(gtf_df['feature'] == 'gene') & (gtf_df['gene_name'].isin(target_genes))].copy()\n",
    "    bam = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    cutdown_bam_path = f\"{save_dir}/{npC.get_dt()}_{'_'.join(target_genes)}_{target_lib}.bam\"\n",
    "    cutdown_bam = pysam.AlignmentFile(cutdown_bam_path, 'wb',\n",
    "                                      template=bam)\n",
    "    for gene_id, gene_name in zip(short_gtf_df['gene_id'], short_gtf_df['gene_name']):\n",
    "        print(f\"Working on {gene_name} @ {npC.get_dt(for_print=True)}\")\n",
    "        chromosome, start, end, strand = \\\n",
    "        short_gtf_df.loc[short_gtf_df['gene_id'] == gene_id, ['chr', 'start', 'end', 'strand', ]].values[0]\n",
    "        gene_is_reverse = (strand == '-')\n",
    "        gene_is_forward = not gene_is_reverse\n",
    "        read_iterator = tqdm(enumerate(bam.fetch(chromosome, start, end)))\n",
    "        for i, read in read_iterator:\n",
    "            if read.get_overlap(start, end) <= MIN_COVERAGE_OF_GENE:\n",
    "                continue  # This will catch reads that are not in the gene! (50 is arbitrary)\n",
    "                #           This is mainly because some reads \"span\" the gene, so they get fetched!\n",
    "            if read.is_reverse != gene_is_reverse:\n",
    "                continue  # This will catch reads that are in the wrong direction!\n",
    "            read.set_tag('gA', gene_id, value_type='Z')\n",
    "            read.set_tag('gN', gene_name, value_type='Z')\n",
    "            read.set_tag('rS', read.reference_start, value_type='i')\n",
    "            read.set_tag('rE', read.reference_end, value_type='i')\n",
    "            read.set_tag('qL', read.query_length, value_type='i')\n",
    "            cutdown_bam.write(read)\n",
    "    for loop_bam in [cutdown_bam]:\n",
    "        loop_bam.close()\n",
    "        pysam.sort(loop_bam.filename, '-o', loop_bam.filename)\n",
    "        pysam.index(loop_bam.filename)  # These samtools methods aren't in the pysam __init__ file, but they work!\n",
    "    bam.close()\n",
    "    print(f\"Done. Saved to: {cutdown_bam_path}\")\n",
    "    return Path(cutdown_bam_path), short_gtf_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "target_genes = (\n",
    "    # 'ets-4',\n",
    "    # 'zip-1',\n",
    "    'xbp-1',\n",
    "    'rps-22',\n",
    ")\n",
    "target_lib = 'newerS6'\n",
    "\n",
    "\n",
    "create_cutdown_bam(target_genes, target_lib)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "def bam_tag_to_df(bam_object_or_path, tags=(), chr = None, start=None, stop = None):\n",
    "    if not isinstance(bam_object_or_path, pysam.AlignmentFile):\n",
    "        bam = pysam.AlignmentFile(bam_object_or_path, 'rb')\n",
    "    else:\n",
    "        bam = bam_object_or_path\n",
    "    read_names = []\n",
    "    chrs = []\n",
    "    chr_starts = []\n",
    "    strands = []\n",
    "    cigars = []\n",
    "    tags_dict = {tag: [] for tag in tags}\n",
    "    for read in bam.fetch(chr, start, stop):\n",
    "        read_names.append(read.query_name)\n",
    "        chrs.append(read.reference_name)\n",
    "        chr_starts.append(read.reference_start)\n",
    "        strands.append(read.is_reverse)\n",
    "        cigars.append(read.cigarstring)\n",
    "        for add_tag in tags:\n",
    "                tags_dict[add_tag].append(read.get_tag(add_tag))\n",
    "    df = pd.DataFrame({'read_id': read_names,\n",
    "                       'chr_id': chrs,\n",
    "                       'chr_pos': chr_starts,\n",
    "                       'strand': strands,\n",
    "                       'cigar': cigars,\n",
    "                       **tags_dict})\n",
    "    df.strand.replace({True: '-', False: '+'}, inplace=True)\n",
    "    return df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "quick_df = bam_tag_to_df(pysam.AlignmentFile(f\"/tmp/{npC.get_dt()}_{'_'.join(target_genes)}_{target_lib}.bam\", 'rb'), tags=['gA', 'gN', 't5', 'rS', 'rE']).rename(columns={'gA': 'gene_id', 'gN': 'gene_name'})\n",
    "quick_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.patches import Rectangle\n",
    "import re\n",
    "get_dt = npC.get_dt\n",
    "\n",
    "\n",
    "def _make_rectangle_patch(genome_start, length, y_center, thickness, color='gray'):\n",
    "    return Rectangle((genome_start, y_center - (thickness / 2)), length, thickness,\n",
    "                     facecolor=color,\n",
    "                     edgecolor=color,\n",
    "                     fill=True,\n",
    "                     lw=0)\n",
    "\n",
    "\n",
    "def _add_patches_from_cigars_and_gen_pos(axes, cigar, gen_start, y, color='black', plot_introns=True):\n",
    "    # Parse the cigar string\n",
    "    parsed_cigar = re.findall(rf'(\\d+)([MDNSIX])', cigar)\n",
    "    mdn_nums = [int(num) for num, char in parsed_cigar if char in \"MDN\"]\n",
    "    gen_end = gen_start + sum(mdn_nums)\n",
    "    genomic_read_length = gen_end - gen_start\n",
    "\n",
    "    genome_loc = gen_start\n",
    "\n",
    "    rectangle_patch_list = []\n",
    "    first_n_length = 0\n",
    "    for length, code in parsed_cigar:\n",
    "        length = int(length)\n",
    "        if code == 'S':\n",
    "            pass\n",
    "        elif code == 'M':\n",
    "            rectangle_patch_list.append(_make_rectangle_patch(genome_loc, length, y, thickness=0.8))\n",
    "            genome_loc += length\n",
    "        elif code == 'D':\n",
    "            if length < 50:\n",
    "                rectangle_patch_list.append(_make_rectangle_patch(genome_loc, length, y, thickness=0.8))\n",
    "            else:\n",
    "                if plot_introns:\n",
    "                    rectangle_patch_list.append(_make_rectangle_patch(genome_loc, length, y, thickness=0.001))\n",
    "            genome_loc += length\n",
    "        elif code == 'I':\n",
    "            pass\n",
    "        elif code == 'N':\n",
    "            if plot_introns:\n",
    "                rectangle_patch_list.append(_make_rectangle_patch(genome_loc, length, y, thickness=0.001))\n",
    "            genome_loc += length\n",
    "    axes.add_collection(PatchCollection(rectangle_patch_list, color=color))\n",
    "    return genomic_read_length\n",
    "\n",
    "\n",
    "def _add_patches_from_cigars_and_gen_pos_gtp(axes, cigar, gen_start, y, color='black', plot_introns=True):\n",
    "    # Parse the cigar string\n",
    "    parsed_cigar = re.findall(rf'(\\d+)([MDNSIX])', cigar)\n",
    "    mdn_nums = [int(num) for num, char in parsed_cigar if char in \"MDN\"]\n",
    "    gen_end = gen_start + sum(mdn_nums)\n",
    "    genomic_read_length = gen_end - gen_start\n",
    "\n",
    "    genome_loc = gen_start\n",
    "\n",
    "    rectangle_patch_list = []\n",
    "    current_length = 0\n",
    "    for length, code in parsed_cigar:\n",
    "        length = int(length)\n",
    "        if code == 'S':\n",
    "            pass\n",
    "        elif code == 'M':\n",
    "            current_length += length\n",
    "            genome_loc += length\n",
    "        elif code == 'D':\n",
    "            current_length += length\n",
    "            genome_loc += length\n",
    "        elif code == 'I':\n",
    "            pass\n",
    "        elif code == 'N':\n",
    "            if plot_introns:\n",
    "                # Only add the accumulated length as a single rectangle if we encountered matches or deletes\n",
    "                if current_length > 0:\n",
    "                    rectangle_patch_list.append(_make_rectangle_patch(genome_loc - current_length, current_length, y, thickness=0.8))\n",
    "                    current_length = 0\n",
    "                rectangle_patch_list.append(_make_rectangle_patch(genome_loc, length, y, thickness=0.001))\n",
    "            genome_loc += length\n",
    "\n",
    "    # Add the accumulated length as a single rectangle if we reached the end of the cigar\n",
    "    if current_length > 0:\n",
    "        rectangle_patch_list.append(_make_rectangle_patch(genome_loc - current_length, current_length, y, thickness=0.8))\n",
    "\n",
    "    axes.add_collection(PatchCollection(rectangle_patch_list, color=color))\n",
    "\n",
    "    return genomic_read_length\n",
    "\n",
    "def subset_df(df, column, fractions):\n",
    "    subset_df = df.groupby(column).apply(lambda x: x.sample(frac=fractions.get(x.name, 1))).reset_index(drop=True)\n",
    "    return subset_df\n",
    "\n",
    "def _row_apply_plot_cigar(row, axes, plot_introns=True, spacing_factor=1.0):\n",
    "    index = row.name * spacing_factor\n",
    "    cigar = row.cigar\n",
    "    gen_start = row.ref_start\n",
    "    is_adapted = row.t5\n",
    "\n",
    "    if is_adapted == '-':\n",
    "        color = 'black'\n",
    "    else:\n",
    "        color = 'red'\n",
    "    return _add_patches_from_cigars_and_gen_pos_gtp(axes, cigar, gen_start, index,\n",
    "                                                    color=color,\n",
    "                                                    plot_introns=plot_introns)\n",
    "\n",
    "def new_plot_reads_from_bam(target_lib_nickname,\n",
    "                            target_gene_name,\n",
    "                            subset_fraction={'+': 1, '-': 1},\n",
    "                            spacing_factor=1.0,\n",
    "                            save_dir=None,\n",
    "                            ):\n",
    "    # First lets make the tmp BAM file that only has the gene of interest:\n",
    "    tmp_bam_path, short_gtf_df = create_cutdown_bam([target_gene_name], target_lib_nickname)\n",
    "    bam_df = bam_tag_to_df(tmp_bam_path, tags=['gA', 'gN', 't5', 'rS', 'rE', 'qL'])\n",
    "    bam_df.rename(columns={'gA': 'gene_name',\n",
    "                           'gN': 'gene_id',\n",
    "                           't5': 't5',\n",
    "                           'rS': 'ref_start',\n",
    "                           'rE': 'ref_end',\n",
    "                           'qL': 'read_length'},\n",
    "                  inplace=True)\n",
    "    \n",
    "    bam_df = subset_df(bam_df, \"t5\", subset_fraction)\n",
    "    read_count = bam_df.shape[0]\n",
    "    print(f\"{read_count:>6,} reads in {target_lib_nickname} that match {target_gene_name}\")\n",
    "    \n",
    "    # Get some gene info:\n",
    "    gene_id, chromosome, genomic_start, genomic_end, gene_strand = short_gtf_df.loc[short_gtf_df['gene_name'] == target_gene_name, ['gene_id', 'chr', 'start', 'end', 'strand']].values[0]\n",
    "    # Now lets plot the reads:\n",
    "    fig, ax = plt.subplots(figsize=(5, (read_count * 0.08 * spacing_factor)))\n",
    "    \n",
    "    if gene_strand == \"-\":\n",
    "        sort_order = [\"t5\", \"ref_end\", \"ref_start\", \"read_length\"]\n",
    "        sort_order_ascending = [False, True, False, False]\n",
    "    else:  # gene_strand == \"+\":\n",
    "        sort_order = [\"t5\", \"ref_start\", \"ref_end\", \"read_length\"]\n",
    "        sort_order_ascending = [False, False, False, False]\n",
    "    \n",
    "    bam_df = bam_df.sort_values(sort_order, ascending=sort_order_ascending).reset_index(drop=True)\n",
    "    tqdm.pandas(desc=\"Plotting Reads...\")\n",
    "    bam_df.progress_apply(lambda row: _row_apply_plot_cigar(row, ax, spacing_factor=spacing_factor), axis=1)\n",
    "\n",
    "    ax.set_xlim(genomic_start, genomic_end)\n",
    "    \n",
    "    number_of_plotted_reads = bam_df.shape[0]\n",
    "    ax.set_ylim(-1, number_of_plotted_reads + 1)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    plt.tight_layout()\n",
    "    if save_dir is not None:\n",
    "        save_path = f\"{save_dir}/{get_dt()}_{target_lib_nickname}_{target_gene_name}.readPlotting\"\n",
    "        for suffix in [\"png\", \"svg\"]:\n",
    "            plt.savefig(f\"{save_path}.{suffix}\", dpi=300)\n",
    "        print(f\"Saved images to: {save_path}.png/svg\")\n",
    "    plt.show()\n",
    "    return bam_df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-01-25T16:59:56.022189647Z",
     "start_time": "2024-01-25T16:59:55.685169112Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for file for 5tera_xrn-1-KD_wt_rerun, at /data16/marcus/working/230327_nanoporeRun_totalRNA_wt_xrn-1-KD_5TERA_rerun/output_dir/cat_files/*cat.sorted.mappedAndPrimary.bam... File Found.\n",
      "Working on rps-22 @ 01/25/24 @ 09:00:20 AM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1478it [00:00, 25051.95it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Saved to: /tmp/240125_rps-22_newerN2.bam\n",
      "    36 reads in newerN2 that match rps-22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Reads...: 100%|██████████| 36/36 [00:00<00:00, 1888.24it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved images to: /home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python/240125_readPlots/rps-22/240125_newerN2_rps-22.readPlotting.png/svg\n"
     ]
    },
    {
     "data": {
      "text/plain": "<Figure size 360x248.832 with 1 Axes>",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAADxCAYAAAD1LG0eAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAIC0lEQVR4nO3dQW7jOBoFYKkwW2dVvZ5cIcj9b1BBzlDrnlXVPp5lo2lZtrst/Xzi9y2FAsIKqSeGerDn8/k8AbC/b9UDABiVAAYoIoABighggCICGKCIAAYo8p9H/vH379/Pr6+vGw2FQ/r4qB7Buvf3/X5W778LNvMxTf87n89/tNcfCuDX19fpx48fzxsVx/fyMk2/f1ePYtnpNE17rueefxdsap6mn0vXHwpgeNivX9Uj6IffxbjmefGyAKYfe+8QT6d9Q9EOmIaXcPRj73A6+s+je3bAjO3Kn4awBwHM2Pb8MCphT8MRBEARAUw/Tic/j6E4gqAfR69pHf3/x3VqaETbosKlhsbWbqwxRxBk2CK41NDY2o05F8AARRxBMDbVMAoJYMamB0whRxBk2KLCpYbG1m7MuR0wGY5Q4TrC/4GnEsCwBxW0ca3sgh1BwB6E77hW5l4AAxQRwABFBDBAEQEMe1BBG9fK3GtBwB5U0Mbm09CgM6ppY1BDgw4J3zGooQH0RwADFBHAAEUEMFRRTRuDGhp0SDVtHGpo0CFVtCG8TdPb0nVHEFBJ+A7h25WsFcAARQQwQBEBDFBEAEMlVbQhfE3T19J1LQiopIo2hM95/ly6LoChkhraMZ1Odz1cHUFAJeF7THfOqwAGKCKAAYoIYIAiAhgqqaEd053zqgUBldTQhiaAoTeqaYfzPk3vS9cdQUBvhO8wBDBAEQEMUEQAAxQRwNAb1bRhaEFAb1TTDudjnj+WrtsBAxSxA4Ye6QIfim9FhiTC91B8KzJAZwQwQBEBDD1SRTsUX8oJSVTRDuXal3LaAQMUsQOGnqmj5Vs5TrIDhp4J33wrcyiAAYoIYIAiAhh6po6Wb2UOvYSDnqmjHcM8L162AwYoYgcMPVNDOwSfhgaJhO8h+DQ0gM4IYIAiAhh6poZ2CD4NDRKpoR2CT0MD6IwdMPRMDe0Q1NAgkfA9BDU0gM4IYIAiAhh6poZ2CGpokEgN7RDU0AA6YwcMCdTRoqmhQTLhG00NDaAzAhigiAAGKCKAIYE+cDQ9YEimDxztWg9YAEMKVbRMK3+9OIKAFMI308q8CWCAIgIYoIgABigigCGFKlqmlXnTgoAUqmi55nnxsgCGNOpoWdTQ4ECEbxY1NID+CGCAIgIYoIgAhjTqaFnU0OBA1NHyqKHBAFTU6p1Odz8kHUHAkQjfeg/MgQAGKCKAAYoIYIAiAhiOREWt3gNzoAUBR6KiFkUAA8+hAnfV2zS9LV13BAE8h/C96tuVrBXAAEUEMEARAQxQRAADz6ECd9XXNH0tXdeCAJ5DBe6qz3n+XLougIF+HaTapoYG5DlA+E6TGhpAdwQwQBEBDFBEAAP9Oki1TQ0NyHOQapsa2p1eXl6m3yFvXk+n0/TrIAt0BElri+e4dY86gmgk3SBJY8V8jejWnAtggCKOIMLN81w9BOAfEsDhzudz9RC4k4clLUcQjVNQ7SVprJivEd2aczvghlYBW7G2aAngRlJVSA0tS9LaYh+OIBpJN0jSWDFfXBLAAEUcQYTzZh1yCeBwamg5PCxpOYJoJFWFksaK+eKSHXBDq4CtWFvjuvbXjwBuJFWF1NCyJK0t/r177k9HEI2kGyRprJiv0dwz3wIYoIgjiHDerEMuARxODS2HhyUtRxCNpKpQ0lgxX6O5Z77tgBtaBWzF2qIlgBtJVSE1tCxJa4vnWdsJO4JoJN0gSWPFfI1qbd4FMEARRxDhvFmHXAI4nBpaDg9LWo4gGklVoaSxYr5GtTbvdsANrQK2Ym2Ny6eh3SmpKqSGliVpbbEPRxCNpBskaayYLy4JYIAijiDCebMOuQRwODW0HB6WtBxBNJKqQkljxXxxyQ64oVXAVqytcV3768cOGKCIHXAjqaupB5wlaW2xDzvgRtINkjRWzBeX7IDDebMOuQRwODW0HB6WtBxBNJKqQkljxXxxyQ644aUWW7G2xqWGBtAZO+BGUlVIDS1L0tpiH3bAjaQbJGmsmC8u2QGH82YdcgngcGpoOTwsaTmCaCRVhZLGivnikh1ww0sttmJtjUsNDaAzdsBXJFSG1NCyJKwp9mUHfEXCjZIwRv5ivmjZAYfzZh1yCeBwamg5PCxpOYK4IqEylDBG/mK+aNkBX+HlFs9mTY1LDQ2gM3bAK3qvDamhZel9PbGNtaMnO+AVvd8svY+PvzNfY1qbdzvgcN6sQy4BHE4NLYeHJS1HECt6rw31Pj7+znyNaW3e7YBXeMHFM1lP41JDA+iMHfANPVeH1NCy9LyWqGEHfEPPN0zPY+OS+aJlBxzOm3XIJYDDqaHl8LCk5QgCoIgAvqHn7mbPY+OS+aLlCOIGLQOexVoa17XjJwHcoXvrSmpoWdTQaDmC6NC9N6mbOYv5omUHHM6bdcglgMOpoeXwsKTlCAKgiADu0L11JbWmLOaLliOIDmk2HJN5HZcaGhRTQxva+9JFRxCwE+FLSwADFBHAAEUEMEARAQw7UUOjpQUBO1FDG9c8zx9L1wUwFFJNG8bb0kVHEFBI+A5jMWsFMEARAQxQRAADFBHAUEg1bRhfSxe1IKCQatoY5nn+XLougGFganDP98iX5TqCgIEJ3+d75HcqgAGKCGCAIgIYoIgAhoGpwT3fI79TLQgYmBpcLTtggCLz+Xy+/x/P85/TNP3cbjgAh/Tf8/n8R3vxoQAG4HkcQQAUEcAARQQwQBEBDFBEAAMUEcAARQQwQBEBDFBEAAMU+T+5jzQS+PB+hQAAAABJRU5ErkJggg==\n"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for file for 5tera_xrn-1-KD_smg-6_rerun, at /data16/marcus/working/230403_nanoporeRun_totalRNA_smg-6_xrn-1-KD_5TERA_rerun/output_dir/cat_files/*cat.sorted.mappedAndPrimary.bam... File Found.\n",
      "Working on rps-22 @ 01/25/24 @ 09:00:20 AM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3918it [00:00, 22207.98it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Saved to: /tmp/240125_rps-22_newerS6.bam\n",
      "    58 reads in newerS6 that match rps-22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Reads...: 100%|██████████| 58/58 [00:00<00:00, 1621.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved images to: /home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python/240125_readPlots/rps-22/240125_newerS6_rps-22.readPlotting.png/svg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": "<Figure size 360x400.896 with 1 Axes>",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAGJCAYAAAC0KukGAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPQ0lEQVR4nO3dvbLbRhIGUNK16VVkx6tXcFnv/wRW+Rk23o2s3NzwloYE+DMAej7OOaHW6xrMAM1W8zN4vlwuJwCO90v1AgBmpQADFFGAAYoowABFFGCAIgowQJF/PfMP//rrr5evX7/utBRu+v59/X//4499//3vrnf/njH7Xk/s++n0v8vl8lv75+dncsDfvn27/Pnnn5sujDvO5+oVAJ3Op9P3y+Xyrf3zpzpgBtT7H9LMXuCP/A+RZt9rrpgBj+7j47X/bYt//7s7+tpn3mtu0gGP7u+/s//9fLLX81r4248OGKCIDnh2X76cTj9+VK+ixsfHsV3pzHvNTTrg2c1cEI6+9pn3mpt0wMxNMoFCCjBzE0OjkBEEQBEFeHYzZ1PlgClmBDE72dTj2Ot5LYyfFGDmjUeJobGXB+8tIwjmLQpiaOzlwbNWgAGKGEEwN9EwCinAzE0OmEJGEMwbjxJDYy8PnrUOGPGoo9hnGjpggCI6YPaRkHmVA6aYDph9JBQaOWCK6YCZm2QChRRg5iaGRiEjCIAiCjD7SMi8ygFTzAiCfci8XrMn8/I6SoYyQiRLDI293bnHjCCoMUIhEkNjb3fOXAEGKGIEwdxEwyikADM3OWAKGUFQY4RIlhgae7tz5jpgaswYyZrxmlmlAwYoogOmXlU+Vg6YI6yMIXTA1KsqSnLAHGHl3HXAzE0ygUIKMHMTQ6OQEQRAEQWYelX5WDlgjrBy7kYQ1JslHzvLdXLN6ygZ3tExLTE0ihlBMI53j4UpvjQUYIAiRhDMTTSMQgowc5MDppARBON491iYGBoNHTDjePeY1rtfH8sW/vajAwYoogOGI8gAz8vrKKGY4juvlbNXgAGKKMAARRRggCIKMBxBBnheXkcJxWSA5+Z1lPBGxNqi/H46/X7rz40gIJHiG+WXhVqrAAMUUYABiijAAEUUYEgk1hbln9Ppn1t/LgUBicTaovx1Pv916891wABFdMDA8+SQnyIHDGxH8X2KHDDAYBRggCIKMEARBRh4nhzyU+SAge3IIT9lKQesAAP7mzy2JoYG1Jm4+J5OYmgAw1GAAYoowABFFGBgf5PH1sTQgDqTx9a8jhJgMDrgxpcvX04/QiIzHx8fp787O4uk693aFvv3jJn3mtt0wI2kB2SLtSZd79aOvvaZ95rbdMDhzudz9RKi2T8qKcDhLpdL1/9/9gLUu3/PmH2vuWYEAVBEAW58BOUVt1hr0vVu7ehrn3mvuc0IonHkt+IjmO16K9nreS2NnxTgRlJUSAytjxga1YwgGkkPiBhaHzE0qinAAEWMIMKJNvWxf1RSgMPJAfeRA6aSEUQjKSokhtZHDI1qOuDGbFGh2a63kr2e19LffnTAAEV0wI2krObROVb6JN1bPO+V51EH3Eh6QJLWivN6d6+crw44nG/WIZcCHO7IGBV9fFjSMoIAKKIAN5KymklrxXm9u1fO1wiiIVXAXtxbtBTgRlJUSAwtS9K9Rb9Hnk8jiEbSA5K0VpzXbB45bwUYoIgRRDjRJsilAIeTA87hw5KWEUQjKSqUtFac12weOW8dcEOqgL24t2jpgAGK6IAbSVlNOeAsSfcW27j3jOqAG0kPSNJacV4zunfmOuBwvlmHXApwODG0HD4saRlBABRRgBtJWc2kteK8ZnTvzI0gGlIF7MW9RUsHDFBEB9xIymrKAWdJurfYztoYQgfcSHpAktaK85rV2rnrgMOJNkEuBTicHHAOH5a0jCAAiijAjaSsZtJacV6zWjt3I4iGVAF7cW/Na2n8pAA3kqJCW8TQkq53a0fH+Gbe65mJoT0h6QHZYq1J17u1o6995r2e2dq5K8AARYwgwok29bF/VFKAw/XmgGcvQEfmqGffa64ZQTSSokJbrDXperd29LXPvNczE0N7wmxRodmut5K9ntfS3350wABFdMCNpKymHHAfOWCOIAf8hKQHRA64jxwwR/A6yjfmm/U+9o9KCnA4MbQ+YmhUMoIAKKIAN5KymnLAfeSAOYIc8BNmy2rOdr2V7PW8vI7yQUlRITG0PmJoHEEM7QlJD4gYWh8xNI7gdZQAAzKCCCfa1Mf+UUkBDicH3EcOmEpGEI2kqJAYWh8xNI4ghvaE2aJCs11vJXs9L6+jBBiMDriRlNWUA+4jB8wR5ICfkPSAyAH3kQPmCF5H+cZ8s97H/lFJAQ4nhtZHDI1KRhAARRTgRlJWUw64jxwwR5ADfsJsWc3ZrreSvZ6X11E+KCkqdHSMij5J9xbHMIJoJD0gSWvFeXFNAQYoYgQRTrQJcinA4Y7MsdLHhyUtI4hGUlQoaa04L67pgBtSBezFvTUvr6MEGIwOuJGU1ZQDzpJ0b3EMHXAj6QFJWivOi2s64HC+WYdcCnA4MbQcPixpGUEAFFGAG0lZzaS14ry4ZgTRkCpgL+6teXkd5YOSokJiaFmS7i2OYQTRSHpAktaK8+KaAgxQxAginGgT5FKAw8kB5/BhScsIopEUFUpaK86LazrghlQBe3FvzcvrKAEGowNuJGU1t8gBJ13v1o7OUc+819ymA24kPSBbrDXperd29LXPvNfcpgMO55v1PvaPSgpwuN4Y2uwF6MgY3+x7zTUjCIAiCnAjKau5xVqTrndrR1/7zHvNbUYQjdmymrNdbyV7PS+vo3xQUlTI6yizJN1bHMMIopH0gCStFefFNQUYoIgRRDjRJsilAIfzOsocPixpGUE0kqJCSWvFeXFNB9yQKmAv7q15eR0lwGB0wI2krKYccJake4tj6IAbSQ9I0lpxXlzTAYfzzTrkUoDDiaHl8GFJywgCoIgC3EjKaiatFefFNSOIhlQBe3FvzcvrKJ+UEBkSQ8uScE/xuC2ePyOIBQkPSsIa+eS83ssW56kAAxQxgggn2gS5FOBwcsA5fFjSMoJYkBAZSlgjn5zXe9niPHXAC6QL2Jp7ipYOGKCIDnhBQmZTDjhLwj3F9tZGFTrgBQkPSsIa+eS85rR27jrgcL5Zh1wKcDgxtBw+LGkZQQAUUYAXJGQ2E9bIJ+c1p7VzN4JYIF3A1txT8/I6yiclRIbE0LIk3FMcywhiQcKDkrBGPjkvWgowQBEjiHCiTZBLAQ4nB5zDhyUtI4gFCZGhhDXyyXnR0gEvkC5ga+6peS397UcHDFBEB7wgIbMpB5wl4Z7iWDrgBQkPSsIa+eS8aOmAw/lmHXIpwOHE0HL4sKRlBAFQRAFekJDZTFgjn5wXLSOIBdIFbM09NS+vo3xSQmRoixhawnXu5egY38x7zW1GEAsSHpQt1phwnXs5+tpn3mtuU4ABihhBhBNt6mP/qKQAh+vNAc9egI7MUc++11wzgliQEBnaYo0J17mXo6995r3mNh3wglkiQ7Nc5wjs9by8jhJgMDrgBQmZTa+jzJJwT3EsHfCChAclYY18cl60dMDhfLMOuRTgcF5HmcOHJS0jCIAiCvCChMxmwhr55LxoGUEskC5ga+6peXkd5QtGjw2JoWUZ/X7ieEYQK0Z/WEZfHz9zXrQUYIAiRhDhRJsglwIcTg44hw9LWkYQK0aPDY2+Pn7mvGjpgFdIGLAl99O8vI4SYDA64BWj5zb9LH0fP0tPNR3witEfFj9L38fP0lNNBxzON+t97B+VFOBwfhW5j19FppIRBEARBXjF6LlNP0vfx8/SU80IYsUMuc0ZrnEU9npeXkf5gtFjQ15HmWX0+4n7tn7mjCBWjP6wjL4+fua88m19hgowQBEjiHCiTZBLAQ7ndZQ5fFjSMoJYMXpsaPT18TPnlW/rM9QBr5AwYEvuJ1o6YIAiOuAVo+c25YCzjH4/cTwd8IrRH5bR18fPnBctHXA436xDLgU4nBhaDh+WtIwgAIoowCtGz22Ovj5+5rxoGUGskDBgS+6neXkd5QtGjw35VeQ+fhWZakYQK0Z/WPwqch+/ikw1BRigiBFEONGmPvaPSgpwOD9L38fP0lPJCGLF6LEhv4rcx68iU00HvGKG2NAM1zgKez2vpb/96IABiuiAV4ye25QD7iMHTDUd8IrRHxY54D5ywFTTAYfzzXof+0clBTicGFofMTQqGUEAFFGAV4ye25QD7iMHTDUjiBUz5DZnuMZR2Ot5eR3li0aODomh9RFDo5oRxB0jPzBiaH3E0KimAAMUMYIIJ9rUx/5RSQEOJwfcRw6YSkYQd4wcHRJD6yOGRjUd8B3vHh169+sbib2el9dRAgxGB3zHyNnNo3Os9Bn5XqKGDviOkR+YkdfGNedFSwcczjfrkEsBDndkjIo+PixpGUEAFFGA7xg5uzny2rjmvGgZQdwhZcBW3Evz8jrKF40cHfI6yj5eR0k1I4g7Rn5gvI6yj9dRUk0BBihiBBFOtKmP/aOSAhzO6yj7eB0llYwg7hg5OuR1lH28jpJqOuA73j069O7XNxJ7PS+vowQYjA74CaPlOOWA+8gBc4S10ZMO+AmjPTxywH3kgDnC2rnrgMP5Zr2P/aOSAhxODK2PGBqVjCAAiijATxgtxykH3EcOmCOsnbsRxBPeMcf5jtc0Kns9L6+j3NAocSK/ipxllPuGcRhBvGCUh2iUdfAY50VLAQYoYgQRTrQJcinA4fwsfQ4flrSMIF4wSpxolHXwGOdFSwf8AskDXuG+mZfXUQIMRgf8ohEynXLAWUa4ZzjOI8+nDvhFIzxII6yBxzmvuTxy3jrgcL5Zh1wKcDgxtBw+LGkZQQAUUYBfNEKmc4Q18DjnNZdHztsI4kXSBzzLPUNLAd7BUXEjMbQsYmi0jCB2cNRD5mHO4rxoKcAARYwgwok2QS4FOJwccA4flrSMIHZwVNxIrCmL86KlA96BZAK3uC/m5XWUAIPRAcMBZICn9/utP9QBwwEU3+ndrLUKMEARBRigiAIMUEQBhgPIAE/vn1t/KAUBB5ABntv5fP7r1p8rwDAw8bW3IYYGaRTftyGGBjASBRigiAIMUEQBBiiiAMPA5IffhhwwpJEffg9ywMAqmeNdyQEDyxTfXckBA4xEAQYoogADFFGAgdPpJPK2MzE0YJnI236WYmg6YIAiOmBgU/LEnz4+Plb/ZqEDBjal+H66txcKMEARBRigiAIMUEQBBjYlT/zp3l5IQQCbkid+nA4YoMj5crk8/g+fz/89nU7/2W85AG/p35fL5bf2D58qwABsxwgCoIgCDFBEAQYoogADFFGAAYoowABFFGCAIgowQBEFGKDI/wG2+NiArV2KQAAAAABJRU5ErkJggg==\n"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for file for 5tera_xrn-1-KD_smg-5_rerun, at /data16/marcus/working/230410_nanoporeRun_totalRNA_smg-5_xrn-1-KD_5TERA_rerun/output_dir/cat_files/*cat.sorted.mappedAndPrimary.bam... File Found.\n",
      "Working on rps-22 @ 01/25/24 @ 09:00:21 AM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2253it [00:00, 21383.87it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Saved to: /tmp/240125_rps-22_newerS5.bam\n",
      "    36 reads in newerS5 that match rps-22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Reads...: 100%|██████████| 36/36 [00:00<00:00, 1664.46it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved images to: /home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python/240125_readPlots/rps-22/240125_newerS5_rps-22.readPlotting.png/svg\n"
     ]
    },
    {
     "data": {
      "text/plain": "<Figure size 360x248.832 with 1 Axes>",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAADxCAYAAAD1LG0eAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAJdUlEQVR4nO3dwY7byBEGYGqR6/jkPcevYOz7v4ENP8Oek5N9t3IIkGBaGkozLXVVsb7vKAwwTTZZKjV/NE/n83kDYL0/ogcA0JUCDBBEAQYIogADBFGAAYIowABB/vGeP/78+fP5y5cvTxoK4b5/jx7Ben/9te5/dTy/bNu2bd+37d/n8/nP8fN3FeAvX75s3759e9yoyOXTp2379St6FOu8vGzbyuu52/nlf07b9ve1z99VgDm4nz+jR3Bszm9fp9PVjxVg1svSCb68rC2KWY6bNDyEY70sRWj1OLIcN2nogOntjZ+GsIICTG8rN6NS7BlYggAIogCz3stL9Aj+a/U4shw3aViCYL2ucayux40YGglFx7LE0AhmCYI40cVIDI1gCjBAEEsQ9CYaRiAFmN7kgAlkCYI40bEsMTSC6YCJ0y2W1e14+T8xNEpZEdkSQ2OFnV8+liDIaUWhEkNjhZ15V4ABgliCoDfJBAIpwPQmhkYgSxDktCKyJYbGCjvzrgMmpyNGto54TNxHDA2CiaH1JIYGCSi+PYmhAeSjAAMEUYABgijAsIoYWk9iaJCAGFpfYmhQiMjacYihQTGK73GIoQHkowADBFGAAYIowJCRyNpxiKFBMSJrxyKGBo2IsaXyddu+XvvcEgQckeKbyh9v1FoFGCCIAgwQRAEGCKIAwxGJsaXye9t+X/tcCgKOSIwtlR+n049rnyvAwGOJwL1mNzRgGcX3NbuhAeSjAAMEUYABgijAwGOJwL1mNzRgGRG4S2/shqYDBgiiAwYeSw74gu0ogTUU3wu2owRIRgEGCKIAA48lhnbBbmjAGmJoF97aDU0HDBBEBwys1TCmJoYG5NCs+G6bGBpAOgowQBAFGFirYUxNDA3IoWFMTQwNIBkd8ODTp0/bryJPaV9eXrafDbuJqipdWzzOi7ci36/SDVJprJivrvbmXQdc3OmNnfaB/BTg4s7nc/QQuJMvS0aWIAZ76zXZVBor5qurvXnXAQ881OJZXFt9vfXrRwcMEEQHPKgUFXpEDK3S8T7a6hhf53PNdTrgQaUb5BFjrXS8j7b62Dufa67TARfnyfoc549ICnBxszG07gVoZYyv+7nmkiWIQaWo0CPGWul4H231sXc+11ynAx50iwp1O95IznVfYmgAyeiAB5WiQnZDq6XStcWce+9NHfCg0g1SaayYr07unWsdcHGerENdCnBxdkOrw5clI0sQAEEU4EGlrGalsWK+Orl3ri1BDKQKeBbXFiMFeFApKiSGVkula4t599yfliAGlW6QSmPFfHVzz3zrgIvzZB3qUoCLE0Orw5clI0sQAEEU4EGlqFClsWK+urlnvi1BDKQKeBbXFiMFeFApKuSlnHO8lJNnu3WNWYIYVLpBvJRzjpdy8my35lwHXJwn63OcPyIpwMV5KeccL+UkkiUIgCAK8KBSVMhbked4KzLPdmvOLUEMukWFuh1vJOeakQI8qBQVEkObI4bGCntdsCWIQaUbRAxtjhgaK+zNuw64OE/W5zh/RFKAixNDmyOGRiRLEABBFOBBpaiQGNocMTRW2Jt3SxCDblGhbscbybnu663lJwV4UCkq5KWctVS6tljDEsSg0g1SaayYLy7pgIvzZB3qUoCL81LOOnxZMrIEARBEAR5UigpVGivmi0uWIAZSBTyLa6svMbQ7VYoK2Q1tjt3QiGYJYlDpBrEb2hy7oRFNB1ycJ+tznD8i6YABguiAi7Md5RzbURJJBzyoFBWyG9ocu6ERTQc86BYV6na8kZzrvsTQ7lQpKmQ3tFoqXVusYQliUOkGqTRWzBeXFGCAIJYgivNkHepSgIuzHWUdviwZWYIYVIoKVRor5otLOuCBVAHP4trqSwztA7LHhsTQasl+PfEce798LEHsyH6zZB8fr5mvnvbmXQEGCGIJojhP1qEuBbg4MbQ6fFkysgSxI3tsKPv4eM189bQ37zrgHRIGPJLrqS8xtA/IHhvyUs45XsrJCmJoH5T9ZvFSzjleyskKYmgACVmCKM6T9TnOH5EU4OK8lHOOl3ISyRLEjuyxIS/lnOOlnKwghvZBHWJDHY4xC+e6LzG0D8geG7IbWi3ZryeeQwztg7LfLNnHx2vmqycxNICELEEU58k61KUAF2c3tDp8WTKyBLEje2wo+/h4zXz1JIb2QRIGPJLrqS8xtA/IHhuyG9ocu6H1lSXCaQliR/abxW5oc+yG1leWuVCAAYJYgijOk/U5zh+RFODi7IY2x25oRLIEsSN7bMhuaHPshtZXlrnQAe/I8JT02TocYxbONSMdMEAQHfCO7LnNLFlG7pP9euKx7rk/dcA7st8s2cfHa+arl3vmWwdcnCfrUJcCXJzd0OrwZcnIEsSOLFGVt2QfH6+Zr17umW8d8A4PuHgk1xMjHTBAEB3wDZmjQ7ajnGM7Sp7t1jWmA74h8w1jO8o5tqPk2W7NuQ64OE/W5zh/RFKAi7Mb2hy7oRHJEsQNmaNDdkObYzc0nu3WnOuAbzh6dOjox5eJc81IBwwQRAd8Q+bokN3Qasl8LfE8e8sQOuAbMt8wmcfGJfPV096864CL82Qd6lKAi7MbWh2+LBlZgrghc3Qo89i4ZL562pt3HfANHnLxKK6lvt769aMDBgiiA75D1viQ3dDm2A2NaDrgO2S9aeyGNsduaETTARfnyfoc549ICnBxdkObYzc0IlmCuEPW+JDd0ObYDY1oOuA7HDk+dORjy8a57ksMDSAZHfCE6FiR3dBqib5eyEcHPCH6Zor+/7yP+WKkAy7Ok3WoSwEuzm5odfiyZGQJAiCIAjwhOtcZ/f95H/PFyBLEBAkE3sP10tdby08KcHOdo1F2QyOaJYjmOhcEu6ERTQdMa5IJRFKAac1uaESyBAEQRAFurnM0ynaURLME0Zxo1DrOdV9iaHzYUeNTYmhEswTBTUctGmJoRNMB05pkApEUYFoTQyOSJQiAIAowNx01PiWGRjRLENwkPvUYzmNfYmgQSAStr71fPpYgYAHFt6+9uVeAAYIowABBFGCAIAowLCCC1tfe3EtBwAIiaL2JoUFB4muH8fXah5YgIDHF9zCu1loFGCCIAgwQRAEGCKIAQ2Lia4fx+9qHUhCQmPjaMZxOpx/XPtcBAwQ5veeVLKfT6V/btv39vOEAHNI/z+fzn+OH7yrAADyOJQiAIAowQBAFGCCIAgwQRAEGCKIAAwRRgAGCKMAAQRRggCD/AWC0h9u7LjbDAAAAAElFTkSuQmCC\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "sup_save_dir = f\"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python/{get_dt()}_readPlots\"\n",
    "for gene in [# \"ets-4\",\n",
    "             # \"xbp-1\",\n",
    "             # \"zip-1\",\n",
    "             \"rps-22\",\n",
    "             ]:\n",
    "    save_dir = f\"{sup_save_dir}/{gene}\"\n",
    "    Path(save_dir).mkdir(exist_ok=True, parents=True)\n",
    "    for lib in [\"newerN2\",\n",
    "                # \"oldN2\",\n",
    "                \"newerS6\",\n",
    "                # \"oldS6\",\n",
    "                \"newerS5\",\n",
    "                ]:\n",
    "        new_plot_reads_from_bam(lib, gene,\n",
    "                                subset_fraction={\"+\": 0.75, \"-\": 0.01},\n",
    "                                save_dir=save_dir,\n",
    "                                spacing_factor=1.2,)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-01-25T17:00:22.010710367Z",
     "start_time": "2024-01-25T17:00:19.892497757Z"
    }
   }
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "execution_count": null
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
